// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/reads.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A bucket over which read coverage has been precomputed. A bucket corresponds
 * to a specific range of the reference sequence.
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.CoverageBucket}
 */
public final class CoverageBucket extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.CoverageBucket)
    CoverageBucketOrBuilder {
private static final long serialVersionUID = 0L;
  // Use CoverageBucket.newBuilder() to construct.
  private CoverageBucket(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private CoverageBucket() {
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new CoverageBucket();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_CoverageBucket_descriptor;
  }

  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_CoverageBucket_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.CoverageBucket.class, com.google.genomics.v1.CoverageBucket.Builder.class);
  }

  public static final int RANGE_FIELD_NUMBER = 1;
  private com.google.genomics.v1.Range range_;
  /**
   * <pre>
   * The genomic coordinate range spanned by this bucket.
   * </pre>
   *
   * <code>.google.genomics.v1.Range range = 1;</code>
   * @return Whether the range field is set.
   */
  @java.lang.Override
  public boolean hasRange() {
    return range_ != null;
  }
  /**
   * <pre>
   * The genomic coordinate range spanned by this bucket.
   * </pre>
   *
   * <code>.google.genomics.v1.Range range = 1;</code>
   * @return The range.
   */
  @java.lang.Override
  public com.google.genomics.v1.Range getRange() {
    return range_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : range_;
  }
  /**
   * <pre>
   * The genomic coordinate range spanned by this bucket.
   * </pre>
   *
   * <code>.google.genomics.v1.Range range = 1;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.RangeOrBuilder getRangeOrBuilder() {
    return range_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : range_;
  }

  public static final int MEAN_COVERAGE_FIELD_NUMBER = 2;
  private float meanCoverage_ = 0F;
  /**
   * <pre>
   * The average number of reads which are aligned to each individual
   * reference base in this bucket.
   * </pre>
   *
   * <code>float mean_coverage = 2;</code>
   * @return The meanCoverage.
   */
  @java.lang.Override
  public float getMeanCoverage() {
    return meanCoverage_;
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (range_ != null) {
      output.writeMessage(1, getRange());
    }
    if (java.lang.Float.floatToRawIntBits(meanCoverage_) != 0) {
      output.writeFloat(2, meanCoverage_);
    }
    getUnknownFields().writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (range_ != null) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(1, getRange());
    }
    if (java.lang.Float.floatToRawIntBits(meanCoverage_) != 0) {
      size += com.google.protobuf.CodedOutputStream
        .computeFloatSize(2, meanCoverage_);
    }
    size += getUnknownFields().getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.CoverageBucket)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.CoverageBucket other = (com.google.genomics.v1.CoverageBucket) obj;

    if (hasRange() != other.hasRange()) return false;
    if (hasRange()) {
      if (!getRange()
          .equals(other.getRange())) return false;
    }
    if (java.lang.Float.floatToIntBits(getMeanCoverage())
        != java.lang.Float.floatToIntBits(
            other.getMeanCoverage())) return false;
    if (!getUnknownFields().equals(other.getUnknownFields())) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    if (hasRange()) {
      hash = (37 * hash) + RANGE_FIELD_NUMBER;
      hash = (53 * hash) + getRange().hashCode();
    }
    hash = (37 * hash) + MEAN_COVERAGE_FIELD_NUMBER;
    hash = (53 * hash) + java.lang.Float.floatToIntBits(
        getMeanCoverage());
    hash = (29 * hash) + getUnknownFields().hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.CoverageBucket parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.CoverageBucket parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.CoverageBucket parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.CoverageBucket parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.CoverageBucket prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A bucket over which read coverage has been precomputed. A bucket corresponds
   * to a specific range of the reference sequence.
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.CoverageBucket}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.CoverageBucket)
      com.google.genomics.v1.CoverageBucketOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_CoverageBucket_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_CoverageBucket_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.CoverageBucket.class, com.google.genomics.v1.CoverageBucket.Builder.class);
    }

    // Construct using com.google.genomics.v1.CoverageBucket.newBuilder()
    private Builder() {

    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);

    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      bitField0_ = 0;
      range_ = null;
      if (rangeBuilder_ != null) {
        rangeBuilder_.dispose();
        rangeBuilder_ = null;
      }
      meanCoverage_ = 0F;
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_CoverageBucket_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.CoverageBucket getDefaultInstanceForType() {
      return com.google.genomics.v1.CoverageBucket.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.CoverageBucket build() {
      com.google.genomics.v1.CoverageBucket result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.CoverageBucket buildPartial() {
      com.google.genomics.v1.CoverageBucket result = new com.google.genomics.v1.CoverageBucket(this);
      if (bitField0_ != 0) { buildPartial0(result); }
      onBuilt();
      return result;
    }

    private void buildPartial0(com.google.genomics.v1.CoverageBucket result) {
      int from_bitField0_ = bitField0_;
      if (((from_bitField0_ & 0x00000001) != 0)) {
        result.range_ = rangeBuilder_ == null
            ? range_
            : rangeBuilder_.build();
      }
      if (((from_bitField0_ & 0x00000002) != 0)) {
        result.meanCoverage_ = meanCoverage_;
      }
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.CoverageBucket) {
        return mergeFrom((com.google.genomics.v1.CoverageBucket)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.CoverageBucket other) {
      if (other == com.google.genomics.v1.CoverageBucket.getDefaultInstance()) return this;
      if (other.hasRange()) {
        mergeRange(other.getRange());
      }
      if (other.getMeanCoverage() != 0F) {
        setMeanCoverage(other.getMeanCoverage());
      }
      this.mergeUnknownFields(other.getUnknownFields());
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 10: {
              input.readMessage(
                  getRangeFieldBuilder().getBuilder(),
                  extensionRegistry);
              bitField0_ |= 0x00000001;
              break;
            } // case 10
            case 21: {
              meanCoverage_ = input.readFloat();
              bitField0_ |= 0x00000002;
              break;
            } // case 21
            default: {
              if (!super.parseUnknownField(input, extensionRegistry, tag)) {
                done = true; // was an endgroup tag
              }
              break;
            } // default:
          } // switch (tag)
        } // while (!done)
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.unwrapIOException();
      } finally {
        onChanged();
      } // finally
      return this;
    }
    private int bitField0_;

    private com.google.genomics.v1.Range range_;
    private com.google.protobuf.SingleFieldBuilderV3<
        com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder> rangeBuilder_;
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     * @return Whether the range field is set.
     */
    public boolean hasRange() {
      return ((bitField0_ & 0x00000001) != 0);
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     * @return The range.
     */
    public com.google.genomics.v1.Range getRange() {
      if (rangeBuilder_ == null) {
        return range_ == null ? com.google.genomics.v1.Range.getDefaultInstance() : range_;
      } else {
        return rangeBuilder_.getMessage();
      }
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    public Builder setRange(com.google.genomics.v1.Range value) {
      if (rangeBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        range_ = value;
      } else {
        rangeBuilder_.setMessage(value);
      }
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    public Builder setRange(
        com.google.genomics.v1.Range.Builder builderForValue) {
      if (rangeBuilder_ == null) {
        range_ = builderForValue.build();
      } else {
        rangeBuilder_.setMessage(builderForValue.build());
      }
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    public Builder mergeRange(com.google.genomics.v1.Range value) {
      if (rangeBuilder_ == null) {
        if (((bitField0_ & 0x00000001) != 0) &&
          range_ != null &&
          range_ != com.google.genomics.v1.Range.getDefaultInstance()) {
          getRangeBuilder().mergeFrom(value);
        } else {
          range_ = value;
        }
      } else {
        rangeBuilder_.mergeFrom(value);
      }
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    public Builder clearRange() {
      bitField0_ = (bitField0_ & ~0x00000001);
      range_ = null;
      if (rangeBuilder_ != null) {
        rangeBuilder_.dispose();
        rangeBuilder_ = null;
      }
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    public com.google.genomics.v1.Range.Builder getRangeBuilder() {
      bitField0_ |= 0x00000001;
      onChanged();
      return getRangeFieldBuilder().getBuilder();
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    public com.google.genomics.v1.RangeOrBuilder getRangeOrBuilder() {
      if (rangeBuilder_ != null) {
        return rangeBuilder_.getMessageOrBuilder();
      } else {
        return range_ == null ?
            com.google.genomics.v1.Range.getDefaultInstance() : range_;
      }
    }
    /**
     * <pre>
     * The genomic coordinate range spanned by this bucket.
     * </pre>
     *
     * <code>.google.genomics.v1.Range range = 1;</code>
     */
    private com.google.protobuf.SingleFieldBuilderV3<
        com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder> 
        getRangeFieldBuilder() {
      if (rangeBuilder_ == null) {
        rangeBuilder_ = new com.google.protobuf.SingleFieldBuilderV3<
            com.google.genomics.v1.Range, com.google.genomics.v1.Range.Builder, com.google.genomics.v1.RangeOrBuilder>(
                getRange(),
                getParentForChildren(),
                isClean());
        range_ = null;
      }
      return rangeBuilder_;
    }

    private float meanCoverage_ ;
    /**
     * <pre>
     * The average number of reads which are aligned to each individual
     * reference base in this bucket.
     * </pre>
     *
     * <code>float mean_coverage = 2;</code>
     * @return The meanCoverage.
     */
    @java.lang.Override
    public float getMeanCoverage() {
      return meanCoverage_;
    }
    /**
     * <pre>
     * The average number of reads which are aligned to each individual
     * reference base in this bucket.
     * </pre>
     *
     * <code>float mean_coverage = 2;</code>
     * @param value The meanCoverage to set.
     * @return This builder for chaining.
     */
    public Builder setMeanCoverage(float value) {
      
      meanCoverage_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The average number of reads which are aligned to each individual
     * reference base in this bucket.
     * </pre>
     *
     * <code>float mean_coverage = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearMeanCoverage() {
      bitField0_ = (bitField0_ & ~0x00000002);
      meanCoverage_ = 0F;
      onChanged();
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.CoverageBucket)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.CoverageBucket)
  private static final com.google.genomics.v1.CoverageBucket DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.CoverageBucket();
  }

  public static com.google.genomics.v1.CoverageBucket getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<CoverageBucket>
      PARSER = new com.google.protobuf.AbstractParser<CoverageBucket>() {
    @java.lang.Override
    public CoverageBucket parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      Builder builder = newBuilder();
      try {
        builder.mergeFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(builder.buildPartial());
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(builder.buildPartial());
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(e)
            .setUnfinishedMessage(builder.buildPartial());
      }
      return builder.buildPartial();
    }
  };

  public static com.google.protobuf.Parser<CoverageBucket> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<CoverageBucket> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.CoverageBucket getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

