// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/reads.proto

package com.google.genomics.v1;

public interface ImportReadGroupSetsRequestOrBuilder extends
    // @@protoc_insertion_point(interface_extends:google.genomics.v1.ImportReadGroupSetsRequest)
    com.google.protobuf.MessageOrBuilder {

  /**
   * <pre>
   * Required. The ID of the dataset these read group sets will belong to. The
   * caller must have WRITE permissions to this dataset.
   * </pre>
   *
   * <code>string dataset_id = 1;</code>
   * @return The datasetId.
   */
  java.lang.String getDatasetId();
  /**
   * <pre>
   * Required. The ID of the dataset these read group sets will belong to. The
   * caller must have WRITE permissions to this dataset.
   * </pre>
   *
   * <code>string dataset_id = 1;</code>
   * @return The bytes for datasetId.
   */
  com.google.protobuf.ByteString
      getDatasetIdBytes();

  /**
   * <pre>
   * The reference set to which the imported read group sets are aligned to, if
   * any. The reference names of this reference set must be a superset of those
   * found in the imported file headers. If no reference set id is provided, a
   * best effort is made to associate with a matching reference set.
   * </pre>
   *
   * <code>string reference_set_id = 4;</code>
   * @return The referenceSetId.
   */
  java.lang.String getReferenceSetId();
  /**
   * <pre>
   * The reference set to which the imported read group sets are aligned to, if
   * any. The reference names of this reference set must be a superset of those
   * found in the imported file headers. If no reference set id is provided, a
   * best effort is made to associate with a matching reference set.
   * </pre>
   *
   * <code>string reference_set_id = 4;</code>
   * @return The bytes for referenceSetId.
   */
  com.google.protobuf.ByteString
      getReferenceSetIdBytes();

  /**
   * <pre>
   * A list of URIs pointing at [BAM
   * files](https://samtools.github.io/hts-specs/SAMv1.pdf)
   * in Google Cloud Storage.
   * Those URIs can include wildcards (*), but do not add or remove
   * matching files before import has completed.
   * Note that Google Cloud Storage object listing is only eventually
   * consistent: files added may be not be immediately visible to
   * everyone. Thus, if using a wildcard it is preferable not to start
   * the import immediately after the files are created.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @return A list containing the sourceUris.
   */
  java.util.List<java.lang.String>
      getSourceUrisList();
  /**
   * <pre>
   * A list of URIs pointing at [BAM
   * files](https://samtools.github.io/hts-specs/SAMv1.pdf)
   * in Google Cloud Storage.
   * Those URIs can include wildcards (*), but do not add or remove
   * matching files before import has completed.
   * Note that Google Cloud Storage object listing is only eventually
   * consistent: files added may be not be immediately visible to
   * everyone. Thus, if using a wildcard it is preferable not to start
   * the import immediately after the files are created.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @return The count of sourceUris.
   */
  int getSourceUrisCount();
  /**
   * <pre>
   * A list of URIs pointing at [BAM
   * files](https://samtools.github.io/hts-specs/SAMv1.pdf)
   * in Google Cloud Storage.
   * Those URIs can include wildcards (*), but do not add or remove
   * matching files before import has completed.
   * Note that Google Cloud Storage object listing is only eventually
   * consistent: files added may be not be immediately visible to
   * everyone. Thus, if using a wildcard it is preferable not to start
   * the import immediately after the files are created.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @param index The index of the element to return.
   * @return The sourceUris at the given index.
   */
  java.lang.String getSourceUris(int index);
  /**
   * <pre>
   * A list of URIs pointing at [BAM
   * files](https://samtools.github.io/hts-specs/SAMv1.pdf)
   * in Google Cloud Storage.
   * Those URIs can include wildcards (*), but do not add or remove
   * matching files before import has completed.
   * Note that Google Cloud Storage object listing is only eventually
   * consistent: files added may be not be immediately visible to
   * everyone. Thus, if using a wildcard it is preferable not to start
   * the import immediately after the files are created.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @param index The index of the value to return.
   * @return The bytes of the sourceUris at the given index.
   */
  com.google.protobuf.ByteString
      getSourceUrisBytes(int index);

  /**
   * <pre>
   * The partition strategy describes how read groups are partitioned into read
   * group sets.
   * </pre>
   *
   * <code>.google.genomics.v1.ImportReadGroupSetsRequest.PartitionStrategy partition_strategy = 5;</code>
   * @return The enum numeric value on the wire for partitionStrategy.
   */
  int getPartitionStrategyValue();
  /**
   * <pre>
   * The partition strategy describes how read groups are partitioned into read
   * group sets.
   * </pre>
   *
   * <code>.google.genomics.v1.ImportReadGroupSetsRequest.PartitionStrategy partition_strategy = 5;</code>
   * @return The partitionStrategy.
   */
  com.google.genomics.v1.ImportReadGroupSetsRequest.PartitionStrategy getPartitionStrategy();
}
