// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/variants.proto

package com.google.genomics.v1;

public interface ImportVariantsRequestOrBuilder extends
    // @@protoc_insertion_point(interface_extends:google.genomics.v1.ImportVariantsRequest)
    com.google.protobuf.MessageOrBuilder {

  /**
   * <pre>
   * Required. The variant set to which variant data should be imported.
   * </pre>
   *
   * <code>string variant_set_id = 1;</code>
   * @return The variantSetId.
   */
  java.lang.String getVariantSetId();
  /**
   * <pre>
   * Required. The variant set to which variant data should be imported.
   * </pre>
   *
   * <code>string variant_set_id = 1;</code>
   * @return The bytes for variantSetId.
   */
  com.google.protobuf.ByteString
      getVariantSetIdBytes();

  /**
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described
   * here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @return A list containing the sourceUris.
   */
  java.util.List<java.lang.String>
      getSourceUrisList();
  /**
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described
   * here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @return The count of sourceUris.
   */
  int getSourceUrisCount();
  /**
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described
   * here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @param index The index of the element to return.
   * @return The sourceUris at the given index.
   */
  java.lang.String getSourceUris(int index);
  /**
   * <pre>
   * A list of URIs referencing variant files in Google Cloud Storage. URIs can
   * include wildcards [as described
   * here](https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames).
   * Note that recursive wildcards ('**') are not supported.
   * </pre>
   *
   * <code>repeated string source_uris = 2;</code>
   * @param index The index of the value to return.
   * @return The bytes of the sourceUris at the given index.
   */
  com.google.protobuf.ByteString
      getSourceUrisBytes(int index);

  /**
   * <pre>
   * The format of the variant data being imported. If unspecified, defaults to
   * to `VCF`.
   * </pre>
   *
   * <code>.google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
   * @return The enum numeric value on the wire for format.
   */
  int getFormatValue();
  /**
   * <pre>
   * The format of the variant data being imported. If unspecified, defaults to
   * to `VCF`.
   * </pre>
   *
   * <code>.google.genomics.v1.ImportVariantsRequest.Format format = 3;</code>
   * @return The format.
   */
  com.google.genomics.v1.ImportVariantsRequest.Format getFormat();

  /**
   * <pre>
   * Convert reference names to the canonical representation.
   * hg19 haploytypes (those reference names containing "_hap")
   * are not modified in any way.
   * All other reference names are modified according to the following rules:
   * The reference name is capitalized.
   * The "chr" prefix is dropped for all autosomes and sex chromsomes.
   * For example "chr17" becomes "17" and "chrX" becomes "X".
   * All mitochondrial chromosomes ("chrM", "chrMT", etc) become "MT".
   * </pre>
   *
   * <code>bool normalize_reference_names = 5;</code>
   * @return The normalizeReferenceNames.
   */
  boolean getNormalizeReferenceNames();

  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */
  int getInfoMergeConfigCount();
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */
  boolean containsInfoMergeConfig(
      java.lang.String key);
  /**
   * Use {@link #getInfoMergeConfigMap()} instead.
   */
  @java.lang.Deprecated
  java.util.Map<java.lang.String, com.google.genomics.v1.InfoMergeOperation>
  getInfoMergeConfig();
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */
  java.util.Map<java.lang.String, com.google.genomics.v1.InfoMergeOperation>
  getInfoMergeConfigMap();
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */
  /* nullable */
com.google.genomics.v1.InfoMergeOperation getInfoMergeConfigOrDefault(
      java.lang.String key,
      /* nullable */
com.google.genomics.v1.InfoMergeOperation         defaultValue);
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */
  com.google.genomics.v1.InfoMergeOperation getInfoMergeConfigOrThrow(
      java.lang.String key);
  /**
   * Use {@link #getInfoMergeConfigValueMap()} instead.
   */
  @java.lang.Deprecated
  java.util.Map<java.lang.String, java.lang.Integer>
  getInfoMergeConfigValue();
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */
  java.util.Map<java.lang.String, java.lang.Integer>
  getInfoMergeConfigValueMap();
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */

  int getInfoMergeConfigValueOrDefault(
      java.lang.String key,
      int defaultValue);
  /**
   * <pre>
   * A mapping between info field keys and the InfoMergeOperations to
   * be performed on them. This is plumbed down to the MergeVariantRequests
   * generated by the resulting import job.
   * </pre>
   *
   * <code>map&lt;string, .google.genomics.v1.InfoMergeOperation&gt; info_merge_config = 6;</code>
   */

  int getInfoMergeConfigValueOrThrow(
      java.lang.String key);
}
