// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/readalignment.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A linear alignment can be represented by one CIGAR string. Describes the
 * mapped position and local alignment of the read to the reference.
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.LinearAlignment}
 */
public final class LinearAlignment extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.LinearAlignment)
    LinearAlignmentOrBuilder {
private static final long serialVersionUID = 0L;
  // Use LinearAlignment.newBuilder() to construct.
  private LinearAlignment(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private LinearAlignment() {
    cigar_ = java.util.Collections.emptyList();
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new LinearAlignment();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  private LinearAlignment(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    this();
    if (extensionRegistry == null) {
      throw new java.lang.NullPointerException();
    }
    int mutable_bitField0_ = 0;
    com.google.protobuf.UnknownFieldSet.Builder unknownFields =
        com.google.protobuf.UnknownFieldSet.newBuilder();
    try {
      boolean done = false;
      while (!done) {
        int tag = input.readTag();
        switch (tag) {
          case 0:
            done = true;
            break;
          case 10: {
            com.google.genomics.v1.Position.Builder subBuilder = null;
            if (position_ != null) {
              subBuilder = position_.toBuilder();
            }
            position_ = input.readMessage(com.google.genomics.v1.Position.parser(), extensionRegistry);
            if (subBuilder != null) {
              subBuilder.mergeFrom(position_);
              position_ = subBuilder.buildPartial();
            }

            break;
          }
          case 16: {

            mappingQuality_ = input.readInt32();
            break;
          }
          case 26: {
            if (!((mutable_bitField0_ & 0x00000001) != 0)) {
              cigar_ = new java.util.ArrayList<com.google.genomics.v1.CigarUnit>();
              mutable_bitField0_ |= 0x00000001;
            }
            cigar_.add(
                input.readMessage(com.google.genomics.v1.CigarUnit.parser(), extensionRegistry));
            break;
          }
          default: {
            if (!parseUnknownField(
                input, unknownFields, extensionRegistry, tag)) {
              done = true;
            }
            break;
          }
        }
      }
    } catch (com.google.protobuf.InvalidProtocolBufferException e) {
      throw e.setUnfinishedMessage(this);
    } catch (com.google.protobuf.UninitializedMessageException e) {
      throw e.asInvalidProtocolBufferException().setUnfinishedMessage(this);
    } catch (java.io.IOException e) {
      throw new com.google.protobuf.InvalidProtocolBufferException(
          e).setUnfinishedMessage(this);
    } finally {
      if (((mutable_bitField0_ & 0x00000001) != 0)) {
        cigar_ = java.util.Collections.unmodifiableList(cigar_);
      }
      this.unknownFields = unknownFields.build();
      makeExtensionsImmutable();
    }
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReadAlignmentProto.internal_static_google_genomics_v1_LinearAlignment_descriptor;
  }

  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReadAlignmentProto.internal_static_google_genomics_v1_LinearAlignment_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.LinearAlignment.class, com.google.genomics.v1.LinearAlignment.Builder.class);
  }

  public static final int POSITION_FIELD_NUMBER = 1;
  private com.google.genomics.v1.Position position_;
  /**
   * <pre>
   * The position of this alignment.
   * </pre>
   *
   * <code>.google.genomics.v1.Position position = 1;</code>
   * @return Whether the position field is set.
   */
  @java.lang.Override
  public boolean hasPosition() {
    return position_ != null;
  }
  /**
   * <pre>
   * The position of this alignment.
   * </pre>
   *
   * <code>.google.genomics.v1.Position position = 1;</code>
   * @return The position.
   */
  @java.lang.Override
  public com.google.genomics.v1.Position getPosition() {
    return position_ == null ? com.google.genomics.v1.Position.getDefaultInstance() : position_;
  }
  /**
   * <pre>
   * The position of this alignment.
   * </pre>
   *
   * <code>.google.genomics.v1.Position position = 1;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.PositionOrBuilder getPositionOrBuilder() {
    return getPosition();
  }

  public static final int MAPPING_QUALITY_FIELD_NUMBER = 2;
  private int mappingQuality_;
  /**
   * <pre>
   * The mapping quality of this alignment. Represents how likely
   * the read maps to this position as opposed to other locations.
   * Specifically, this is -10 log10 Pr(mapping position is wrong), rounded to
   * the nearest integer.
   * </pre>
   *
   * <code>int32 mapping_quality = 2;</code>
   * @return The mappingQuality.
   */
  @java.lang.Override
  public int getMappingQuality() {
    return mappingQuality_;
  }

  public static final int CIGAR_FIELD_NUMBER = 3;
  private java.util.List<com.google.genomics.v1.CigarUnit> cigar_;
  /**
   * <pre>
   * Represents the local alignment of this sequence (alignment matches, indels,
   * etc) against the reference.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
   */
  @java.lang.Override
  public java.util.List<com.google.genomics.v1.CigarUnit> getCigarList() {
    return cigar_;
  }
  /**
   * <pre>
   * Represents the local alignment of this sequence (alignment matches, indels,
   * etc) against the reference.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
   */
  @java.lang.Override
  public java.util.List<? extends com.google.genomics.v1.CigarUnitOrBuilder> 
      getCigarOrBuilderList() {
    return cigar_;
  }
  /**
   * <pre>
   * Represents the local alignment of this sequence (alignment matches, indels,
   * etc) against the reference.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
   */
  @java.lang.Override
  public int getCigarCount() {
    return cigar_.size();
  }
  /**
   * <pre>
   * Represents the local alignment of this sequence (alignment matches, indels,
   * etc) against the reference.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.CigarUnit getCigar(int index) {
    return cigar_.get(index);
  }
  /**
   * <pre>
   * Represents the local alignment of this sequence (alignment matches, indels,
   * etc) against the reference.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.CigarUnitOrBuilder getCigarOrBuilder(
      int index) {
    return cigar_.get(index);
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (position_ != null) {
      output.writeMessage(1, getPosition());
    }
    if (mappingQuality_ != 0) {
      output.writeInt32(2, mappingQuality_);
    }
    for (int i = 0; i < cigar_.size(); i++) {
      output.writeMessage(3, cigar_.get(i));
    }
    unknownFields.writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (position_ != null) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(1, getPosition());
    }
    if (mappingQuality_ != 0) {
      size += com.google.protobuf.CodedOutputStream
        .computeInt32Size(2, mappingQuality_);
    }
    for (int i = 0; i < cigar_.size(); i++) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(3, cigar_.get(i));
    }
    size += unknownFields.getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.LinearAlignment)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.LinearAlignment other = (com.google.genomics.v1.LinearAlignment) obj;

    if (hasPosition() != other.hasPosition()) return false;
    if (hasPosition()) {
      if (!getPosition()
          .equals(other.getPosition())) return false;
    }
    if (getMappingQuality()
        != other.getMappingQuality()) return false;
    if (!getCigarList()
        .equals(other.getCigarList())) return false;
    if (!unknownFields.equals(other.unknownFields)) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    if (hasPosition()) {
      hash = (37 * hash) + POSITION_FIELD_NUMBER;
      hash = (53 * hash) + getPosition().hashCode();
    }
    hash = (37 * hash) + MAPPING_QUALITY_FIELD_NUMBER;
    hash = (53 * hash) + getMappingQuality();
    if (getCigarCount() > 0) {
      hash = (37 * hash) + CIGAR_FIELD_NUMBER;
      hash = (53 * hash) + getCigarList().hashCode();
    }
    hash = (29 * hash) + unknownFields.hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.LinearAlignment parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.LinearAlignment parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.LinearAlignment parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.LinearAlignment parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.LinearAlignment prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A linear alignment can be represented by one CIGAR string. Describes the
   * mapped position and local alignment of the read to the reference.
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.LinearAlignment}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.LinearAlignment)
      com.google.genomics.v1.LinearAlignmentOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadAlignmentProto.internal_static_google_genomics_v1_LinearAlignment_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadAlignmentProto.internal_static_google_genomics_v1_LinearAlignment_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.LinearAlignment.class, com.google.genomics.v1.LinearAlignment.Builder.class);
    }

    // Construct using com.google.genomics.v1.LinearAlignment.newBuilder()
    private Builder() {
      maybeForceBuilderInitialization();
    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);
      maybeForceBuilderInitialization();
    }
    private void maybeForceBuilderInitialization() {
      if (com.google.protobuf.GeneratedMessageV3
              .alwaysUseFieldBuilders) {
        getCigarFieldBuilder();
      }
    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      if (positionBuilder_ == null) {
        position_ = null;
      } else {
        position_ = null;
        positionBuilder_ = null;
      }
      mappingQuality_ = 0;

      if (cigarBuilder_ == null) {
        cigar_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
      } else {
        cigarBuilder_.clear();
      }
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReadAlignmentProto.internal_static_google_genomics_v1_LinearAlignment_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.LinearAlignment getDefaultInstanceForType() {
      return com.google.genomics.v1.LinearAlignment.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.LinearAlignment build() {
      com.google.genomics.v1.LinearAlignment result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.LinearAlignment buildPartial() {
      com.google.genomics.v1.LinearAlignment result = new com.google.genomics.v1.LinearAlignment(this);
      int from_bitField0_ = bitField0_;
      if (positionBuilder_ == null) {
        result.position_ = position_;
      } else {
        result.position_ = positionBuilder_.build();
      }
      result.mappingQuality_ = mappingQuality_;
      if (cigarBuilder_ == null) {
        if (((bitField0_ & 0x00000001) != 0)) {
          cigar_ = java.util.Collections.unmodifiableList(cigar_);
          bitField0_ = (bitField0_ & ~0x00000001);
        }
        result.cigar_ = cigar_;
      } else {
        result.cigar_ = cigarBuilder_.build();
      }
      onBuilt();
      return result;
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.LinearAlignment) {
        return mergeFrom((com.google.genomics.v1.LinearAlignment)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.LinearAlignment other) {
      if (other == com.google.genomics.v1.LinearAlignment.getDefaultInstance()) return this;
      if (other.hasPosition()) {
        mergePosition(other.getPosition());
      }
      if (other.getMappingQuality() != 0) {
        setMappingQuality(other.getMappingQuality());
      }
      if (cigarBuilder_ == null) {
        if (!other.cigar_.isEmpty()) {
          if (cigar_.isEmpty()) {
            cigar_ = other.cigar_;
            bitField0_ = (bitField0_ & ~0x00000001);
          } else {
            ensureCigarIsMutable();
            cigar_.addAll(other.cigar_);
          }
          onChanged();
        }
      } else {
        if (!other.cigar_.isEmpty()) {
          if (cigarBuilder_.isEmpty()) {
            cigarBuilder_.dispose();
            cigarBuilder_ = null;
            cigar_ = other.cigar_;
            bitField0_ = (bitField0_ & ~0x00000001);
            cigarBuilder_ = 
              com.google.protobuf.GeneratedMessageV3.alwaysUseFieldBuilders ?
                 getCigarFieldBuilder() : null;
          } else {
            cigarBuilder_.addAllMessages(other.cigar_);
          }
        }
      }
      this.mergeUnknownFields(other.unknownFields);
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      com.google.genomics.v1.LinearAlignment parsedMessage = null;
      try {
        parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        parsedMessage = (com.google.genomics.v1.LinearAlignment) e.getUnfinishedMessage();
        throw e.unwrapIOException();
      } finally {
        if (parsedMessage != null) {
          mergeFrom(parsedMessage);
        }
      }
      return this;
    }
    private int bitField0_;

    private com.google.genomics.v1.Position position_;
    private com.google.protobuf.SingleFieldBuilderV3<
        com.google.genomics.v1.Position, com.google.genomics.v1.Position.Builder, com.google.genomics.v1.PositionOrBuilder> positionBuilder_;
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     * @return Whether the position field is set.
     */
    public boolean hasPosition() {
      return positionBuilder_ != null || position_ != null;
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     * @return The position.
     */
    public com.google.genomics.v1.Position getPosition() {
      if (positionBuilder_ == null) {
        return position_ == null ? com.google.genomics.v1.Position.getDefaultInstance() : position_;
      } else {
        return positionBuilder_.getMessage();
      }
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    public Builder setPosition(com.google.genomics.v1.Position value) {
      if (positionBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        position_ = value;
        onChanged();
      } else {
        positionBuilder_.setMessage(value);
      }

      return this;
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    public Builder setPosition(
        com.google.genomics.v1.Position.Builder builderForValue) {
      if (positionBuilder_ == null) {
        position_ = builderForValue.build();
        onChanged();
      } else {
        positionBuilder_.setMessage(builderForValue.build());
      }

      return this;
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    public Builder mergePosition(com.google.genomics.v1.Position value) {
      if (positionBuilder_ == null) {
        if (position_ != null) {
          position_ =
            com.google.genomics.v1.Position.newBuilder(position_).mergeFrom(value).buildPartial();
        } else {
          position_ = value;
        }
        onChanged();
      } else {
        positionBuilder_.mergeFrom(value);
      }

      return this;
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    public Builder clearPosition() {
      if (positionBuilder_ == null) {
        position_ = null;
        onChanged();
      } else {
        position_ = null;
        positionBuilder_ = null;
      }

      return this;
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    public com.google.genomics.v1.Position.Builder getPositionBuilder() {
      
      onChanged();
      return getPositionFieldBuilder().getBuilder();
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    public com.google.genomics.v1.PositionOrBuilder getPositionOrBuilder() {
      if (positionBuilder_ != null) {
        return positionBuilder_.getMessageOrBuilder();
      } else {
        return position_ == null ?
            com.google.genomics.v1.Position.getDefaultInstance() : position_;
      }
    }
    /**
     * <pre>
     * The position of this alignment.
     * </pre>
     *
     * <code>.google.genomics.v1.Position position = 1;</code>
     */
    private com.google.protobuf.SingleFieldBuilderV3<
        com.google.genomics.v1.Position, com.google.genomics.v1.Position.Builder, com.google.genomics.v1.PositionOrBuilder> 
        getPositionFieldBuilder() {
      if (positionBuilder_ == null) {
        positionBuilder_ = new com.google.protobuf.SingleFieldBuilderV3<
            com.google.genomics.v1.Position, com.google.genomics.v1.Position.Builder, com.google.genomics.v1.PositionOrBuilder>(
                getPosition(),
                getParentForChildren(),
                isClean());
        position_ = null;
      }
      return positionBuilder_;
    }

    private int mappingQuality_ ;
    /**
     * <pre>
     * The mapping quality of this alignment. Represents how likely
     * the read maps to this position as opposed to other locations.
     * Specifically, this is -10 log10 Pr(mapping position is wrong), rounded to
     * the nearest integer.
     * </pre>
     *
     * <code>int32 mapping_quality = 2;</code>
     * @return The mappingQuality.
     */
    @java.lang.Override
    public int getMappingQuality() {
      return mappingQuality_;
    }
    /**
     * <pre>
     * The mapping quality of this alignment. Represents how likely
     * the read maps to this position as opposed to other locations.
     * Specifically, this is -10 log10 Pr(mapping position is wrong), rounded to
     * the nearest integer.
     * </pre>
     *
     * <code>int32 mapping_quality = 2;</code>
     * @param value The mappingQuality to set.
     * @return This builder for chaining.
     */
    public Builder setMappingQuality(int value) {
      
      mappingQuality_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The mapping quality of this alignment. Represents how likely
     * the read maps to this position as opposed to other locations.
     * Specifically, this is -10 log10 Pr(mapping position is wrong), rounded to
     * the nearest integer.
     * </pre>
     *
     * <code>int32 mapping_quality = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearMappingQuality() {
      
      mappingQuality_ = 0;
      onChanged();
      return this;
    }

    private java.util.List<com.google.genomics.v1.CigarUnit> cigar_ =
      java.util.Collections.emptyList();
    private void ensureCigarIsMutable() {
      if (!((bitField0_ & 0x00000001) != 0)) {
        cigar_ = new java.util.ArrayList<com.google.genomics.v1.CigarUnit>(cigar_);
        bitField0_ |= 0x00000001;
       }
    }

    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.CigarUnit, com.google.genomics.v1.CigarUnit.Builder, com.google.genomics.v1.CigarUnitOrBuilder> cigarBuilder_;

    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public java.util.List<com.google.genomics.v1.CigarUnit> getCigarList() {
      if (cigarBuilder_ == null) {
        return java.util.Collections.unmodifiableList(cigar_);
      } else {
        return cigarBuilder_.getMessageList();
      }
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public int getCigarCount() {
      if (cigarBuilder_ == null) {
        return cigar_.size();
      } else {
        return cigarBuilder_.getCount();
      }
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public com.google.genomics.v1.CigarUnit getCigar(int index) {
      if (cigarBuilder_ == null) {
        return cigar_.get(index);
      } else {
        return cigarBuilder_.getMessage(index);
      }
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder setCigar(
        int index, com.google.genomics.v1.CigarUnit value) {
      if (cigarBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureCigarIsMutable();
        cigar_.set(index, value);
        onChanged();
      } else {
        cigarBuilder_.setMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder setCigar(
        int index, com.google.genomics.v1.CigarUnit.Builder builderForValue) {
      if (cigarBuilder_ == null) {
        ensureCigarIsMutable();
        cigar_.set(index, builderForValue.build());
        onChanged();
      } else {
        cigarBuilder_.setMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder addCigar(com.google.genomics.v1.CigarUnit value) {
      if (cigarBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureCigarIsMutable();
        cigar_.add(value);
        onChanged();
      } else {
        cigarBuilder_.addMessage(value);
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder addCigar(
        int index, com.google.genomics.v1.CigarUnit value) {
      if (cigarBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureCigarIsMutable();
        cigar_.add(index, value);
        onChanged();
      } else {
        cigarBuilder_.addMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder addCigar(
        com.google.genomics.v1.CigarUnit.Builder builderForValue) {
      if (cigarBuilder_ == null) {
        ensureCigarIsMutable();
        cigar_.add(builderForValue.build());
        onChanged();
      } else {
        cigarBuilder_.addMessage(builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder addCigar(
        int index, com.google.genomics.v1.CigarUnit.Builder builderForValue) {
      if (cigarBuilder_ == null) {
        ensureCigarIsMutable();
        cigar_.add(index, builderForValue.build());
        onChanged();
      } else {
        cigarBuilder_.addMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder addAllCigar(
        java.lang.Iterable<? extends com.google.genomics.v1.CigarUnit> values) {
      if (cigarBuilder_ == null) {
        ensureCigarIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, cigar_);
        onChanged();
      } else {
        cigarBuilder_.addAllMessages(values);
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder clearCigar() {
      if (cigarBuilder_ == null) {
        cigar_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
        onChanged();
      } else {
        cigarBuilder_.clear();
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public Builder removeCigar(int index) {
      if (cigarBuilder_ == null) {
        ensureCigarIsMutable();
        cigar_.remove(index);
        onChanged();
      } else {
        cigarBuilder_.remove(index);
      }
      return this;
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public com.google.genomics.v1.CigarUnit.Builder getCigarBuilder(
        int index) {
      return getCigarFieldBuilder().getBuilder(index);
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public com.google.genomics.v1.CigarUnitOrBuilder getCigarOrBuilder(
        int index) {
      if (cigarBuilder_ == null) {
        return cigar_.get(index);  } else {
        return cigarBuilder_.getMessageOrBuilder(index);
      }
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public java.util.List<? extends com.google.genomics.v1.CigarUnitOrBuilder> 
         getCigarOrBuilderList() {
      if (cigarBuilder_ != null) {
        return cigarBuilder_.getMessageOrBuilderList();
      } else {
        return java.util.Collections.unmodifiableList(cigar_);
      }
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public com.google.genomics.v1.CigarUnit.Builder addCigarBuilder() {
      return getCigarFieldBuilder().addBuilder(
          com.google.genomics.v1.CigarUnit.getDefaultInstance());
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public com.google.genomics.v1.CigarUnit.Builder addCigarBuilder(
        int index) {
      return getCigarFieldBuilder().addBuilder(
          index, com.google.genomics.v1.CigarUnit.getDefaultInstance());
    }
    /**
     * <pre>
     * Represents the local alignment of this sequence (alignment matches, indels,
     * etc) against the reference.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.CigarUnit cigar = 3;</code>
     */
    public java.util.List<com.google.genomics.v1.CigarUnit.Builder> 
         getCigarBuilderList() {
      return getCigarFieldBuilder().getBuilderList();
    }
    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.CigarUnit, com.google.genomics.v1.CigarUnit.Builder, com.google.genomics.v1.CigarUnitOrBuilder> 
        getCigarFieldBuilder() {
      if (cigarBuilder_ == null) {
        cigarBuilder_ = new com.google.protobuf.RepeatedFieldBuilderV3<
            com.google.genomics.v1.CigarUnit, com.google.genomics.v1.CigarUnit.Builder, com.google.genomics.v1.CigarUnitOrBuilder>(
                cigar_,
                ((bitField0_ & 0x00000001) != 0),
                getParentForChildren(),
                isClean());
        cigar_ = null;
      }
      return cigarBuilder_;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.LinearAlignment)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.LinearAlignment)
  private static final com.google.genomics.v1.LinearAlignment DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.LinearAlignment();
  }

  public static com.google.genomics.v1.LinearAlignment getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<LinearAlignment>
      PARSER = new com.google.protobuf.AbstractParser<LinearAlignment>() {
    @java.lang.Override
    public LinearAlignment parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return new LinearAlignment(input, extensionRegistry);
    }
  };

  public static com.google.protobuf.Parser<LinearAlignment> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<LinearAlignment> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.LinearAlignment getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

