// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/readgroup.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A read group is all the data that's processed the same way by the sequencer.
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.ReadGroup}
 */
public final class ReadGroup extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.ReadGroup)
    ReadGroupOrBuilder {
private static final long serialVersionUID = 0L;
  // Use ReadGroup.newBuilder() to construct.
  private ReadGroup(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private ReadGroup() {
    id_ = "";
    datasetId_ = "";
    name_ = "";
    description_ = "";
    sampleId_ = "";
    programs_ = java.util.Collections.emptyList();
    referenceSetId_ = "";
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new ReadGroup();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  private ReadGroup(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    this();
    if (extensionRegistry == null) {
      throw new java.lang.NullPointerException();
    }
    int mutable_bitField0_ = 0;
    com.google.protobuf.UnknownFieldSet.Builder unknownFields =
        com.google.protobuf.UnknownFieldSet.newBuilder();
    try {
      boolean done = false;
      while (!done) {
        int tag = input.readTag();
        switch (tag) {
          case 0:
            done = true;
            break;
          case 10: {
            java.lang.String s = input.readStringRequireUtf8();

            id_ = s;
            break;
          }
          case 18: {
            java.lang.String s = input.readStringRequireUtf8();

            datasetId_ = s;
            break;
          }
          case 26: {
            java.lang.String s = input.readStringRequireUtf8();

            name_ = s;
            break;
          }
          case 34: {
            java.lang.String s = input.readStringRequireUtf8();

            description_ = s;
            break;
          }
          case 42: {
            java.lang.String s = input.readStringRequireUtf8();

            sampleId_ = s;
            break;
          }
          case 50: {
            com.google.genomics.v1.ReadGroup.Experiment.Builder subBuilder = null;
            if (experiment_ != null) {
              subBuilder = experiment_.toBuilder();
            }
            experiment_ = input.readMessage(com.google.genomics.v1.ReadGroup.Experiment.parser(), extensionRegistry);
            if (subBuilder != null) {
              subBuilder.mergeFrom(experiment_);
              experiment_ = subBuilder.buildPartial();
            }

            break;
          }
          case 56: {

            predictedInsertSize_ = input.readInt32();
            break;
          }
          case 82: {
            if (!((mutable_bitField0_ & 0x00000001) != 0)) {
              programs_ = new java.util.ArrayList<com.google.genomics.v1.ReadGroup.Program>();
              mutable_bitField0_ |= 0x00000001;
            }
            programs_.add(
                input.readMessage(com.google.genomics.v1.ReadGroup.Program.parser(), extensionRegistry));
            break;
          }
          case 90: {
            java.lang.String s = input.readStringRequireUtf8();

            referenceSetId_ = s;
            break;
          }
          case 98: {
            if (!((mutable_bitField0_ & 0x00000002) != 0)) {
              info_ = com.google.protobuf.MapField.newMapField(
                  InfoDefaultEntryHolder.defaultEntry);
              mutable_bitField0_ |= 0x00000002;
            }
            com.google.protobuf.MapEntry<java.lang.String, com.google.protobuf.ListValue>
            info__ = input.readMessage(
                InfoDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
            info_.getMutableMap().put(
                info__.getKey(), info__.getValue());
            break;
          }
          default: {
            if (!parseUnknownField(
                input, unknownFields, extensionRegistry, tag)) {
              done = true;
            }
            break;
          }
        }
      }
    } catch (com.google.protobuf.InvalidProtocolBufferException e) {
      throw e.setUnfinishedMessage(this);
    } catch (com.google.protobuf.UninitializedMessageException e) {
      throw e.asInvalidProtocolBufferException().setUnfinishedMessage(this);
    } catch (java.io.IOException e) {
      throw new com.google.protobuf.InvalidProtocolBufferException(
          e).setUnfinishedMessage(this);
    } finally {
      if (((mutable_bitField0_ & 0x00000001) != 0)) {
        programs_ = java.util.Collections.unmodifiableList(programs_);
      }
      this.unknownFields = unknownFields.build();
      makeExtensionsImmutable();
    }
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_descriptor;
  }

  @SuppressWarnings({"rawtypes"})
  @java.lang.Override
  protected com.google.protobuf.MapField internalGetMapField(
      int number) {
    switch (number) {
      case 12:
        return internalGetInfo();
      default:
        throw new RuntimeException(
            "Invalid map field number: " + number);
    }
  }
  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.ReadGroup.class, com.google.genomics.v1.ReadGroup.Builder.class);
  }

  public interface ExperimentOrBuilder extends
      // @@protoc_insertion_point(interface_extends:google.genomics.v1.ReadGroup.Experiment)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <pre>
     * A client-supplied library identifier; a library is a collection of DNA
     * fragments which have been prepared for sequencing from a sample. This
     * field is important for quality control as error or bias can be introduced
     * during sample preparation.
     * </pre>
     *
     * <code>string library_id = 1;</code>
     * @return The libraryId.
     */
    java.lang.String getLibraryId();
    /**
     * <pre>
     * A client-supplied library identifier; a library is a collection of DNA
     * fragments which have been prepared for sequencing from a sample. This
     * field is important for quality control as error or bias can be introduced
     * during sample preparation.
     * </pre>
     *
     * <code>string library_id = 1;</code>
     * @return The bytes for libraryId.
     */
    com.google.protobuf.ByteString
        getLibraryIdBytes();

    /**
     * <pre>
     * The platform unit used as part of this experiment, for example
     * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
     * &#64;RG PU field in the SAM spec.
     * </pre>
     *
     * <code>string platform_unit = 2;</code>
     * @return The platformUnit.
     */
    java.lang.String getPlatformUnit();
    /**
     * <pre>
     * The platform unit used as part of this experiment, for example
     * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
     * &#64;RG PU field in the SAM spec.
     * </pre>
     *
     * <code>string platform_unit = 2;</code>
     * @return The bytes for platformUnit.
     */
    com.google.protobuf.ByteString
        getPlatformUnitBytes();

    /**
     * <pre>
     * The sequencing center used as part of this experiment.
     * </pre>
     *
     * <code>string sequencing_center = 3;</code>
     * @return The sequencingCenter.
     */
    java.lang.String getSequencingCenter();
    /**
     * <pre>
     * The sequencing center used as part of this experiment.
     * </pre>
     *
     * <code>string sequencing_center = 3;</code>
     * @return The bytes for sequencingCenter.
     */
    com.google.protobuf.ByteString
        getSequencingCenterBytes();

    /**
     * <pre>
     * The instrument model used as part of this experiment. This maps to
     * sequencing technology in the SAM spec.
     * </pre>
     *
     * <code>string instrument_model = 4;</code>
     * @return The instrumentModel.
     */
    java.lang.String getInstrumentModel();
    /**
     * <pre>
     * The instrument model used as part of this experiment. This maps to
     * sequencing technology in the SAM spec.
     * </pre>
     *
     * <code>string instrument_model = 4;</code>
     * @return The bytes for instrumentModel.
     */
    com.google.protobuf.ByteString
        getInstrumentModelBytes();
  }
  /**
   * Protobuf type {@code google.genomics.v1.ReadGroup.Experiment}
   */
  public static final class Experiment extends
      com.google.protobuf.GeneratedMessageV3 implements
      // @@protoc_insertion_point(message_implements:google.genomics.v1.ReadGroup.Experiment)
      ExperimentOrBuilder {
  private static final long serialVersionUID = 0L;
    // Use Experiment.newBuilder() to construct.
    private Experiment(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
      super(builder);
    }
    private Experiment() {
      libraryId_ = "";
      platformUnit_ = "";
      sequencingCenter_ = "";
      instrumentModel_ = "";
    }

    @java.lang.Override
    @SuppressWarnings({"unused"})
    protected java.lang.Object newInstance(
        UnusedPrivateParameter unused) {
      return new Experiment();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return this.unknownFields;
    }
    private Experiment(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      this();
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      com.google.protobuf.UnknownFieldSet.Builder unknownFields =
          com.google.protobuf.UnknownFieldSet.newBuilder();
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 10: {
              java.lang.String s = input.readStringRequireUtf8();

              libraryId_ = s;
              break;
            }
            case 18: {
              java.lang.String s = input.readStringRequireUtf8();

              platformUnit_ = s;
              break;
            }
            case 26: {
              java.lang.String s = input.readStringRequireUtf8();

              sequencingCenter_ = s;
              break;
            }
            case 34: {
              java.lang.String s = input.readStringRequireUtf8();

              instrumentModel_ = s;
              break;
            }
            default: {
              if (!parseUnknownField(
                  input, unknownFields, extensionRegistry, tag)) {
                done = true;
              }
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(this);
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(this);
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(
            e).setUnfinishedMessage(this);
      } finally {
        this.unknownFields = unknownFields.build();
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Experiment_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Experiment_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.ReadGroup.Experiment.class, com.google.genomics.v1.ReadGroup.Experiment.Builder.class);
    }

    public static final int LIBRARY_ID_FIELD_NUMBER = 1;
    private volatile java.lang.Object libraryId_;
    /**
     * <pre>
     * A client-supplied library identifier; a library is a collection of DNA
     * fragments which have been prepared for sequencing from a sample. This
     * field is important for quality control as error or bias can be introduced
     * during sample preparation.
     * </pre>
     *
     * <code>string library_id = 1;</code>
     * @return The libraryId.
     */
    @java.lang.Override
    public java.lang.String getLibraryId() {
      java.lang.Object ref = libraryId_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        libraryId_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * A client-supplied library identifier; a library is a collection of DNA
     * fragments which have been prepared for sequencing from a sample. This
     * field is important for quality control as error or bias can be introduced
     * during sample preparation.
     * </pre>
     *
     * <code>string library_id = 1;</code>
     * @return The bytes for libraryId.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getLibraryIdBytes() {
      java.lang.Object ref = libraryId_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        libraryId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int PLATFORM_UNIT_FIELD_NUMBER = 2;
    private volatile java.lang.Object platformUnit_;
    /**
     * <pre>
     * The platform unit used as part of this experiment, for example
     * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
     * &#64;RG PU field in the SAM spec.
     * </pre>
     *
     * <code>string platform_unit = 2;</code>
     * @return The platformUnit.
     */
    @java.lang.Override
    public java.lang.String getPlatformUnit() {
      java.lang.Object ref = platformUnit_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        platformUnit_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The platform unit used as part of this experiment, for example
     * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
     * &#64;RG PU field in the SAM spec.
     * </pre>
     *
     * <code>string platform_unit = 2;</code>
     * @return The bytes for platformUnit.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getPlatformUnitBytes() {
      java.lang.Object ref = platformUnit_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        platformUnit_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int SEQUENCING_CENTER_FIELD_NUMBER = 3;
    private volatile java.lang.Object sequencingCenter_;
    /**
     * <pre>
     * The sequencing center used as part of this experiment.
     * </pre>
     *
     * <code>string sequencing_center = 3;</code>
     * @return The sequencingCenter.
     */
    @java.lang.Override
    public java.lang.String getSequencingCenter() {
      java.lang.Object ref = sequencingCenter_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        sequencingCenter_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The sequencing center used as part of this experiment.
     * </pre>
     *
     * <code>string sequencing_center = 3;</code>
     * @return The bytes for sequencingCenter.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getSequencingCenterBytes() {
      java.lang.Object ref = sequencingCenter_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        sequencingCenter_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int INSTRUMENT_MODEL_FIELD_NUMBER = 4;
    private volatile java.lang.Object instrumentModel_;
    /**
     * <pre>
     * The instrument model used as part of this experiment. This maps to
     * sequencing technology in the SAM spec.
     * </pre>
     *
     * <code>string instrument_model = 4;</code>
     * @return The instrumentModel.
     */
    @java.lang.Override
    public java.lang.String getInstrumentModel() {
      java.lang.Object ref = instrumentModel_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        instrumentModel_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The instrument model used as part of this experiment. This maps to
     * sequencing technology in the SAM spec.
     * </pre>
     *
     * <code>string instrument_model = 4;</code>
     * @return The bytes for instrumentModel.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getInstrumentModelBytes() {
      java.lang.Object ref = instrumentModel_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        instrumentModel_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    private byte memoizedIsInitialized = -1;
    @java.lang.Override
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    @java.lang.Override
    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(libraryId_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 1, libraryId_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(platformUnit_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 2, platformUnit_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(sequencingCenter_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 3, sequencingCenter_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(instrumentModel_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 4, instrumentModel_);
      }
      unknownFields.writeTo(output);
    }

    @java.lang.Override
    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(libraryId_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, libraryId_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(platformUnit_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(2, platformUnit_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(sequencingCenter_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(3, sequencingCenter_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(instrumentModel_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(4, instrumentModel_);
      }
      size += unknownFields.getSerializedSize();
      memoizedSize = size;
      return size;
    }

    @java.lang.Override
    public boolean equals(final java.lang.Object obj) {
      if (obj == this) {
       return true;
      }
      if (!(obj instanceof com.google.genomics.v1.ReadGroup.Experiment)) {
        return super.equals(obj);
      }
      com.google.genomics.v1.ReadGroup.Experiment other = (com.google.genomics.v1.ReadGroup.Experiment) obj;

      if (!getLibraryId()
          .equals(other.getLibraryId())) return false;
      if (!getPlatformUnit()
          .equals(other.getPlatformUnit())) return false;
      if (!getSequencingCenter()
          .equals(other.getSequencingCenter())) return false;
      if (!getInstrumentModel()
          .equals(other.getInstrumentModel())) return false;
      if (!unknownFields.equals(other.unknownFields)) return false;
      return true;
    }

    @java.lang.Override
    public int hashCode() {
      if (memoizedHashCode != 0) {
        return memoizedHashCode;
      }
      int hash = 41;
      hash = (19 * hash) + getDescriptor().hashCode();
      hash = (37 * hash) + LIBRARY_ID_FIELD_NUMBER;
      hash = (53 * hash) + getLibraryId().hashCode();
      hash = (37 * hash) + PLATFORM_UNIT_FIELD_NUMBER;
      hash = (53 * hash) + getPlatformUnit().hashCode();
      hash = (37 * hash) + SEQUENCING_CENTER_FIELD_NUMBER;
      hash = (53 * hash) + getSequencingCenter().hashCode();
      hash = (37 * hash) + INSTRUMENT_MODEL_FIELD_NUMBER;
      hash = (53 * hash) + getInstrumentModel().hashCode();
      hash = (29 * hash) + unknownFields.hashCode();
      memoizedHashCode = hash;
      return hash;
    }

    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        java.nio.ByteBuffer data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        java.nio.ByteBuffer data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseDelimitedWithIOException(PARSER, input);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input);
    }
    public static com.google.genomics.v1.ReadGroup.Experiment parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input, extensionRegistry);
    }

    @java.lang.Override
    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(com.google.genomics.v1.ReadGroup.Experiment prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    @java.lang.Override
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code google.genomics.v1.ReadGroup.Experiment}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:google.genomics.v1.ReadGroup.Experiment)
        com.google.genomics.v1.ReadGroup.ExperimentOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Experiment_descriptor;
      }

      @java.lang.Override
      protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Experiment_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                com.google.genomics.v1.ReadGroup.Experiment.class, com.google.genomics.v1.ReadGroup.Experiment.Builder.class);
      }

      // Construct using com.google.genomics.v1.ReadGroup.Experiment.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessageV3
                .alwaysUseFieldBuilders) {
        }
      }
      @java.lang.Override
      public Builder clear() {
        super.clear();
        libraryId_ = "";

        platformUnit_ = "";

        sequencingCenter_ = "";

        instrumentModel_ = "";

        return this;
      }

      @java.lang.Override
      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Experiment_descriptor;
      }

      @java.lang.Override
      public com.google.genomics.v1.ReadGroup.Experiment getDefaultInstanceForType() {
        return com.google.genomics.v1.ReadGroup.Experiment.getDefaultInstance();
      }

      @java.lang.Override
      public com.google.genomics.v1.ReadGroup.Experiment build() {
        com.google.genomics.v1.ReadGroup.Experiment result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      @java.lang.Override
      public com.google.genomics.v1.ReadGroup.Experiment buildPartial() {
        com.google.genomics.v1.ReadGroup.Experiment result = new com.google.genomics.v1.ReadGroup.Experiment(this);
        result.libraryId_ = libraryId_;
        result.platformUnit_ = platformUnit_;
        result.sequencingCenter_ = sequencingCenter_;
        result.instrumentModel_ = instrumentModel_;
        onBuilt();
        return result;
      }

      @java.lang.Override
      public Builder clone() {
        return super.clone();
      }
      @java.lang.Override
      public Builder setField(
          com.google.protobuf.Descriptors.FieldDescriptor field,
          java.lang.Object value) {
        return super.setField(field, value);
      }
      @java.lang.Override
      public Builder clearField(
          com.google.protobuf.Descriptors.FieldDescriptor field) {
        return super.clearField(field);
      }
      @java.lang.Override
      public Builder clearOneof(
          com.google.protobuf.Descriptors.OneofDescriptor oneof) {
        return super.clearOneof(oneof);
      }
      @java.lang.Override
      public Builder setRepeatedField(
          com.google.protobuf.Descriptors.FieldDescriptor field,
          int index, java.lang.Object value) {
        return super.setRepeatedField(field, index, value);
      }
      @java.lang.Override
      public Builder addRepeatedField(
          com.google.protobuf.Descriptors.FieldDescriptor field,
          java.lang.Object value) {
        return super.addRepeatedField(field, value);
      }
      @java.lang.Override
      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof com.google.genomics.v1.ReadGroup.Experiment) {
          return mergeFrom((com.google.genomics.v1.ReadGroup.Experiment)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(com.google.genomics.v1.ReadGroup.Experiment other) {
        if (other == com.google.genomics.v1.ReadGroup.Experiment.getDefaultInstance()) return this;
        if (!other.getLibraryId().isEmpty()) {
          libraryId_ = other.libraryId_;
          onChanged();
        }
        if (!other.getPlatformUnit().isEmpty()) {
          platformUnit_ = other.platformUnit_;
          onChanged();
        }
        if (!other.getSequencingCenter().isEmpty()) {
          sequencingCenter_ = other.sequencingCenter_;
          onChanged();
        }
        if (!other.getInstrumentModel().isEmpty()) {
          instrumentModel_ = other.instrumentModel_;
          onChanged();
        }
        this.mergeUnknownFields(other.unknownFields);
        onChanged();
        return this;
      }

      @java.lang.Override
      public final boolean isInitialized() {
        return true;
      }

      @java.lang.Override
      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        com.google.genomics.v1.ReadGroup.Experiment parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (com.google.genomics.v1.ReadGroup.Experiment) e.getUnfinishedMessage();
          throw e.unwrapIOException();
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }

      private java.lang.Object libraryId_ = "";
      /**
       * <pre>
       * A client-supplied library identifier; a library is a collection of DNA
       * fragments which have been prepared for sequencing from a sample. This
       * field is important for quality control as error or bias can be introduced
       * during sample preparation.
       * </pre>
       *
       * <code>string library_id = 1;</code>
       * @return The libraryId.
       */
      public java.lang.String getLibraryId() {
        java.lang.Object ref = libraryId_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          libraryId_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * A client-supplied library identifier; a library is a collection of DNA
       * fragments which have been prepared for sequencing from a sample. This
       * field is important for quality control as error or bias can be introduced
       * during sample preparation.
       * </pre>
       *
       * <code>string library_id = 1;</code>
       * @return The bytes for libraryId.
       */
      public com.google.protobuf.ByteString
          getLibraryIdBytes() {
        java.lang.Object ref = libraryId_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          libraryId_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * A client-supplied library identifier; a library is a collection of DNA
       * fragments which have been prepared for sequencing from a sample. This
       * field is important for quality control as error or bias can be introduced
       * during sample preparation.
       * </pre>
       *
       * <code>string library_id = 1;</code>
       * @param value The libraryId to set.
       * @return This builder for chaining.
       */
      public Builder setLibraryId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        libraryId_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * A client-supplied library identifier; a library is a collection of DNA
       * fragments which have been prepared for sequencing from a sample. This
       * field is important for quality control as error or bias can be introduced
       * during sample preparation.
       * </pre>
       *
       * <code>string library_id = 1;</code>
       * @return This builder for chaining.
       */
      public Builder clearLibraryId() {
        
        libraryId_ = getDefaultInstance().getLibraryId();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * A client-supplied library identifier; a library is a collection of DNA
       * fragments which have been prepared for sequencing from a sample. This
       * field is important for quality control as error or bias can be introduced
       * during sample preparation.
       * </pre>
       *
       * <code>string library_id = 1;</code>
       * @param value The bytes for libraryId to set.
       * @return This builder for chaining.
       */
      public Builder setLibraryIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        libraryId_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object platformUnit_ = "";
      /**
       * <pre>
       * The platform unit used as part of this experiment, for example
       * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
       * &#64;RG PU field in the SAM spec.
       * </pre>
       *
       * <code>string platform_unit = 2;</code>
       * @return The platformUnit.
       */
      public java.lang.String getPlatformUnit() {
        java.lang.Object ref = platformUnit_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          platformUnit_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The platform unit used as part of this experiment, for example
       * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
       * &#64;RG PU field in the SAM spec.
       * </pre>
       *
       * <code>string platform_unit = 2;</code>
       * @return The bytes for platformUnit.
       */
      public com.google.protobuf.ByteString
          getPlatformUnitBytes() {
        java.lang.Object ref = platformUnit_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          platformUnit_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The platform unit used as part of this experiment, for example
       * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
       * &#64;RG PU field in the SAM spec.
       * </pre>
       *
       * <code>string platform_unit = 2;</code>
       * @param value The platformUnit to set.
       * @return This builder for chaining.
       */
      public Builder setPlatformUnit(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        platformUnit_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The platform unit used as part of this experiment, for example
       * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
       * &#64;RG PU field in the SAM spec.
       * </pre>
       *
       * <code>string platform_unit = 2;</code>
       * @return This builder for chaining.
       */
      public Builder clearPlatformUnit() {
        
        platformUnit_ = getDefaultInstance().getPlatformUnit();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The platform unit used as part of this experiment, for example
       * flowcell-barcode.lane for Illumina or slide for SOLiD. Corresponds to the
       * &#64;RG PU field in the SAM spec.
       * </pre>
       *
       * <code>string platform_unit = 2;</code>
       * @param value The bytes for platformUnit to set.
       * @return This builder for chaining.
       */
      public Builder setPlatformUnitBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        platformUnit_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object sequencingCenter_ = "";
      /**
       * <pre>
       * The sequencing center used as part of this experiment.
       * </pre>
       *
       * <code>string sequencing_center = 3;</code>
       * @return The sequencingCenter.
       */
      public java.lang.String getSequencingCenter() {
        java.lang.Object ref = sequencingCenter_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          sequencingCenter_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The sequencing center used as part of this experiment.
       * </pre>
       *
       * <code>string sequencing_center = 3;</code>
       * @return The bytes for sequencingCenter.
       */
      public com.google.protobuf.ByteString
          getSequencingCenterBytes() {
        java.lang.Object ref = sequencingCenter_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          sequencingCenter_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The sequencing center used as part of this experiment.
       * </pre>
       *
       * <code>string sequencing_center = 3;</code>
       * @param value The sequencingCenter to set.
       * @return This builder for chaining.
       */
      public Builder setSequencingCenter(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        sequencingCenter_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The sequencing center used as part of this experiment.
       * </pre>
       *
       * <code>string sequencing_center = 3;</code>
       * @return This builder for chaining.
       */
      public Builder clearSequencingCenter() {
        
        sequencingCenter_ = getDefaultInstance().getSequencingCenter();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The sequencing center used as part of this experiment.
       * </pre>
       *
       * <code>string sequencing_center = 3;</code>
       * @param value The bytes for sequencingCenter to set.
       * @return This builder for chaining.
       */
      public Builder setSequencingCenterBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        sequencingCenter_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object instrumentModel_ = "";
      /**
       * <pre>
       * The instrument model used as part of this experiment. This maps to
       * sequencing technology in the SAM spec.
       * </pre>
       *
       * <code>string instrument_model = 4;</code>
       * @return The instrumentModel.
       */
      public java.lang.String getInstrumentModel() {
        java.lang.Object ref = instrumentModel_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          instrumentModel_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The instrument model used as part of this experiment. This maps to
       * sequencing technology in the SAM spec.
       * </pre>
       *
       * <code>string instrument_model = 4;</code>
       * @return The bytes for instrumentModel.
       */
      public com.google.protobuf.ByteString
          getInstrumentModelBytes() {
        java.lang.Object ref = instrumentModel_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          instrumentModel_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The instrument model used as part of this experiment. This maps to
       * sequencing technology in the SAM spec.
       * </pre>
       *
       * <code>string instrument_model = 4;</code>
       * @param value The instrumentModel to set.
       * @return This builder for chaining.
       */
      public Builder setInstrumentModel(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        instrumentModel_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The instrument model used as part of this experiment. This maps to
       * sequencing technology in the SAM spec.
       * </pre>
       *
       * <code>string instrument_model = 4;</code>
       * @return This builder for chaining.
       */
      public Builder clearInstrumentModel() {
        
        instrumentModel_ = getDefaultInstance().getInstrumentModel();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The instrument model used as part of this experiment. This maps to
       * sequencing technology in the SAM spec.
       * </pre>
       *
       * <code>string instrument_model = 4;</code>
       * @param value The bytes for instrumentModel to set.
       * @return This builder for chaining.
       */
      public Builder setInstrumentModelBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        instrumentModel_ = value;
        onChanged();
        return this;
      }
      @java.lang.Override
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return super.setUnknownFields(unknownFields);
      }

      @java.lang.Override
      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return super.mergeUnknownFields(unknownFields);
      }


      // @@protoc_insertion_point(builder_scope:google.genomics.v1.ReadGroup.Experiment)
    }

    // @@protoc_insertion_point(class_scope:google.genomics.v1.ReadGroup.Experiment)
    private static final com.google.genomics.v1.ReadGroup.Experiment DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new com.google.genomics.v1.ReadGroup.Experiment();
    }

    public static com.google.genomics.v1.ReadGroup.Experiment getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<Experiment>
        PARSER = new com.google.protobuf.AbstractParser<Experiment>() {
      @java.lang.Override
      public Experiment parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return new Experiment(input, extensionRegistry);
      }
    };

    public static com.google.protobuf.Parser<Experiment> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<Experiment> getParserForType() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroup.Experiment getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface ProgramOrBuilder extends
      // @@protoc_insertion_point(interface_extends:google.genomics.v1.ReadGroup.Program)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <pre>
     * The command line used to run this program.
     * </pre>
     *
     * <code>string command_line = 1;</code>
     * @return The commandLine.
     */
    java.lang.String getCommandLine();
    /**
     * <pre>
     * The command line used to run this program.
     * </pre>
     *
     * <code>string command_line = 1;</code>
     * @return The bytes for commandLine.
     */
    com.google.protobuf.ByteString
        getCommandLineBytes();

    /**
     * <pre>
     * The user specified locally unique ID of the program. Used along with
     * `prevProgramId` to define an ordering between programs.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return The id.
     */
    java.lang.String getId();
    /**
     * <pre>
     * The user specified locally unique ID of the program. Used along with
     * `prevProgramId` to define an ordering between programs.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return The bytes for id.
     */
    com.google.protobuf.ByteString
        getIdBytes();

    /**
     * <pre>
     * The display name of the program. This is typically the colloquial name of
     * the tool used, for example 'bwa' or 'picard'.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return The name.
     */
    java.lang.String getName();
    /**
     * <pre>
     * The display name of the program. This is typically the colloquial name of
     * the tool used, for example 'bwa' or 'picard'.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return The bytes for name.
     */
    com.google.protobuf.ByteString
        getNameBytes();

    /**
     * <pre>
     * The ID of the program run before this one.
     * </pre>
     *
     * <code>string prev_program_id = 4;</code>
     * @return The prevProgramId.
     */
    java.lang.String getPrevProgramId();
    /**
     * <pre>
     * The ID of the program run before this one.
     * </pre>
     *
     * <code>string prev_program_id = 4;</code>
     * @return The bytes for prevProgramId.
     */
    com.google.protobuf.ByteString
        getPrevProgramIdBytes();

    /**
     * <pre>
     * The version of the program run.
     * </pre>
     *
     * <code>string version = 5;</code>
     * @return The version.
     */
    java.lang.String getVersion();
    /**
     * <pre>
     * The version of the program run.
     * </pre>
     *
     * <code>string version = 5;</code>
     * @return The bytes for version.
     */
    com.google.protobuf.ByteString
        getVersionBytes();
  }
  /**
   * Protobuf type {@code google.genomics.v1.ReadGroup.Program}
   */
  public static final class Program extends
      com.google.protobuf.GeneratedMessageV3 implements
      // @@protoc_insertion_point(message_implements:google.genomics.v1.ReadGroup.Program)
      ProgramOrBuilder {
  private static final long serialVersionUID = 0L;
    // Use Program.newBuilder() to construct.
    private Program(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
      super(builder);
    }
    private Program() {
      commandLine_ = "";
      id_ = "";
      name_ = "";
      prevProgramId_ = "";
      version_ = "";
    }

    @java.lang.Override
    @SuppressWarnings({"unused"})
    protected java.lang.Object newInstance(
        UnusedPrivateParameter unused) {
      return new Program();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return this.unknownFields;
    }
    private Program(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      this();
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      com.google.protobuf.UnknownFieldSet.Builder unknownFields =
          com.google.protobuf.UnknownFieldSet.newBuilder();
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 10: {
              java.lang.String s = input.readStringRequireUtf8();

              commandLine_ = s;
              break;
            }
            case 18: {
              java.lang.String s = input.readStringRequireUtf8();

              id_ = s;
              break;
            }
            case 26: {
              java.lang.String s = input.readStringRequireUtf8();

              name_ = s;
              break;
            }
            case 34: {
              java.lang.String s = input.readStringRequireUtf8();

              prevProgramId_ = s;
              break;
            }
            case 42: {
              java.lang.String s = input.readStringRequireUtf8();

              version_ = s;
              break;
            }
            default: {
              if (!parseUnknownField(
                  input, unknownFields, extensionRegistry, tag)) {
                done = true;
              }
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(this);
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(this);
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(
            e).setUnfinishedMessage(this);
      } finally {
        this.unknownFields = unknownFields.build();
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Program_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Program_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.ReadGroup.Program.class, com.google.genomics.v1.ReadGroup.Program.Builder.class);
    }

    public static final int COMMAND_LINE_FIELD_NUMBER = 1;
    private volatile java.lang.Object commandLine_;
    /**
     * <pre>
     * The command line used to run this program.
     * </pre>
     *
     * <code>string command_line = 1;</code>
     * @return The commandLine.
     */
    @java.lang.Override
    public java.lang.String getCommandLine() {
      java.lang.Object ref = commandLine_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        commandLine_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The command line used to run this program.
     * </pre>
     *
     * <code>string command_line = 1;</code>
     * @return The bytes for commandLine.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getCommandLineBytes() {
      java.lang.Object ref = commandLine_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        commandLine_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ID_FIELD_NUMBER = 2;
    private volatile java.lang.Object id_;
    /**
     * <pre>
     * The user specified locally unique ID of the program. Used along with
     * `prevProgramId` to define an ordering between programs.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return The id.
     */
    @java.lang.Override
    public java.lang.String getId() {
      java.lang.Object ref = id_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        id_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The user specified locally unique ID of the program. Used along with
     * `prevProgramId` to define an ordering between programs.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return The bytes for id.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getIdBytes() {
      java.lang.Object ref = id_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        id_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int NAME_FIELD_NUMBER = 3;
    private volatile java.lang.Object name_;
    /**
     * <pre>
     * The display name of the program. This is typically the colloquial name of
     * the tool used, for example 'bwa' or 'picard'.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return The name.
     */
    @java.lang.Override
    public java.lang.String getName() {
      java.lang.Object ref = name_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        name_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The display name of the program. This is typically the colloquial name of
     * the tool used, for example 'bwa' or 'picard'.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return The bytes for name.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getNameBytes() {
      java.lang.Object ref = name_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        name_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int PREV_PROGRAM_ID_FIELD_NUMBER = 4;
    private volatile java.lang.Object prevProgramId_;
    /**
     * <pre>
     * The ID of the program run before this one.
     * </pre>
     *
     * <code>string prev_program_id = 4;</code>
     * @return The prevProgramId.
     */
    @java.lang.Override
    public java.lang.String getPrevProgramId() {
      java.lang.Object ref = prevProgramId_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        prevProgramId_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The ID of the program run before this one.
     * </pre>
     *
     * <code>string prev_program_id = 4;</code>
     * @return The bytes for prevProgramId.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getPrevProgramIdBytes() {
      java.lang.Object ref = prevProgramId_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        prevProgramId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int VERSION_FIELD_NUMBER = 5;
    private volatile java.lang.Object version_;
    /**
     * <pre>
     * The version of the program run.
     * </pre>
     *
     * <code>string version = 5;</code>
     * @return The version.
     */
    @java.lang.Override
    public java.lang.String getVersion() {
      java.lang.Object ref = version_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        version_ = s;
        return s;
      }
    }
    /**
     * <pre>
     * The version of the program run.
     * </pre>
     *
     * <code>string version = 5;</code>
     * @return The bytes for version.
     */
    @java.lang.Override
    public com.google.protobuf.ByteString
        getVersionBytes() {
      java.lang.Object ref = version_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        version_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    private byte memoizedIsInitialized = -1;
    @java.lang.Override
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    @java.lang.Override
    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(commandLine_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 1, commandLine_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 2, id_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 3, name_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(prevProgramId_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 4, prevProgramId_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(version_)) {
        com.google.protobuf.GeneratedMessageV3.writeString(output, 5, version_);
      }
      unknownFields.writeTo(output);
    }

    @java.lang.Override
    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(commandLine_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, commandLine_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(2, id_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(3, name_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(prevProgramId_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(4, prevProgramId_);
      }
      if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(version_)) {
        size += com.google.protobuf.GeneratedMessageV3.computeStringSize(5, version_);
      }
      size += unknownFields.getSerializedSize();
      memoizedSize = size;
      return size;
    }

    @java.lang.Override
    public boolean equals(final java.lang.Object obj) {
      if (obj == this) {
       return true;
      }
      if (!(obj instanceof com.google.genomics.v1.ReadGroup.Program)) {
        return super.equals(obj);
      }
      com.google.genomics.v1.ReadGroup.Program other = (com.google.genomics.v1.ReadGroup.Program) obj;

      if (!getCommandLine()
          .equals(other.getCommandLine())) return false;
      if (!getId()
          .equals(other.getId())) return false;
      if (!getName()
          .equals(other.getName())) return false;
      if (!getPrevProgramId()
          .equals(other.getPrevProgramId())) return false;
      if (!getVersion()
          .equals(other.getVersion())) return false;
      if (!unknownFields.equals(other.unknownFields)) return false;
      return true;
    }

    @java.lang.Override
    public int hashCode() {
      if (memoizedHashCode != 0) {
        return memoizedHashCode;
      }
      int hash = 41;
      hash = (19 * hash) + getDescriptor().hashCode();
      hash = (37 * hash) + COMMAND_LINE_FIELD_NUMBER;
      hash = (53 * hash) + getCommandLine().hashCode();
      hash = (37 * hash) + ID_FIELD_NUMBER;
      hash = (53 * hash) + getId().hashCode();
      hash = (37 * hash) + NAME_FIELD_NUMBER;
      hash = (53 * hash) + getName().hashCode();
      hash = (37 * hash) + PREV_PROGRAM_ID_FIELD_NUMBER;
      hash = (53 * hash) + getPrevProgramId().hashCode();
      hash = (37 * hash) + VERSION_FIELD_NUMBER;
      hash = (53 * hash) + getVersion().hashCode();
      hash = (29 * hash) + unknownFields.hashCode();
      memoizedHashCode = hash;
      return hash;
    }

    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        java.nio.ByteBuffer data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        java.nio.ByteBuffer data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseDelimitedWithIOException(PARSER, input);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input);
    }
    public static com.google.genomics.v1.ReadGroup.Program parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return com.google.protobuf.GeneratedMessageV3
          .parseWithIOException(PARSER, input, extensionRegistry);
    }

    @java.lang.Override
    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(com.google.genomics.v1.ReadGroup.Program prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    @java.lang.Override
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code google.genomics.v1.ReadGroup.Program}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:google.genomics.v1.ReadGroup.Program)
        com.google.genomics.v1.ReadGroup.ProgramOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Program_descriptor;
      }

      @java.lang.Override
      protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Program_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                com.google.genomics.v1.ReadGroup.Program.class, com.google.genomics.v1.ReadGroup.Program.Builder.class);
      }

      // Construct using com.google.genomics.v1.ReadGroup.Program.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessageV3
                .alwaysUseFieldBuilders) {
        }
      }
      @java.lang.Override
      public Builder clear() {
        super.clear();
        commandLine_ = "";

        id_ = "";

        name_ = "";

        prevProgramId_ = "";

        version_ = "";

        return this;
      }

      @java.lang.Override
      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_Program_descriptor;
      }

      @java.lang.Override
      public com.google.genomics.v1.ReadGroup.Program getDefaultInstanceForType() {
        return com.google.genomics.v1.ReadGroup.Program.getDefaultInstance();
      }

      @java.lang.Override
      public com.google.genomics.v1.ReadGroup.Program build() {
        com.google.genomics.v1.ReadGroup.Program result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      @java.lang.Override
      public com.google.genomics.v1.ReadGroup.Program buildPartial() {
        com.google.genomics.v1.ReadGroup.Program result = new com.google.genomics.v1.ReadGroup.Program(this);
        result.commandLine_ = commandLine_;
        result.id_ = id_;
        result.name_ = name_;
        result.prevProgramId_ = prevProgramId_;
        result.version_ = version_;
        onBuilt();
        return result;
      }

      @java.lang.Override
      public Builder clone() {
        return super.clone();
      }
      @java.lang.Override
      public Builder setField(
          com.google.protobuf.Descriptors.FieldDescriptor field,
          java.lang.Object value) {
        return super.setField(field, value);
      }
      @java.lang.Override
      public Builder clearField(
          com.google.protobuf.Descriptors.FieldDescriptor field) {
        return super.clearField(field);
      }
      @java.lang.Override
      public Builder clearOneof(
          com.google.protobuf.Descriptors.OneofDescriptor oneof) {
        return super.clearOneof(oneof);
      }
      @java.lang.Override
      public Builder setRepeatedField(
          com.google.protobuf.Descriptors.FieldDescriptor field,
          int index, java.lang.Object value) {
        return super.setRepeatedField(field, index, value);
      }
      @java.lang.Override
      public Builder addRepeatedField(
          com.google.protobuf.Descriptors.FieldDescriptor field,
          java.lang.Object value) {
        return super.addRepeatedField(field, value);
      }
      @java.lang.Override
      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof com.google.genomics.v1.ReadGroup.Program) {
          return mergeFrom((com.google.genomics.v1.ReadGroup.Program)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(com.google.genomics.v1.ReadGroup.Program other) {
        if (other == com.google.genomics.v1.ReadGroup.Program.getDefaultInstance()) return this;
        if (!other.getCommandLine().isEmpty()) {
          commandLine_ = other.commandLine_;
          onChanged();
        }
        if (!other.getId().isEmpty()) {
          id_ = other.id_;
          onChanged();
        }
        if (!other.getName().isEmpty()) {
          name_ = other.name_;
          onChanged();
        }
        if (!other.getPrevProgramId().isEmpty()) {
          prevProgramId_ = other.prevProgramId_;
          onChanged();
        }
        if (!other.getVersion().isEmpty()) {
          version_ = other.version_;
          onChanged();
        }
        this.mergeUnknownFields(other.unknownFields);
        onChanged();
        return this;
      }

      @java.lang.Override
      public final boolean isInitialized() {
        return true;
      }

      @java.lang.Override
      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        com.google.genomics.v1.ReadGroup.Program parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (com.google.genomics.v1.ReadGroup.Program) e.getUnfinishedMessage();
          throw e.unwrapIOException();
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }

      private java.lang.Object commandLine_ = "";
      /**
       * <pre>
       * The command line used to run this program.
       * </pre>
       *
       * <code>string command_line = 1;</code>
       * @return The commandLine.
       */
      public java.lang.String getCommandLine() {
        java.lang.Object ref = commandLine_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          commandLine_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The command line used to run this program.
       * </pre>
       *
       * <code>string command_line = 1;</code>
       * @return The bytes for commandLine.
       */
      public com.google.protobuf.ByteString
          getCommandLineBytes() {
        java.lang.Object ref = commandLine_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          commandLine_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The command line used to run this program.
       * </pre>
       *
       * <code>string command_line = 1;</code>
       * @param value The commandLine to set.
       * @return This builder for chaining.
       */
      public Builder setCommandLine(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        commandLine_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The command line used to run this program.
       * </pre>
       *
       * <code>string command_line = 1;</code>
       * @return This builder for chaining.
       */
      public Builder clearCommandLine() {
        
        commandLine_ = getDefaultInstance().getCommandLine();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The command line used to run this program.
       * </pre>
       *
       * <code>string command_line = 1;</code>
       * @param value The bytes for commandLine to set.
       * @return This builder for chaining.
       */
      public Builder setCommandLineBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        commandLine_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object id_ = "";
      /**
       * <pre>
       * The user specified locally unique ID of the program. Used along with
       * `prevProgramId` to define an ordering between programs.
       * </pre>
       *
       * <code>string id = 2;</code>
       * @return The id.
       */
      public java.lang.String getId() {
        java.lang.Object ref = id_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          id_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The user specified locally unique ID of the program. Used along with
       * `prevProgramId` to define an ordering between programs.
       * </pre>
       *
       * <code>string id = 2;</code>
       * @return The bytes for id.
       */
      public com.google.protobuf.ByteString
          getIdBytes() {
        java.lang.Object ref = id_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          id_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The user specified locally unique ID of the program. Used along with
       * `prevProgramId` to define an ordering between programs.
       * </pre>
       *
       * <code>string id = 2;</code>
       * @param value The id to set.
       * @return This builder for chaining.
       */
      public Builder setId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        id_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The user specified locally unique ID of the program. Used along with
       * `prevProgramId` to define an ordering between programs.
       * </pre>
       *
       * <code>string id = 2;</code>
       * @return This builder for chaining.
       */
      public Builder clearId() {
        
        id_ = getDefaultInstance().getId();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The user specified locally unique ID of the program. Used along with
       * `prevProgramId` to define an ordering between programs.
       * </pre>
       *
       * <code>string id = 2;</code>
       * @param value The bytes for id to set.
       * @return This builder for chaining.
       */
      public Builder setIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        id_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object name_ = "";
      /**
       * <pre>
       * The display name of the program. This is typically the colloquial name of
       * the tool used, for example 'bwa' or 'picard'.
       * </pre>
       *
       * <code>string name = 3;</code>
       * @return The name.
       */
      public java.lang.String getName() {
        java.lang.Object ref = name_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          name_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The display name of the program. This is typically the colloquial name of
       * the tool used, for example 'bwa' or 'picard'.
       * </pre>
       *
       * <code>string name = 3;</code>
       * @return The bytes for name.
       */
      public com.google.protobuf.ByteString
          getNameBytes() {
        java.lang.Object ref = name_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          name_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The display name of the program. This is typically the colloquial name of
       * the tool used, for example 'bwa' or 'picard'.
       * </pre>
       *
       * <code>string name = 3;</code>
       * @param value The name to set.
       * @return This builder for chaining.
       */
      public Builder setName(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        name_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The display name of the program. This is typically the colloquial name of
       * the tool used, for example 'bwa' or 'picard'.
       * </pre>
       *
       * <code>string name = 3;</code>
       * @return This builder for chaining.
       */
      public Builder clearName() {
        
        name_ = getDefaultInstance().getName();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The display name of the program. This is typically the colloquial name of
       * the tool used, for example 'bwa' or 'picard'.
       * </pre>
       *
       * <code>string name = 3;</code>
       * @param value The bytes for name to set.
       * @return This builder for chaining.
       */
      public Builder setNameBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        name_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object prevProgramId_ = "";
      /**
       * <pre>
       * The ID of the program run before this one.
       * </pre>
       *
       * <code>string prev_program_id = 4;</code>
       * @return The prevProgramId.
       */
      public java.lang.String getPrevProgramId() {
        java.lang.Object ref = prevProgramId_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          prevProgramId_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The ID of the program run before this one.
       * </pre>
       *
       * <code>string prev_program_id = 4;</code>
       * @return The bytes for prevProgramId.
       */
      public com.google.protobuf.ByteString
          getPrevProgramIdBytes() {
        java.lang.Object ref = prevProgramId_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          prevProgramId_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The ID of the program run before this one.
       * </pre>
       *
       * <code>string prev_program_id = 4;</code>
       * @param value The prevProgramId to set.
       * @return This builder for chaining.
       */
      public Builder setPrevProgramId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        prevProgramId_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The ID of the program run before this one.
       * </pre>
       *
       * <code>string prev_program_id = 4;</code>
       * @return This builder for chaining.
       */
      public Builder clearPrevProgramId() {
        
        prevProgramId_ = getDefaultInstance().getPrevProgramId();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The ID of the program run before this one.
       * </pre>
       *
       * <code>string prev_program_id = 4;</code>
       * @param value The bytes for prevProgramId to set.
       * @return This builder for chaining.
       */
      public Builder setPrevProgramIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        prevProgramId_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object version_ = "";
      /**
       * <pre>
       * The version of the program run.
       * </pre>
       *
       * <code>string version = 5;</code>
       * @return The version.
       */
      public java.lang.String getVersion() {
        java.lang.Object ref = version_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          version_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <pre>
       * The version of the program run.
       * </pre>
       *
       * <code>string version = 5;</code>
       * @return The bytes for version.
       */
      public com.google.protobuf.ByteString
          getVersionBytes() {
        java.lang.Object ref = version_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          version_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <pre>
       * The version of the program run.
       * </pre>
       *
       * <code>string version = 5;</code>
       * @param value The version to set.
       * @return This builder for chaining.
       */
      public Builder setVersion(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        version_ = value;
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The version of the program run.
       * </pre>
       *
       * <code>string version = 5;</code>
       * @return This builder for chaining.
       */
      public Builder clearVersion() {
        
        version_ = getDefaultInstance().getVersion();
        onChanged();
        return this;
      }
      /**
       * <pre>
       * The version of the program run.
       * </pre>
       *
       * <code>string version = 5;</code>
       * @param value The bytes for version to set.
       * @return This builder for chaining.
       */
      public Builder setVersionBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        version_ = value;
        onChanged();
        return this;
      }
      @java.lang.Override
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return super.setUnknownFields(unknownFields);
      }

      @java.lang.Override
      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return super.mergeUnknownFields(unknownFields);
      }


      // @@protoc_insertion_point(builder_scope:google.genomics.v1.ReadGroup.Program)
    }

    // @@protoc_insertion_point(class_scope:google.genomics.v1.ReadGroup.Program)
    private static final com.google.genomics.v1.ReadGroup.Program DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new com.google.genomics.v1.ReadGroup.Program();
    }

    public static com.google.genomics.v1.ReadGroup.Program getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<Program>
        PARSER = new com.google.protobuf.AbstractParser<Program>() {
      @java.lang.Override
      public Program parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return new Program(input, extensionRegistry);
      }
    };

    public static com.google.protobuf.Parser<Program> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<Program> getParserForType() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroup.Program getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public static final int ID_FIELD_NUMBER = 1;
  private volatile java.lang.Object id_;
  /**
   * <pre>
   * The server-generated read group ID, unique for all read groups.
   * Note: This is different than the &#64;RG ID field in the SAM spec. For that
   * value, see [name][google.genomics.v1.ReadGroup.name].
   * </pre>
   *
   * <code>string id = 1;</code>
   * @return The id.
   */
  @java.lang.Override
  public java.lang.String getId() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      id_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The server-generated read group ID, unique for all read groups.
   * Note: This is different than the &#64;RG ID field in the SAM spec. For that
   * value, see [name][google.genomics.v1.ReadGroup.name].
   * </pre>
   *
   * <code>string id = 1;</code>
   * @return The bytes for id.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getIdBytes() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      id_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int DATASET_ID_FIELD_NUMBER = 2;
  private volatile java.lang.Object datasetId_;
  /**
   * <pre>
   * The dataset to which this read group belongs.
   * </pre>
   *
   * <code>string dataset_id = 2;</code>
   * @return The datasetId.
   */
  @java.lang.Override
  public java.lang.String getDatasetId() {
    java.lang.Object ref = datasetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      datasetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The dataset to which this read group belongs.
   * </pre>
   *
   * <code>string dataset_id = 2;</code>
   * @return The bytes for datasetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getDatasetIdBytes() {
    java.lang.Object ref = datasetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      datasetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int NAME_FIELD_NUMBER = 3;
  private volatile java.lang.Object name_;
  /**
   * <pre>
   * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
   * </pre>
   *
   * <code>string name = 3;</code>
   * @return The name.
   */
  @java.lang.Override
  public java.lang.String getName() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      name_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
   * </pre>
   *
   * <code>string name = 3;</code>
   * @return The bytes for name.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getNameBytes() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      name_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int DESCRIPTION_FIELD_NUMBER = 4;
  private volatile java.lang.Object description_;
  /**
   * <pre>
   * A free-form text description of this read group.
   * </pre>
   *
   * <code>string description = 4;</code>
   * @return The description.
   */
  @java.lang.Override
  public java.lang.String getDescription() {
    java.lang.Object ref = description_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      description_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * A free-form text description of this read group.
   * </pre>
   *
   * <code>string description = 4;</code>
   * @return The bytes for description.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getDescriptionBytes() {
    java.lang.Object ref = description_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      description_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int SAMPLE_ID_FIELD_NUMBER = 5;
  private volatile java.lang.Object sampleId_;
  /**
   * <pre>
   * A client-supplied sample identifier for the reads in this read group.
   * </pre>
   *
   * <code>string sample_id = 5;</code>
   * @return The sampleId.
   */
  @java.lang.Override
  public java.lang.String getSampleId() {
    java.lang.Object ref = sampleId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      sampleId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * A client-supplied sample identifier for the reads in this read group.
   * </pre>
   *
   * <code>string sample_id = 5;</code>
   * @return The bytes for sampleId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getSampleIdBytes() {
    java.lang.Object ref = sampleId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      sampleId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int EXPERIMENT_FIELD_NUMBER = 6;
  private com.google.genomics.v1.ReadGroup.Experiment experiment_;
  /**
   * <pre>
   * The experiment used to generate this read group.
   * </pre>
   *
   * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
   * @return Whether the experiment field is set.
   */
  @java.lang.Override
  public boolean hasExperiment() {
    return experiment_ != null;
  }
  /**
   * <pre>
   * The experiment used to generate this read group.
   * </pre>
   *
   * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
   * @return The experiment.
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadGroup.Experiment getExperiment() {
    return experiment_ == null ? com.google.genomics.v1.ReadGroup.Experiment.getDefaultInstance() : experiment_;
  }
  /**
   * <pre>
   * The experiment used to generate this read group.
   * </pre>
   *
   * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadGroup.ExperimentOrBuilder getExperimentOrBuilder() {
    return getExperiment();
  }

  public static final int PREDICTED_INSERT_SIZE_FIELD_NUMBER = 7;
  private int predictedInsertSize_;
  /**
   * <pre>
   * The predicted insert size of this read group. The insert size is the length
   * the sequenced DNA fragment from end-to-end, not including the adapters.
   * </pre>
   *
   * <code>int32 predicted_insert_size = 7;</code>
   * @return The predictedInsertSize.
   */
  @java.lang.Override
  public int getPredictedInsertSize() {
    return predictedInsertSize_;
  }

  public static final int PROGRAMS_FIELD_NUMBER = 10;
  private java.util.List<com.google.genomics.v1.ReadGroup.Program> programs_;
  /**
   * <pre>
   * The programs used to generate this read group. Programs are always
   * identical for all read groups within a read group set. For this reason,
   * only the first read group in a returned set will have this field
   * populated.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
   */
  @java.lang.Override
  public java.util.List<com.google.genomics.v1.ReadGroup.Program> getProgramsList() {
    return programs_;
  }
  /**
   * <pre>
   * The programs used to generate this read group. Programs are always
   * identical for all read groups within a read group set. For this reason,
   * only the first read group in a returned set will have this field
   * populated.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
   */
  @java.lang.Override
  public java.util.List<? extends com.google.genomics.v1.ReadGroup.ProgramOrBuilder> 
      getProgramsOrBuilderList() {
    return programs_;
  }
  /**
   * <pre>
   * The programs used to generate this read group. Programs are always
   * identical for all read groups within a read group set. For this reason,
   * only the first read group in a returned set will have this field
   * populated.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
   */
  @java.lang.Override
  public int getProgramsCount() {
    return programs_.size();
  }
  /**
   * <pre>
   * The programs used to generate this read group. Programs are always
   * identical for all read groups within a read group set. For this reason,
   * only the first read group in a returned set will have this field
   * populated.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadGroup.Program getPrograms(int index) {
    return programs_.get(index);
  }
  /**
   * <pre>
   * The programs used to generate this read group. Programs are always
   * identical for all read groups within a read group set. For this reason,
   * only the first read group in a returned set will have this field
   * populated.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadGroup.ProgramOrBuilder getProgramsOrBuilder(
      int index) {
    return programs_.get(index);
  }

  public static final int REFERENCE_SET_ID_FIELD_NUMBER = 11;
  private volatile java.lang.Object referenceSetId_;
  /**
   * <pre>
   * The reference set the reads in this read group are aligned to.
   * </pre>
   *
   * <code>string reference_set_id = 11;</code>
   * @return The referenceSetId.
   */
  @java.lang.Override
  public java.lang.String getReferenceSetId() {
    java.lang.Object ref = referenceSetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      referenceSetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The reference set the reads in this read group are aligned to.
   * </pre>
   *
   * <code>string reference_set_id = 11;</code>
   * @return The bytes for referenceSetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getReferenceSetIdBytes() {
    java.lang.Object ref = referenceSetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      referenceSetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int INFO_FIELD_NUMBER = 12;
  private static final class InfoDefaultEntryHolder {
    static final com.google.protobuf.MapEntry<
        java.lang.String, com.google.protobuf.ListValue> defaultEntry =
            com.google.protobuf.MapEntry
            .<java.lang.String, com.google.protobuf.ListValue>newDefaultInstance(
                com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_InfoEntry_descriptor, 
                com.google.protobuf.WireFormat.FieldType.STRING,
                "",
                com.google.protobuf.WireFormat.FieldType.MESSAGE,
                com.google.protobuf.ListValue.getDefaultInstance());
  }
  private com.google.protobuf.MapField<
      java.lang.String, com.google.protobuf.ListValue> info_;
  private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
  internalGetInfo() {
    if (info_ == null) {
      return com.google.protobuf.MapField.emptyMapField(
          InfoDefaultEntryHolder.defaultEntry);
    }
    return info_;
  }

  public int getInfoCount() {
    return internalGetInfo().getMap().size();
  }
  /**
   * <pre>
   * A map of additional read group information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */

  @java.lang.Override
  public boolean containsInfo(
      java.lang.String key) {
    if (key == null) { throw new NullPointerException("map key"); }
    return internalGetInfo().getMap().containsKey(key);
  }
  /**
   * Use {@link #getInfoMap()} instead.
   */
  @java.lang.Override
  @java.lang.Deprecated
  public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfo() {
    return getInfoMap();
  }
  /**
   * <pre>
   * A map of additional read group information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */
  @java.lang.Override

  public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfoMap() {
    return internalGetInfo().getMap();
  }
  /**
   * <pre>
   * A map of additional read group information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */
  @java.lang.Override

  public com.google.protobuf.ListValue getInfoOrDefault(
      java.lang.String key,
      com.google.protobuf.ListValue defaultValue) {
    if (key == null) { throw new NullPointerException("map key"); }
    java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
        internalGetInfo().getMap();
    return map.containsKey(key) ? map.get(key) : defaultValue;
  }
  /**
   * <pre>
   * A map of additional read group information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
   */
  @java.lang.Override

  public com.google.protobuf.ListValue getInfoOrThrow(
      java.lang.String key) {
    if (key == null) { throw new NullPointerException("map key"); }
    java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
        internalGetInfo().getMap();
    if (!map.containsKey(key)) {
      throw new java.lang.IllegalArgumentException();
    }
    return map.get(key);
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 1, id_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(datasetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 2, datasetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 3, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(description_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 4, description_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(sampleId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 5, sampleId_);
    }
    if (experiment_ != null) {
      output.writeMessage(6, getExperiment());
    }
    if (predictedInsertSize_ != 0) {
      output.writeInt32(7, predictedInsertSize_);
    }
    for (int i = 0; i < programs_.size(); i++) {
      output.writeMessage(10, programs_.get(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(referenceSetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 11, referenceSetId_);
    }
    com.google.protobuf.GeneratedMessageV3
      .serializeStringMapTo(
        output,
        internalGetInfo(),
        InfoDefaultEntryHolder.defaultEntry,
        12);
    unknownFields.writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, id_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(datasetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(2, datasetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(3, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(description_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(4, description_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(sampleId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(5, sampleId_);
    }
    if (experiment_ != null) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(6, getExperiment());
    }
    if (predictedInsertSize_ != 0) {
      size += com.google.protobuf.CodedOutputStream
        .computeInt32Size(7, predictedInsertSize_);
    }
    for (int i = 0; i < programs_.size(); i++) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(10, programs_.get(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(referenceSetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(11, referenceSetId_);
    }
    for (java.util.Map.Entry<java.lang.String, com.google.protobuf.ListValue> entry
         : internalGetInfo().getMap().entrySet()) {
      com.google.protobuf.MapEntry<java.lang.String, com.google.protobuf.ListValue>
      info__ = InfoDefaultEntryHolder.defaultEntry.newBuilderForType()
          .setKey(entry.getKey())
          .setValue(entry.getValue())
          .build();
      size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(12, info__);
    }
    size += unknownFields.getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.ReadGroup)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.ReadGroup other = (com.google.genomics.v1.ReadGroup) obj;

    if (!getId()
        .equals(other.getId())) return false;
    if (!getDatasetId()
        .equals(other.getDatasetId())) return false;
    if (!getName()
        .equals(other.getName())) return false;
    if (!getDescription()
        .equals(other.getDescription())) return false;
    if (!getSampleId()
        .equals(other.getSampleId())) return false;
    if (hasExperiment() != other.hasExperiment()) return false;
    if (hasExperiment()) {
      if (!getExperiment()
          .equals(other.getExperiment())) return false;
    }
    if (getPredictedInsertSize()
        != other.getPredictedInsertSize()) return false;
    if (!getProgramsList()
        .equals(other.getProgramsList())) return false;
    if (!getReferenceSetId()
        .equals(other.getReferenceSetId())) return false;
    if (!internalGetInfo().equals(
        other.internalGetInfo())) return false;
    if (!unknownFields.equals(other.unknownFields)) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    hash = (37 * hash) + ID_FIELD_NUMBER;
    hash = (53 * hash) + getId().hashCode();
    hash = (37 * hash) + DATASET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getDatasetId().hashCode();
    hash = (37 * hash) + NAME_FIELD_NUMBER;
    hash = (53 * hash) + getName().hashCode();
    hash = (37 * hash) + DESCRIPTION_FIELD_NUMBER;
    hash = (53 * hash) + getDescription().hashCode();
    hash = (37 * hash) + SAMPLE_ID_FIELD_NUMBER;
    hash = (53 * hash) + getSampleId().hashCode();
    if (hasExperiment()) {
      hash = (37 * hash) + EXPERIMENT_FIELD_NUMBER;
      hash = (53 * hash) + getExperiment().hashCode();
    }
    hash = (37 * hash) + PREDICTED_INSERT_SIZE_FIELD_NUMBER;
    hash = (53 * hash) + getPredictedInsertSize();
    if (getProgramsCount() > 0) {
      hash = (37 * hash) + PROGRAMS_FIELD_NUMBER;
      hash = (53 * hash) + getProgramsList().hashCode();
    }
    hash = (37 * hash) + REFERENCE_SET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getReferenceSetId().hashCode();
    if (!internalGetInfo().getMap().isEmpty()) {
      hash = (37 * hash) + INFO_FIELD_NUMBER;
      hash = (53 * hash) + internalGetInfo().hashCode();
    }
    hash = (29 * hash) + unknownFields.hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.ReadGroup parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroup parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.ReadGroup parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.ReadGroup parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.ReadGroup prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A read group is all the data that's processed the same way by the sequencer.
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.ReadGroup}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.ReadGroup)
      com.google.genomics.v1.ReadGroupOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 12:
          return internalGetInfo();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMutableMapField(
        int number) {
      switch (number) {
        case 12:
          return internalGetMutableInfo();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.ReadGroup.class, com.google.genomics.v1.ReadGroup.Builder.class);
    }

    // Construct using com.google.genomics.v1.ReadGroup.newBuilder()
    private Builder() {
      maybeForceBuilderInitialization();
    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);
      maybeForceBuilderInitialization();
    }
    private void maybeForceBuilderInitialization() {
      if (com.google.protobuf.GeneratedMessageV3
              .alwaysUseFieldBuilders) {
        getProgramsFieldBuilder();
      }
    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      id_ = "";

      datasetId_ = "";

      name_ = "";

      description_ = "";

      sampleId_ = "";

      if (experimentBuilder_ == null) {
        experiment_ = null;
      } else {
        experiment_ = null;
        experimentBuilder_ = null;
      }
      predictedInsertSize_ = 0;

      if (programsBuilder_ == null) {
        programs_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
      } else {
        programsBuilder_.clear();
      }
      referenceSetId_ = "";

      internalGetMutableInfo().clear();
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReadGroupProto.internal_static_google_genomics_v1_ReadGroup_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroup getDefaultInstanceForType() {
      return com.google.genomics.v1.ReadGroup.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroup build() {
      com.google.genomics.v1.ReadGroup result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroup buildPartial() {
      com.google.genomics.v1.ReadGroup result = new com.google.genomics.v1.ReadGroup(this);
      int from_bitField0_ = bitField0_;
      result.id_ = id_;
      result.datasetId_ = datasetId_;
      result.name_ = name_;
      result.description_ = description_;
      result.sampleId_ = sampleId_;
      if (experimentBuilder_ == null) {
        result.experiment_ = experiment_;
      } else {
        result.experiment_ = experimentBuilder_.build();
      }
      result.predictedInsertSize_ = predictedInsertSize_;
      if (programsBuilder_ == null) {
        if (((bitField0_ & 0x00000001) != 0)) {
          programs_ = java.util.Collections.unmodifiableList(programs_);
          bitField0_ = (bitField0_ & ~0x00000001);
        }
        result.programs_ = programs_;
      } else {
        result.programs_ = programsBuilder_.build();
      }
      result.referenceSetId_ = referenceSetId_;
      result.info_ = internalGetInfo();
      result.info_.makeImmutable();
      onBuilt();
      return result;
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.ReadGroup) {
        return mergeFrom((com.google.genomics.v1.ReadGroup)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.ReadGroup other) {
      if (other == com.google.genomics.v1.ReadGroup.getDefaultInstance()) return this;
      if (!other.getId().isEmpty()) {
        id_ = other.id_;
        onChanged();
      }
      if (!other.getDatasetId().isEmpty()) {
        datasetId_ = other.datasetId_;
        onChanged();
      }
      if (!other.getName().isEmpty()) {
        name_ = other.name_;
        onChanged();
      }
      if (!other.getDescription().isEmpty()) {
        description_ = other.description_;
        onChanged();
      }
      if (!other.getSampleId().isEmpty()) {
        sampleId_ = other.sampleId_;
        onChanged();
      }
      if (other.hasExperiment()) {
        mergeExperiment(other.getExperiment());
      }
      if (other.getPredictedInsertSize() != 0) {
        setPredictedInsertSize(other.getPredictedInsertSize());
      }
      if (programsBuilder_ == null) {
        if (!other.programs_.isEmpty()) {
          if (programs_.isEmpty()) {
            programs_ = other.programs_;
            bitField0_ = (bitField0_ & ~0x00000001);
          } else {
            ensureProgramsIsMutable();
            programs_.addAll(other.programs_);
          }
          onChanged();
        }
      } else {
        if (!other.programs_.isEmpty()) {
          if (programsBuilder_.isEmpty()) {
            programsBuilder_.dispose();
            programsBuilder_ = null;
            programs_ = other.programs_;
            bitField0_ = (bitField0_ & ~0x00000001);
            programsBuilder_ = 
              com.google.protobuf.GeneratedMessageV3.alwaysUseFieldBuilders ?
                 getProgramsFieldBuilder() : null;
          } else {
            programsBuilder_.addAllMessages(other.programs_);
          }
        }
      }
      if (!other.getReferenceSetId().isEmpty()) {
        referenceSetId_ = other.referenceSetId_;
        onChanged();
      }
      internalGetMutableInfo().mergeFrom(
          other.internalGetInfo());
      this.mergeUnknownFields(other.unknownFields);
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      com.google.genomics.v1.ReadGroup parsedMessage = null;
      try {
        parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        parsedMessage = (com.google.genomics.v1.ReadGroup) e.getUnfinishedMessage();
        throw e.unwrapIOException();
      } finally {
        if (parsedMessage != null) {
          mergeFrom(parsedMessage);
        }
      }
      return this;
    }
    private int bitField0_;

    private java.lang.Object id_ = "";
    /**
     * <pre>
     * The server-generated read group ID, unique for all read groups.
     * Note: This is different than the &#64;RG ID field in the SAM spec. For that
     * value, see [name][google.genomics.v1.ReadGroup.name].
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return The id.
     */
    public java.lang.String getId() {
      java.lang.Object ref = id_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        id_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The server-generated read group ID, unique for all read groups.
     * Note: This is different than the &#64;RG ID field in the SAM spec. For that
     * value, see [name][google.genomics.v1.ReadGroup.name].
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return The bytes for id.
     */
    public com.google.protobuf.ByteString
        getIdBytes() {
      java.lang.Object ref = id_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        id_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The server-generated read group ID, unique for all read groups.
     * Note: This is different than the &#64;RG ID field in the SAM spec. For that
     * value, see [name][google.genomics.v1.ReadGroup.name].
     * </pre>
     *
     * <code>string id = 1;</code>
     * @param value The id to set.
     * @return This builder for chaining.
     */
    public Builder setId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      id_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated read group ID, unique for all read groups.
     * Note: This is different than the &#64;RG ID field in the SAM spec. For that
     * value, see [name][google.genomics.v1.ReadGroup.name].
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return This builder for chaining.
     */
    public Builder clearId() {
      
      id_ = getDefaultInstance().getId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated read group ID, unique for all read groups.
     * Note: This is different than the &#64;RG ID field in the SAM spec. For that
     * value, see [name][google.genomics.v1.ReadGroup.name].
     * </pre>
     *
     * <code>string id = 1;</code>
     * @param value The bytes for id to set.
     * @return This builder for chaining.
     */
    public Builder setIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      id_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object datasetId_ = "";
    /**
     * <pre>
     * The dataset to which this read group belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @return The datasetId.
     */
    public java.lang.String getDatasetId() {
      java.lang.Object ref = datasetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        datasetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The dataset to which this read group belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @return The bytes for datasetId.
     */
    public com.google.protobuf.ByteString
        getDatasetIdBytes() {
      java.lang.Object ref = datasetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        datasetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The dataset to which this read group belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @param value The datasetId to set.
     * @return This builder for chaining.
     */
    public Builder setDatasetId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      datasetId_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The dataset to which this read group belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearDatasetId() {
      
      datasetId_ = getDefaultInstance().getDatasetId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The dataset to which this read group belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @param value The bytes for datasetId to set.
     * @return This builder for chaining.
     */
    public Builder setDatasetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      datasetId_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object name_ = "";
    /**
     * <pre>
     * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return The name.
     */
    public java.lang.String getName() {
      java.lang.Object ref = name_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        name_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return The bytes for name.
     */
    public com.google.protobuf.ByteString
        getNameBytes() {
      java.lang.Object ref = name_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        name_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @param value The name to set.
     * @return This builder for chaining.
     */
    public Builder setName(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      name_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @return This builder for chaining.
     */
    public Builder clearName() {
      
      name_ = getDefaultInstance().getName();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The read group name. This corresponds to the &#64;RG ID field in the SAM spec.
     * </pre>
     *
     * <code>string name = 3;</code>
     * @param value The bytes for name to set.
     * @return This builder for chaining.
     */
    public Builder setNameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      name_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object description_ = "";
    /**
     * <pre>
     * A free-form text description of this read group.
     * </pre>
     *
     * <code>string description = 4;</code>
     * @return The description.
     */
    public java.lang.String getDescription() {
      java.lang.Object ref = description_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        description_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * A free-form text description of this read group.
     * </pre>
     *
     * <code>string description = 4;</code>
     * @return The bytes for description.
     */
    public com.google.protobuf.ByteString
        getDescriptionBytes() {
      java.lang.Object ref = description_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        description_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * A free-form text description of this read group.
     * </pre>
     *
     * <code>string description = 4;</code>
     * @param value The description to set.
     * @return This builder for chaining.
     */
    public Builder setDescription(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      description_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * A free-form text description of this read group.
     * </pre>
     *
     * <code>string description = 4;</code>
     * @return This builder for chaining.
     */
    public Builder clearDescription() {
      
      description_ = getDefaultInstance().getDescription();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * A free-form text description of this read group.
     * </pre>
     *
     * <code>string description = 4;</code>
     * @param value The bytes for description to set.
     * @return This builder for chaining.
     */
    public Builder setDescriptionBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      description_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object sampleId_ = "";
    /**
     * <pre>
     * A client-supplied sample identifier for the reads in this read group.
     * </pre>
     *
     * <code>string sample_id = 5;</code>
     * @return The sampleId.
     */
    public java.lang.String getSampleId() {
      java.lang.Object ref = sampleId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        sampleId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * A client-supplied sample identifier for the reads in this read group.
     * </pre>
     *
     * <code>string sample_id = 5;</code>
     * @return The bytes for sampleId.
     */
    public com.google.protobuf.ByteString
        getSampleIdBytes() {
      java.lang.Object ref = sampleId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        sampleId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * A client-supplied sample identifier for the reads in this read group.
     * </pre>
     *
     * <code>string sample_id = 5;</code>
     * @param value The sampleId to set.
     * @return This builder for chaining.
     */
    public Builder setSampleId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      sampleId_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * A client-supplied sample identifier for the reads in this read group.
     * </pre>
     *
     * <code>string sample_id = 5;</code>
     * @return This builder for chaining.
     */
    public Builder clearSampleId() {
      
      sampleId_ = getDefaultInstance().getSampleId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * A client-supplied sample identifier for the reads in this read group.
     * </pre>
     *
     * <code>string sample_id = 5;</code>
     * @param value The bytes for sampleId to set.
     * @return This builder for chaining.
     */
    public Builder setSampleIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      sampleId_ = value;
      onChanged();
      return this;
    }

    private com.google.genomics.v1.ReadGroup.Experiment experiment_;
    private com.google.protobuf.SingleFieldBuilderV3<
        com.google.genomics.v1.ReadGroup.Experiment, com.google.genomics.v1.ReadGroup.Experiment.Builder, com.google.genomics.v1.ReadGroup.ExperimentOrBuilder> experimentBuilder_;
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     * @return Whether the experiment field is set.
     */
    public boolean hasExperiment() {
      return experimentBuilder_ != null || experiment_ != null;
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     * @return The experiment.
     */
    public com.google.genomics.v1.ReadGroup.Experiment getExperiment() {
      if (experimentBuilder_ == null) {
        return experiment_ == null ? com.google.genomics.v1.ReadGroup.Experiment.getDefaultInstance() : experiment_;
      } else {
        return experimentBuilder_.getMessage();
      }
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    public Builder setExperiment(com.google.genomics.v1.ReadGroup.Experiment value) {
      if (experimentBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        experiment_ = value;
        onChanged();
      } else {
        experimentBuilder_.setMessage(value);
      }

      return this;
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    public Builder setExperiment(
        com.google.genomics.v1.ReadGroup.Experiment.Builder builderForValue) {
      if (experimentBuilder_ == null) {
        experiment_ = builderForValue.build();
        onChanged();
      } else {
        experimentBuilder_.setMessage(builderForValue.build());
      }

      return this;
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    public Builder mergeExperiment(com.google.genomics.v1.ReadGroup.Experiment value) {
      if (experimentBuilder_ == null) {
        if (experiment_ != null) {
          experiment_ =
            com.google.genomics.v1.ReadGroup.Experiment.newBuilder(experiment_).mergeFrom(value).buildPartial();
        } else {
          experiment_ = value;
        }
        onChanged();
      } else {
        experimentBuilder_.mergeFrom(value);
      }

      return this;
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    public Builder clearExperiment() {
      if (experimentBuilder_ == null) {
        experiment_ = null;
        onChanged();
      } else {
        experiment_ = null;
        experimentBuilder_ = null;
      }

      return this;
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    public com.google.genomics.v1.ReadGroup.Experiment.Builder getExperimentBuilder() {
      
      onChanged();
      return getExperimentFieldBuilder().getBuilder();
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    public com.google.genomics.v1.ReadGroup.ExperimentOrBuilder getExperimentOrBuilder() {
      if (experimentBuilder_ != null) {
        return experimentBuilder_.getMessageOrBuilder();
      } else {
        return experiment_ == null ?
            com.google.genomics.v1.ReadGroup.Experiment.getDefaultInstance() : experiment_;
      }
    }
    /**
     * <pre>
     * The experiment used to generate this read group.
     * </pre>
     *
     * <code>.google.genomics.v1.ReadGroup.Experiment experiment = 6;</code>
     */
    private com.google.protobuf.SingleFieldBuilderV3<
        com.google.genomics.v1.ReadGroup.Experiment, com.google.genomics.v1.ReadGroup.Experiment.Builder, com.google.genomics.v1.ReadGroup.ExperimentOrBuilder> 
        getExperimentFieldBuilder() {
      if (experimentBuilder_ == null) {
        experimentBuilder_ = new com.google.protobuf.SingleFieldBuilderV3<
            com.google.genomics.v1.ReadGroup.Experiment, com.google.genomics.v1.ReadGroup.Experiment.Builder, com.google.genomics.v1.ReadGroup.ExperimentOrBuilder>(
                getExperiment(),
                getParentForChildren(),
                isClean());
        experiment_ = null;
      }
      return experimentBuilder_;
    }

    private int predictedInsertSize_ ;
    /**
     * <pre>
     * The predicted insert size of this read group. The insert size is the length
     * the sequenced DNA fragment from end-to-end, not including the adapters.
     * </pre>
     *
     * <code>int32 predicted_insert_size = 7;</code>
     * @return The predictedInsertSize.
     */
    @java.lang.Override
    public int getPredictedInsertSize() {
      return predictedInsertSize_;
    }
    /**
     * <pre>
     * The predicted insert size of this read group. The insert size is the length
     * the sequenced DNA fragment from end-to-end, not including the adapters.
     * </pre>
     *
     * <code>int32 predicted_insert_size = 7;</code>
     * @param value The predictedInsertSize to set.
     * @return This builder for chaining.
     */
    public Builder setPredictedInsertSize(int value) {
      
      predictedInsertSize_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The predicted insert size of this read group. The insert size is the length
     * the sequenced DNA fragment from end-to-end, not including the adapters.
     * </pre>
     *
     * <code>int32 predicted_insert_size = 7;</code>
     * @return This builder for chaining.
     */
    public Builder clearPredictedInsertSize() {
      
      predictedInsertSize_ = 0;
      onChanged();
      return this;
    }

    private java.util.List<com.google.genomics.v1.ReadGroup.Program> programs_ =
      java.util.Collections.emptyList();
    private void ensureProgramsIsMutable() {
      if (!((bitField0_ & 0x00000001) != 0)) {
        programs_ = new java.util.ArrayList<com.google.genomics.v1.ReadGroup.Program>(programs_);
        bitField0_ |= 0x00000001;
       }
    }

    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.ReadGroup.Program, com.google.genomics.v1.ReadGroup.Program.Builder, com.google.genomics.v1.ReadGroup.ProgramOrBuilder> programsBuilder_;

    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public java.util.List<com.google.genomics.v1.ReadGroup.Program> getProgramsList() {
      if (programsBuilder_ == null) {
        return java.util.Collections.unmodifiableList(programs_);
      } else {
        return programsBuilder_.getMessageList();
      }
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public int getProgramsCount() {
      if (programsBuilder_ == null) {
        return programs_.size();
      } else {
        return programsBuilder_.getCount();
      }
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public com.google.genomics.v1.ReadGroup.Program getPrograms(int index) {
      if (programsBuilder_ == null) {
        return programs_.get(index);
      } else {
        return programsBuilder_.getMessage(index);
      }
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder setPrograms(
        int index, com.google.genomics.v1.ReadGroup.Program value) {
      if (programsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureProgramsIsMutable();
        programs_.set(index, value);
        onChanged();
      } else {
        programsBuilder_.setMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder setPrograms(
        int index, com.google.genomics.v1.ReadGroup.Program.Builder builderForValue) {
      if (programsBuilder_ == null) {
        ensureProgramsIsMutable();
        programs_.set(index, builderForValue.build());
        onChanged();
      } else {
        programsBuilder_.setMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder addPrograms(com.google.genomics.v1.ReadGroup.Program value) {
      if (programsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureProgramsIsMutable();
        programs_.add(value);
        onChanged();
      } else {
        programsBuilder_.addMessage(value);
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder addPrograms(
        int index, com.google.genomics.v1.ReadGroup.Program value) {
      if (programsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureProgramsIsMutable();
        programs_.add(index, value);
        onChanged();
      } else {
        programsBuilder_.addMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder addPrograms(
        com.google.genomics.v1.ReadGroup.Program.Builder builderForValue) {
      if (programsBuilder_ == null) {
        ensureProgramsIsMutable();
        programs_.add(builderForValue.build());
        onChanged();
      } else {
        programsBuilder_.addMessage(builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder addPrograms(
        int index, com.google.genomics.v1.ReadGroup.Program.Builder builderForValue) {
      if (programsBuilder_ == null) {
        ensureProgramsIsMutable();
        programs_.add(index, builderForValue.build());
        onChanged();
      } else {
        programsBuilder_.addMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder addAllPrograms(
        java.lang.Iterable<? extends com.google.genomics.v1.ReadGroup.Program> values) {
      if (programsBuilder_ == null) {
        ensureProgramsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, programs_);
        onChanged();
      } else {
        programsBuilder_.addAllMessages(values);
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder clearPrograms() {
      if (programsBuilder_ == null) {
        programs_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
        onChanged();
      } else {
        programsBuilder_.clear();
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public Builder removePrograms(int index) {
      if (programsBuilder_ == null) {
        ensureProgramsIsMutable();
        programs_.remove(index);
        onChanged();
      } else {
        programsBuilder_.remove(index);
      }
      return this;
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public com.google.genomics.v1.ReadGroup.Program.Builder getProgramsBuilder(
        int index) {
      return getProgramsFieldBuilder().getBuilder(index);
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public com.google.genomics.v1.ReadGroup.ProgramOrBuilder getProgramsOrBuilder(
        int index) {
      if (programsBuilder_ == null) {
        return programs_.get(index);  } else {
        return programsBuilder_.getMessageOrBuilder(index);
      }
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public java.util.List<? extends com.google.genomics.v1.ReadGroup.ProgramOrBuilder> 
         getProgramsOrBuilderList() {
      if (programsBuilder_ != null) {
        return programsBuilder_.getMessageOrBuilderList();
      } else {
        return java.util.Collections.unmodifiableList(programs_);
      }
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public com.google.genomics.v1.ReadGroup.Program.Builder addProgramsBuilder() {
      return getProgramsFieldBuilder().addBuilder(
          com.google.genomics.v1.ReadGroup.Program.getDefaultInstance());
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public com.google.genomics.v1.ReadGroup.Program.Builder addProgramsBuilder(
        int index) {
      return getProgramsFieldBuilder().addBuilder(
          index, com.google.genomics.v1.ReadGroup.Program.getDefaultInstance());
    }
    /**
     * <pre>
     * The programs used to generate this read group. Programs are always
     * identical for all read groups within a read group set. For this reason,
     * only the first read group in a returned set will have this field
     * populated.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup.Program programs = 10;</code>
     */
    public java.util.List<com.google.genomics.v1.ReadGroup.Program.Builder> 
         getProgramsBuilderList() {
      return getProgramsFieldBuilder().getBuilderList();
    }
    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.ReadGroup.Program, com.google.genomics.v1.ReadGroup.Program.Builder, com.google.genomics.v1.ReadGroup.ProgramOrBuilder> 
        getProgramsFieldBuilder() {
      if (programsBuilder_ == null) {
        programsBuilder_ = new com.google.protobuf.RepeatedFieldBuilderV3<
            com.google.genomics.v1.ReadGroup.Program, com.google.genomics.v1.ReadGroup.Program.Builder, com.google.genomics.v1.ReadGroup.ProgramOrBuilder>(
                programs_,
                ((bitField0_ & 0x00000001) != 0),
                getParentForChildren(),
                isClean());
        programs_ = null;
      }
      return programsBuilder_;
    }

    private java.lang.Object referenceSetId_ = "";
    /**
     * <pre>
     * The reference set the reads in this read group are aligned to.
     * </pre>
     *
     * <code>string reference_set_id = 11;</code>
     * @return The referenceSetId.
     */
    public java.lang.String getReferenceSetId() {
      java.lang.Object ref = referenceSetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        referenceSetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The reference set the reads in this read group are aligned to.
     * </pre>
     *
     * <code>string reference_set_id = 11;</code>
     * @return The bytes for referenceSetId.
     */
    public com.google.protobuf.ByteString
        getReferenceSetIdBytes() {
      java.lang.Object ref = referenceSetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        referenceSetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The reference set the reads in this read group are aligned to.
     * </pre>
     *
     * <code>string reference_set_id = 11;</code>
     * @param value The referenceSetId to set.
     * @return This builder for chaining.
     */
    public Builder setReferenceSetId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      referenceSetId_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The reference set the reads in this read group are aligned to.
     * </pre>
     *
     * <code>string reference_set_id = 11;</code>
     * @return This builder for chaining.
     */
    public Builder clearReferenceSetId() {
      
      referenceSetId_ = getDefaultInstance().getReferenceSetId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The reference set the reads in this read group are aligned to.
     * </pre>
     *
     * <code>string reference_set_id = 11;</code>
     * @param value The bytes for referenceSetId to set.
     * @return This builder for chaining.
     */
    public Builder setReferenceSetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      referenceSetId_ = value;
      onChanged();
      return this;
    }

    private com.google.protobuf.MapField<
        java.lang.String, com.google.protobuf.ListValue> info_;
    private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
    internalGetInfo() {
      if (info_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            InfoDefaultEntryHolder.defaultEntry);
      }
      return info_;
    }
    private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
    internalGetMutableInfo() {
      onChanged();;
      if (info_ == null) {
        info_ = com.google.protobuf.MapField.newMapField(
            InfoDefaultEntryHolder.defaultEntry);
      }
      if (!info_.isMutable()) {
        info_ = info_.copy();
      }
      return info_;
    }

    public int getInfoCount() {
      return internalGetInfo().getMap().size();
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */

    @java.lang.Override
    public boolean containsInfo(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      return internalGetInfo().getMap().containsKey(key);
    }
    /**
     * Use {@link #getInfoMap()} instead.
     */
    @java.lang.Override
    @java.lang.Deprecated
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfo() {
      return getInfoMap();
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    @java.lang.Override

    public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfoMap() {
      return internalGetInfo().getMap();
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    @java.lang.Override

    public com.google.protobuf.ListValue getInfoOrDefault(
        java.lang.String key,
        com.google.protobuf.ListValue defaultValue) {
      if (key == null) { throw new NullPointerException("map key"); }
      java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
          internalGetInfo().getMap();
      return map.containsKey(key) ? map.get(key) : defaultValue;
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    @java.lang.Override

    public com.google.protobuf.ListValue getInfoOrThrow(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
          internalGetInfo().getMap();
      if (!map.containsKey(key)) {
        throw new java.lang.IllegalArgumentException();
      }
      return map.get(key);
    }

    public Builder clearInfo() {
      internalGetMutableInfo().getMutableMap()
          .clear();
      return this;
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */

    public Builder removeInfo(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      internalGetMutableInfo().getMutableMap()
          .remove(key);
      return this;
    }
    /**
     * Use alternate mutation accessors instead.
     */
    @java.lang.Deprecated
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue>
    getMutableInfo() {
      return internalGetMutableInfo().getMutableMap();
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */
    public Builder putInfo(
        java.lang.String key,
        com.google.protobuf.ListValue value) {
      if (key == null) { throw new NullPointerException("map key"); }
      if (value == null) {
  throw new NullPointerException("map value");
}

      internalGetMutableInfo().getMutableMap()
          .put(key, value);
      return this;
    }
    /**
     * <pre>
     * A map of additional read group information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 12;</code>
     */

    public Builder putAllInfo(
        java.util.Map<java.lang.String, com.google.protobuf.ListValue> values) {
      internalGetMutableInfo().getMutableMap()
          .putAll(values);
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.ReadGroup)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.ReadGroup)
  private static final com.google.genomics.v1.ReadGroup DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.ReadGroup();
  }

  public static com.google.genomics.v1.ReadGroup getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<ReadGroup>
      PARSER = new com.google.protobuf.AbstractParser<ReadGroup>() {
    @java.lang.Override
    public ReadGroup parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return new ReadGroup(input, extensionRegistry);
    }
  };

  public static com.google.protobuf.Parser<ReadGroup> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<ReadGroup> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.ReadGroup getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

