// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/readgroupset.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A read group set is a logical collection of read groups, which are
 * collections of reads produced by a sequencer. A read group set typically
 * models reads corresponding to one sample, sequenced one way, and aligned one
 * way.
 * * A read group set belongs to one dataset.
 * * A read group belongs to one read group set.
 * * A read belongs to one read group.
 * For more genomics resource definitions, see [Fundamentals of Google
 * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.ReadGroupSet}
 */
public final class ReadGroupSet extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.ReadGroupSet)
    ReadGroupSetOrBuilder {
private static final long serialVersionUID = 0L;
  // Use ReadGroupSet.newBuilder() to construct.
  private ReadGroupSet(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private ReadGroupSet() {
    id_ = "";
    datasetId_ = "";
    referenceSetId_ = "";
    name_ = "";
    filename_ = "";
    readGroups_ = java.util.Collections.emptyList();
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new ReadGroupSet();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  private ReadGroupSet(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    this();
    if (extensionRegistry == null) {
      throw new java.lang.NullPointerException();
    }
    int mutable_bitField0_ = 0;
    com.google.protobuf.UnknownFieldSet.Builder unknownFields =
        com.google.protobuf.UnknownFieldSet.newBuilder();
    try {
      boolean done = false;
      while (!done) {
        int tag = input.readTag();
        switch (tag) {
          case 0:
            done = true;
            break;
          case 10: {
            java.lang.String s = input.readStringRequireUtf8();

            id_ = s;
            break;
          }
          case 18: {
            java.lang.String s = input.readStringRequireUtf8();

            datasetId_ = s;
            break;
          }
          case 26: {
            java.lang.String s = input.readStringRequireUtf8();

            referenceSetId_ = s;
            break;
          }
          case 34: {
            java.lang.String s = input.readStringRequireUtf8();

            name_ = s;
            break;
          }
          case 42: {
            java.lang.String s = input.readStringRequireUtf8();

            filename_ = s;
            break;
          }
          case 50: {
            if (!((mutable_bitField0_ & 0x00000001) != 0)) {
              readGroups_ = new java.util.ArrayList<com.google.genomics.v1.ReadGroup>();
              mutable_bitField0_ |= 0x00000001;
            }
            readGroups_.add(
                input.readMessage(com.google.genomics.v1.ReadGroup.parser(), extensionRegistry));
            break;
          }
          case 58: {
            if (!((mutable_bitField0_ & 0x00000002) != 0)) {
              info_ = com.google.protobuf.MapField.newMapField(
                  InfoDefaultEntryHolder.defaultEntry);
              mutable_bitField0_ |= 0x00000002;
            }
            com.google.protobuf.MapEntry<java.lang.String, com.google.protobuf.ListValue>
            info__ = input.readMessage(
                InfoDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
            info_.getMutableMap().put(
                info__.getKey(), info__.getValue());
            break;
          }
          default: {
            if (!parseUnknownField(
                input, unknownFields, extensionRegistry, tag)) {
              done = true;
            }
            break;
          }
        }
      }
    } catch (com.google.protobuf.InvalidProtocolBufferException e) {
      throw e.setUnfinishedMessage(this);
    } catch (com.google.protobuf.UninitializedMessageException e) {
      throw e.asInvalidProtocolBufferException().setUnfinishedMessage(this);
    } catch (java.io.IOException e) {
      throw new com.google.protobuf.InvalidProtocolBufferException(
          e).setUnfinishedMessage(this);
    } finally {
      if (((mutable_bitField0_ & 0x00000001) != 0)) {
        readGroups_ = java.util.Collections.unmodifiableList(readGroups_);
      }
      this.unknownFields = unknownFields.build();
      makeExtensionsImmutable();
    }
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReadGroupSetProto.internal_static_google_genomics_v1_ReadGroupSet_descriptor;
  }

  @SuppressWarnings({"rawtypes"})
  @java.lang.Override
  protected com.google.protobuf.MapField internalGetMapField(
      int number) {
    switch (number) {
      case 7:
        return internalGetInfo();
      default:
        throw new RuntimeException(
            "Invalid map field number: " + number);
    }
  }
  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReadGroupSetProto.internal_static_google_genomics_v1_ReadGroupSet_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.ReadGroupSet.class, com.google.genomics.v1.ReadGroupSet.Builder.class);
  }

  public static final int ID_FIELD_NUMBER = 1;
  private volatile java.lang.Object id_;
  /**
   * <pre>
   * The server-generated read group set ID, unique for all read group sets.
   * </pre>
   *
   * <code>string id = 1;</code>
   * @return The id.
   */
  @java.lang.Override
  public java.lang.String getId() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      id_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The server-generated read group set ID, unique for all read group sets.
   * </pre>
   *
   * <code>string id = 1;</code>
   * @return The bytes for id.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getIdBytes() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      id_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int DATASET_ID_FIELD_NUMBER = 2;
  private volatile java.lang.Object datasetId_;
  /**
   * <pre>
   * The dataset to which this read group set belongs.
   * </pre>
   *
   * <code>string dataset_id = 2;</code>
   * @return The datasetId.
   */
  @java.lang.Override
  public java.lang.String getDatasetId() {
    java.lang.Object ref = datasetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      datasetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The dataset to which this read group set belongs.
   * </pre>
   *
   * <code>string dataset_id = 2;</code>
   * @return The bytes for datasetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getDatasetIdBytes() {
    java.lang.Object ref = datasetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      datasetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int REFERENCE_SET_ID_FIELD_NUMBER = 3;
  private volatile java.lang.Object referenceSetId_;
  /**
   * <pre>
   * The reference set to which the reads in this read group set are aligned.
   * </pre>
   *
   * <code>string reference_set_id = 3;</code>
   * @return The referenceSetId.
   */
  @java.lang.Override
  public java.lang.String getReferenceSetId() {
    java.lang.Object ref = referenceSetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      referenceSetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The reference set to which the reads in this read group set are aligned.
   * </pre>
   *
   * <code>string reference_set_id = 3;</code>
   * @return The bytes for referenceSetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getReferenceSetIdBytes() {
    java.lang.Object ref = referenceSetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      referenceSetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int NAME_FIELD_NUMBER = 4;
  private volatile java.lang.Object name_;
  /**
   * <pre>
   * The read group set name. By default this will be initialized to the sample
   * name of the sequenced data contained in this set.
   * </pre>
   *
   * <code>string name = 4;</code>
   * @return The name.
   */
  @java.lang.Override
  public java.lang.String getName() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      name_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The read group set name. By default this will be initialized to the sample
   * name of the sequenced data contained in this set.
   * </pre>
   *
   * <code>string name = 4;</code>
   * @return The bytes for name.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getNameBytes() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      name_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int FILENAME_FIELD_NUMBER = 5;
  private volatile java.lang.Object filename_;
  /**
   * <pre>
   * The filename of the original source file for this read group set, if any.
   * </pre>
   *
   * <code>string filename = 5;</code>
   * @return The filename.
   */
  @java.lang.Override
  public java.lang.String getFilename() {
    java.lang.Object ref = filename_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      filename_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The filename of the original source file for this read group set, if any.
   * </pre>
   *
   * <code>string filename = 5;</code>
   * @return The bytes for filename.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getFilenameBytes() {
    java.lang.Object ref = filename_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      filename_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int READ_GROUPS_FIELD_NUMBER = 6;
  private java.util.List<com.google.genomics.v1.ReadGroup> readGroups_;
  /**
   * <pre>
   * The read groups in this set. There are typically 1-10 read groups in a read
   * group set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
   */
  @java.lang.Override
  public java.util.List<com.google.genomics.v1.ReadGroup> getReadGroupsList() {
    return readGroups_;
  }
  /**
   * <pre>
   * The read groups in this set. There are typically 1-10 read groups in a read
   * group set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
   */
  @java.lang.Override
  public java.util.List<? extends com.google.genomics.v1.ReadGroupOrBuilder> 
      getReadGroupsOrBuilderList() {
    return readGroups_;
  }
  /**
   * <pre>
   * The read groups in this set. There are typically 1-10 read groups in a read
   * group set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
   */
  @java.lang.Override
  public int getReadGroupsCount() {
    return readGroups_.size();
  }
  /**
   * <pre>
   * The read groups in this set. There are typically 1-10 read groups in a read
   * group set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadGroup getReadGroups(int index) {
    return readGroups_.get(index);
  }
  /**
   * <pre>
   * The read groups in this set. There are typically 1-10 read groups in a read
   * group set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadGroupOrBuilder getReadGroupsOrBuilder(
      int index) {
    return readGroups_.get(index);
  }

  public static final int INFO_FIELD_NUMBER = 7;
  private static final class InfoDefaultEntryHolder {
    static final com.google.protobuf.MapEntry<
        java.lang.String, com.google.protobuf.ListValue> defaultEntry =
            com.google.protobuf.MapEntry
            .<java.lang.String, com.google.protobuf.ListValue>newDefaultInstance(
                com.google.genomics.v1.ReadGroupSetProto.internal_static_google_genomics_v1_ReadGroupSet_InfoEntry_descriptor, 
                com.google.protobuf.WireFormat.FieldType.STRING,
                "",
                com.google.protobuf.WireFormat.FieldType.MESSAGE,
                com.google.protobuf.ListValue.getDefaultInstance());
  }
  private com.google.protobuf.MapField<
      java.lang.String, com.google.protobuf.ListValue> info_;
  private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
  internalGetInfo() {
    if (info_ == null) {
      return com.google.protobuf.MapField.emptyMapField(
          InfoDefaultEntryHolder.defaultEntry);
    }
    return info_;
  }

  public int getInfoCount() {
    return internalGetInfo().getMap().size();
  }
  /**
   * <pre>
   * A map of additional read group set information.
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
   */

  @java.lang.Override
  public boolean containsInfo(
      java.lang.String key) {
    if (key == null) { throw new NullPointerException("map key"); }
    return internalGetInfo().getMap().containsKey(key);
  }
  /**
   * Use {@link #getInfoMap()} instead.
   */
  @java.lang.Override
  @java.lang.Deprecated
  public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfo() {
    return getInfoMap();
  }
  /**
   * <pre>
   * A map of additional read group set information.
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
   */
  @java.lang.Override

  public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfoMap() {
    return internalGetInfo().getMap();
  }
  /**
   * <pre>
   * A map of additional read group set information.
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
   */
  @java.lang.Override

  public com.google.protobuf.ListValue getInfoOrDefault(
      java.lang.String key,
      com.google.protobuf.ListValue defaultValue) {
    if (key == null) { throw new NullPointerException("map key"); }
    java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
        internalGetInfo().getMap();
    return map.containsKey(key) ? map.get(key) : defaultValue;
  }
  /**
   * <pre>
   * A map of additional read group set information.
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
   */
  @java.lang.Override

  public com.google.protobuf.ListValue getInfoOrThrow(
      java.lang.String key) {
    if (key == null) { throw new NullPointerException("map key"); }
    java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
        internalGetInfo().getMap();
    if (!map.containsKey(key)) {
      throw new java.lang.IllegalArgumentException();
    }
    return map.get(key);
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 1, id_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(datasetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 2, datasetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(referenceSetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 3, referenceSetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 4, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(filename_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 5, filename_);
    }
    for (int i = 0; i < readGroups_.size(); i++) {
      output.writeMessage(6, readGroups_.get(i));
    }
    com.google.protobuf.GeneratedMessageV3
      .serializeStringMapTo(
        output,
        internalGetInfo(),
        InfoDefaultEntryHolder.defaultEntry,
        7);
    unknownFields.writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, id_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(datasetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(2, datasetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(referenceSetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(3, referenceSetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(4, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(filename_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(5, filename_);
    }
    for (int i = 0; i < readGroups_.size(); i++) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(6, readGroups_.get(i));
    }
    for (java.util.Map.Entry<java.lang.String, com.google.protobuf.ListValue> entry
         : internalGetInfo().getMap().entrySet()) {
      com.google.protobuf.MapEntry<java.lang.String, com.google.protobuf.ListValue>
      info__ = InfoDefaultEntryHolder.defaultEntry.newBuilderForType()
          .setKey(entry.getKey())
          .setValue(entry.getValue())
          .build();
      size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(7, info__);
    }
    size += unknownFields.getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.ReadGroupSet)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.ReadGroupSet other = (com.google.genomics.v1.ReadGroupSet) obj;

    if (!getId()
        .equals(other.getId())) return false;
    if (!getDatasetId()
        .equals(other.getDatasetId())) return false;
    if (!getReferenceSetId()
        .equals(other.getReferenceSetId())) return false;
    if (!getName()
        .equals(other.getName())) return false;
    if (!getFilename()
        .equals(other.getFilename())) return false;
    if (!getReadGroupsList()
        .equals(other.getReadGroupsList())) return false;
    if (!internalGetInfo().equals(
        other.internalGetInfo())) return false;
    if (!unknownFields.equals(other.unknownFields)) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    hash = (37 * hash) + ID_FIELD_NUMBER;
    hash = (53 * hash) + getId().hashCode();
    hash = (37 * hash) + DATASET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getDatasetId().hashCode();
    hash = (37 * hash) + REFERENCE_SET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getReferenceSetId().hashCode();
    hash = (37 * hash) + NAME_FIELD_NUMBER;
    hash = (53 * hash) + getName().hashCode();
    hash = (37 * hash) + FILENAME_FIELD_NUMBER;
    hash = (53 * hash) + getFilename().hashCode();
    if (getReadGroupsCount() > 0) {
      hash = (37 * hash) + READ_GROUPS_FIELD_NUMBER;
      hash = (53 * hash) + getReadGroupsList().hashCode();
    }
    if (!internalGetInfo().getMap().isEmpty()) {
      hash = (37 * hash) + INFO_FIELD_NUMBER;
      hash = (53 * hash) + internalGetInfo().hashCode();
    }
    hash = (29 * hash) + unknownFields.hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroupSet parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.ReadGroupSet parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.ReadGroupSet parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.ReadGroupSet prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A read group set is a logical collection of read groups, which are
   * collections of reads produced by a sequencer. A read group set typically
   * models reads corresponding to one sample, sequenced one way, and aligned one
   * way.
   * * A read group set belongs to one dataset.
   * * A read group belongs to one read group set.
   * * A read belongs to one read group.
   * For more genomics resource definitions, see [Fundamentals of Google
   * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.ReadGroupSet}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.ReadGroupSet)
      com.google.genomics.v1.ReadGroupSetOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadGroupSetProto.internal_static_google_genomics_v1_ReadGroupSet_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 7:
          return internalGetInfo();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMutableMapField(
        int number) {
      switch (number) {
        case 7:
          return internalGetMutableInfo();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadGroupSetProto.internal_static_google_genomics_v1_ReadGroupSet_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.ReadGroupSet.class, com.google.genomics.v1.ReadGroupSet.Builder.class);
    }

    // Construct using com.google.genomics.v1.ReadGroupSet.newBuilder()
    private Builder() {
      maybeForceBuilderInitialization();
    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);
      maybeForceBuilderInitialization();
    }
    private void maybeForceBuilderInitialization() {
      if (com.google.protobuf.GeneratedMessageV3
              .alwaysUseFieldBuilders) {
        getReadGroupsFieldBuilder();
      }
    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      id_ = "";

      datasetId_ = "";

      referenceSetId_ = "";

      name_ = "";

      filename_ = "";

      if (readGroupsBuilder_ == null) {
        readGroups_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
      } else {
        readGroupsBuilder_.clear();
      }
      internalGetMutableInfo().clear();
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReadGroupSetProto.internal_static_google_genomics_v1_ReadGroupSet_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroupSet getDefaultInstanceForType() {
      return com.google.genomics.v1.ReadGroupSet.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroupSet build() {
      com.google.genomics.v1.ReadGroupSet result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.ReadGroupSet buildPartial() {
      com.google.genomics.v1.ReadGroupSet result = new com.google.genomics.v1.ReadGroupSet(this);
      int from_bitField0_ = bitField0_;
      result.id_ = id_;
      result.datasetId_ = datasetId_;
      result.referenceSetId_ = referenceSetId_;
      result.name_ = name_;
      result.filename_ = filename_;
      if (readGroupsBuilder_ == null) {
        if (((bitField0_ & 0x00000001) != 0)) {
          readGroups_ = java.util.Collections.unmodifiableList(readGroups_);
          bitField0_ = (bitField0_ & ~0x00000001);
        }
        result.readGroups_ = readGroups_;
      } else {
        result.readGroups_ = readGroupsBuilder_.build();
      }
      result.info_ = internalGetInfo();
      result.info_.makeImmutable();
      onBuilt();
      return result;
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.ReadGroupSet) {
        return mergeFrom((com.google.genomics.v1.ReadGroupSet)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.ReadGroupSet other) {
      if (other == com.google.genomics.v1.ReadGroupSet.getDefaultInstance()) return this;
      if (!other.getId().isEmpty()) {
        id_ = other.id_;
        onChanged();
      }
      if (!other.getDatasetId().isEmpty()) {
        datasetId_ = other.datasetId_;
        onChanged();
      }
      if (!other.getReferenceSetId().isEmpty()) {
        referenceSetId_ = other.referenceSetId_;
        onChanged();
      }
      if (!other.getName().isEmpty()) {
        name_ = other.name_;
        onChanged();
      }
      if (!other.getFilename().isEmpty()) {
        filename_ = other.filename_;
        onChanged();
      }
      if (readGroupsBuilder_ == null) {
        if (!other.readGroups_.isEmpty()) {
          if (readGroups_.isEmpty()) {
            readGroups_ = other.readGroups_;
            bitField0_ = (bitField0_ & ~0x00000001);
          } else {
            ensureReadGroupsIsMutable();
            readGroups_.addAll(other.readGroups_);
          }
          onChanged();
        }
      } else {
        if (!other.readGroups_.isEmpty()) {
          if (readGroupsBuilder_.isEmpty()) {
            readGroupsBuilder_.dispose();
            readGroupsBuilder_ = null;
            readGroups_ = other.readGroups_;
            bitField0_ = (bitField0_ & ~0x00000001);
            readGroupsBuilder_ = 
              com.google.protobuf.GeneratedMessageV3.alwaysUseFieldBuilders ?
                 getReadGroupsFieldBuilder() : null;
          } else {
            readGroupsBuilder_.addAllMessages(other.readGroups_);
          }
        }
      }
      internalGetMutableInfo().mergeFrom(
          other.internalGetInfo());
      this.mergeUnknownFields(other.unknownFields);
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      com.google.genomics.v1.ReadGroupSet parsedMessage = null;
      try {
        parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        parsedMessage = (com.google.genomics.v1.ReadGroupSet) e.getUnfinishedMessage();
        throw e.unwrapIOException();
      } finally {
        if (parsedMessage != null) {
          mergeFrom(parsedMessage);
        }
      }
      return this;
    }
    private int bitField0_;

    private java.lang.Object id_ = "";
    /**
     * <pre>
     * The server-generated read group set ID, unique for all read group sets.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return The id.
     */
    public java.lang.String getId() {
      java.lang.Object ref = id_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        id_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The server-generated read group set ID, unique for all read group sets.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return The bytes for id.
     */
    public com.google.protobuf.ByteString
        getIdBytes() {
      java.lang.Object ref = id_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        id_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The server-generated read group set ID, unique for all read group sets.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @param value The id to set.
     * @return This builder for chaining.
     */
    public Builder setId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      id_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated read group set ID, unique for all read group sets.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return This builder for chaining.
     */
    public Builder clearId() {
      
      id_ = getDefaultInstance().getId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated read group set ID, unique for all read group sets.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @param value The bytes for id to set.
     * @return This builder for chaining.
     */
    public Builder setIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      id_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object datasetId_ = "";
    /**
     * <pre>
     * The dataset to which this read group set belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @return The datasetId.
     */
    public java.lang.String getDatasetId() {
      java.lang.Object ref = datasetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        datasetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The dataset to which this read group set belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @return The bytes for datasetId.
     */
    public com.google.protobuf.ByteString
        getDatasetIdBytes() {
      java.lang.Object ref = datasetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        datasetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The dataset to which this read group set belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @param value The datasetId to set.
     * @return This builder for chaining.
     */
    public Builder setDatasetId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      datasetId_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The dataset to which this read group set belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearDatasetId() {
      
      datasetId_ = getDefaultInstance().getDatasetId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The dataset to which this read group set belongs.
     * </pre>
     *
     * <code>string dataset_id = 2;</code>
     * @param value The bytes for datasetId to set.
     * @return This builder for chaining.
     */
    public Builder setDatasetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      datasetId_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object referenceSetId_ = "";
    /**
     * <pre>
     * The reference set to which the reads in this read group set are aligned.
     * </pre>
     *
     * <code>string reference_set_id = 3;</code>
     * @return The referenceSetId.
     */
    public java.lang.String getReferenceSetId() {
      java.lang.Object ref = referenceSetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        referenceSetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The reference set to which the reads in this read group set are aligned.
     * </pre>
     *
     * <code>string reference_set_id = 3;</code>
     * @return The bytes for referenceSetId.
     */
    public com.google.protobuf.ByteString
        getReferenceSetIdBytes() {
      java.lang.Object ref = referenceSetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        referenceSetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The reference set to which the reads in this read group set are aligned.
     * </pre>
     *
     * <code>string reference_set_id = 3;</code>
     * @param value The referenceSetId to set.
     * @return This builder for chaining.
     */
    public Builder setReferenceSetId(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      referenceSetId_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The reference set to which the reads in this read group set are aligned.
     * </pre>
     *
     * <code>string reference_set_id = 3;</code>
     * @return This builder for chaining.
     */
    public Builder clearReferenceSetId() {
      
      referenceSetId_ = getDefaultInstance().getReferenceSetId();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The reference set to which the reads in this read group set are aligned.
     * </pre>
     *
     * <code>string reference_set_id = 3;</code>
     * @param value The bytes for referenceSetId to set.
     * @return This builder for chaining.
     */
    public Builder setReferenceSetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      referenceSetId_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object name_ = "";
    /**
     * <pre>
     * The read group set name. By default this will be initialized to the sample
     * name of the sequenced data contained in this set.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @return The name.
     */
    public java.lang.String getName() {
      java.lang.Object ref = name_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        name_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The read group set name. By default this will be initialized to the sample
     * name of the sequenced data contained in this set.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @return The bytes for name.
     */
    public com.google.protobuf.ByteString
        getNameBytes() {
      java.lang.Object ref = name_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        name_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The read group set name. By default this will be initialized to the sample
     * name of the sequenced data contained in this set.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @param value The name to set.
     * @return This builder for chaining.
     */
    public Builder setName(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      name_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The read group set name. By default this will be initialized to the sample
     * name of the sequenced data contained in this set.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @return This builder for chaining.
     */
    public Builder clearName() {
      
      name_ = getDefaultInstance().getName();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The read group set name. By default this will be initialized to the sample
     * name of the sequenced data contained in this set.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @param value The bytes for name to set.
     * @return This builder for chaining.
     */
    public Builder setNameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      name_ = value;
      onChanged();
      return this;
    }

    private java.lang.Object filename_ = "";
    /**
     * <pre>
     * The filename of the original source file for this read group set, if any.
     * </pre>
     *
     * <code>string filename = 5;</code>
     * @return The filename.
     */
    public java.lang.String getFilename() {
      java.lang.Object ref = filename_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        filename_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The filename of the original source file for this read group set, if any.
     * </pre>
     *
     * <code>string filename = 5;</code>
     * @return The bytes for filename.
     */
    public com.google.protobuf.ByteString
        getFilenameBytes() {
      java.lang.Object ref = filename_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        filename_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The filename of the original source file for this read group set, if any.
     * </pre>
     *
     * <code>string filename = 5;</code>
     * @param value The filename to set.
     * @return This builder for chaining.
     */
    public Builder setFilename(
        java.lang.String value) {
      if (value == null) {
    throw new NullPointerException();
  }
  
      filename_ = value;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The filename of the original source file for this read group set, if any.
     * </pre>
     *
     * <code>string filename = 5;</code>
     * @return This builder for chaining.
     */
    public Builder clearFilename() {
      
      filename_ = getDefaultInstance().getFilename();
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The filename of the original source file for this read group set, if any.
     * </pre>
     *
     * <code>string filename = 5;</code>
     * @param value The bytes for filename to set.
     * @return This builder for chaining.
     */
    public Builder setFilenameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
      
      filename_ = value;
      onChanged();
      return this;
    }

    private java.util.List<com.google.genomics.v1.ReadGroup> readGroups_ =
      java.util.Collections.emptyList();
    private void ensureReadGroupsIsMutable() {
      if (!((bitField0_ & 0x00000001) != 0)) {
        readGroups_ = new java.util.ArrayList<com.google.genomics.v1.ReadGroup>(readGroups_);
        bitField0_ |= 0x00000001;
       }
    }

    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.ReadGroup, com.google.genomics.v1.ReadGroup.Builder, com.google.genomics.v1.ReadGroupOrBuilder> readGroupsBuilder_;

    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public java.util.List<com.google.genomics.v1.ReadGroup> getReadGroupsList() {
      if (readGroupsBuilder_ == null) {
        return java.util.Collections.unmodifiableList(readGroups_);
      } else {
        return readGroupsBuilder_.getMessageList();
      }
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public int getReadGroupsCount() {
      if (readGroupsBuilder_ == null) {
        return readGroups_.size();
      } else {
        return readGroupsBuilder_.getCount();
      }
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public com.google.genomics.v1.ReadGroup getReadGroups(int index) {
      if (readGroupsBuilder_ == null) {
        return readGroups_.get(index);
      } else {
        return readGroupsBuilder_.getMessage(index);
      }
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder setReadGroups(
        int index, com.google.genomics.v1.ReadGroup value) {
      if (readGroupsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureReadGroupsIsMutable();
        readGroups_.set(index, value);
        onChanged();
      } else {
        readGroupsBuilder_.setMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder setReadGroups(
        int index, com.google.genomics.v1.ReadGroup.Builder builderForValue) {
      if (readGroupsBuilder_ == null) {
        ensureReadGroupsIsMutable();
        readGroups_.set(index, builderForValue.build());
        onChanged();
      } else {
        readGroupsBuilder_.setMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder addReadGroups(com.google.genomics.v1.ReadGroup value) {
      if (readGroupsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureReadGroupsIsMutable();
        readGroups_.add(value);
        onChanged();
      } else {
        readGroupsBuilder_.addMessage(value);
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder addReadGroups(
        int index, com.google.genomics.v1.ReadGroup value) {
      if (readGroupsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureReadGroupsIsMutable();
        readGroups_.add(index, value);
        onChanged();
      } else {
        readGroupsBuilder_.addMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder addReadGroups(
        com.google.genomics.v1.ReadGroup.Builder builderForValue) {
      if (readGroupsBuilder_ == null) {
        ensureReadGroupsIsMutable();
        readGroups_.add(builderForValue.build());
        onChanged();
      } else {
        readGroupsBuilder_.addMessage(builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder addReadGroups(
        int index, com.google.genomics.v1.ReadGroup.Builder builderForValue) {
      if (readGroupsBuilder_ == null) {
        ensureReadGroupsIsMutable();
        readGroups_.add(index, builderForValue.build());
        onChanged();
      } else {
        readGroupsBuilder_.addMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder addAllReadGroups(
        java.lang.Iterable<? extends com.google.genomics.v1.ReadGroup> values) {
      if (readGroupsBuilder_ == null) {
        ensureReadGroupsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, readGroups_);
        onChanged();
      } else {
        readGroupsBuilder_.addAllMessages(values);
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder clearReadGroups() {
      if (readGroupsBuilder_ == null) {
        readGroups_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
        onChanged();
      } else {
        readGroupsBuilder_.clear();
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public Builder removeReadGroups(int index) {
      if (readGroupsBuilder_ == null) {
        ensureReadGroupsIsMutable();
        readGroups_.remove(index);
        onChanged();
      } else {
        readGroupsBuilder_.remove(index);
      }
      return this;
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public com.google.genomics.v1.ReadGroup.Builder getReadGroupsBuilder(
        int index) {
      return getReadGroupsFieldBuilder().getBuilder(index);
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public com.google.genomics.v1.ReadGroupOrBuilder getReadGroupsOrBuilder(
        int index) {
      if (readGroupsBuilder_ == null) {
        return readGroups_.get(index);  } else {
        return readGroupsBuilder_.getMessageOrBuilder(index);
      }
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public java.util.List<? extends com.google.genomics.v1.ReadGroupOrBuilder> 
         getReadGroupsOrBuilderList() {
      if (readGroupsBuilder_ != null) {
        return readGroupsBuilder_.getMessageOrBuilderList();
      } else {
        return java.util.Collections.unmodifiableList(readGroups_);
      }
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public com.google.genomics.v1.ReadGroup.Builder addReadGroupsBuilder() {
      return getReadGroupsFieldBuilder().addBuilder(
          com.google.genomics.v1.ReadGroup.getDefaultInstance());
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public com.google.genomics.v1.ReadGroup.Builder addReadGroupsBuilder(
        int index) {
      return getReadGroupsFieldBuilder().addBuilder(
          index, com.google.genomics.v1.ReadGroup.getDefaultInstance());
    }
    /**
     * <pre>
     * The read groups in this set. There are typically 1-10 read groups in a read
     * group set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReadGroup read_groups = 6;</code>
     */
    public java.util.List<com.google.genomics.v1.ReadGroup.Builder> 
         getReadGroupsBuilderList() {
      return getReadGroupsFieldBuilder().getBuilderList();
    }
    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.ReadGroup, com.google.genomics.v1.ReadGroup.Builder, com.google.genomics.v1.ReadGroupOrBuilder> 
        getReadGroupsFieldBuilder() {
      if (readGroupsBuilder_ == null) {
        readGroupsBuilder_ = new com.google.protobuf.RepeatedFieldBuilderV3<
            com.google.genomics.v1.ReadGroup, com.google.genomics.v1.ReadGroup.Builder, com.google.genomics.v1.ReadGroupOrBuilder>(
                readGroups_,
                ((bitField0_ & 0x00000001) != 0),
                getParentForChildren(),
                isClean());
        readGroups_ = null;
      }
      return readGroupsBuilder_;
    }

    private com.google.protobuf.MapField<
        java.lang.String, com.google.protobuf.ListValue> info_;
    private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
    internalGetInfo() {
      if (info_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            InfoDefaultEntryHolder.defaultEntry);
      }
      return info_;
    }
    private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
    internalGetMutableInfo() {
      onChanged();;
      if (info_ == null) {
        info_ = com.google.protobuf.MapField.newMapField(
            InfoDefaultEntryHolder.defaultEntry);
      }
      if (!info_.isMutable()) {
        info_ = info_.copy();
      }
      return info_;
    }

    public int getInfoCount() {
      return internalGetInfo().getMap().size();
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */

    @java.lang.Override
    public boolean containsInfo(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      return internalGetInfo().getMap().containsKey(key);
    }
    /**
     * Use {@link #getInfoMap()} instead.
     */
    @java.lang.Override
    @java.lang.Deprecated
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfo() {
      return getInfoMap();
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */
    @java.lang.Override

    public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfoMap() {
      return internalGetInfo().getMap();
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */
    @java.lang.Override

    public com.google.protobuf.ListValue getInfoOrDefault(
        java.lang.String key,
        com.google.protobuf.ListValue defaultValue) {
      if (key == null) { throw new NullPointerException("map key"); }
      java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
          internalGetInfo().getMap();
      return map.containsKey(key) ? map.get(key) : defaultValue;
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */
    @java.lang.Override

    public com.google.protobuf.ListValue getInfoOrThrow(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
          internalGetInfo().getMap();
      if (!map.containsKey(key)) {
        throw new java.lang.IllegalArgumentException();
      }
      return map.get(key);
    }

    public Builder clearInfo() {
      internalGetMutableInfo().getMutableMap()
          .clear();
      return this;
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */

    public Builder removeInfo(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      internalGetMutableInfo().getMutableMap()
          .remove(key);
      return this;
    }
    /**
     * Use alternate mutation accessors instead.
     */
    @java.lang.Deprecated
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue>
    getMutableInfo() {
      return internalGetMutableInfo().getMutableMap();
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */
    public Builder putInfo(
        java.lang.String key,
        com.google.protobuf.ListValue value) {
      if (key == null) { throw new NullPointerException("map key"); }
      if (value == null) {
  throw new NullPointerException("map value");
}

      internalGetMutableInfo().getMutableMap()
          .put(key, value);
      return this;
    }
    /**
     * <pre>
     * A map of additional read group set information.
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 7;</code>
     */

    public Builder putAllInfo(
        java.util.Map<java.lang.String, com.google.protobuf.ListValue> values) {
      internalGetMutableInfo().getMutableMap()
          .putAll(values);
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.ReadGroupSet)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.ReadGroupSet)
  private static final com.google.genomics.v1.ReadGroupSet DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.ReadGroupSet();
  }

  public static com.google.genomics.v1.ReadGroupSet getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<ReadGroupSet>
      PARSER = new com.google.protobuf.AbstractParser<ReadGroupSet>() {
    @java.lang.Override
    public ReadGroupSet parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return new ReadGroupSet(input, extensionRegistry);
    }
  };

  public static com.google.protobuf.Parser<ReadGroupSet> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<ReadGroupSet> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.ReadGroupSet getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

