// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/references.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A reference is a canonical assembled DNA sequence, intended to act as a
 * reference coordinate space for other genomic annotations. A single reference
 * might represent the human chromosome 1 or mitochandrial DNA, for instance. A
 * reference belongs to one or more reference sets.
 * For more genomics resource definitions, see [Fundamentals of Google
 * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.Reference}
 */
public final class Reference extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.Reference)
    ReferenceOrBuilder {
private static final long serialVersionUID = 0L;
  // Use Reference.newBuilder() to construct.
  private Reference(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private Reference() {
    id_ = "";
    md5Checksum_ = "";
    name_ = "";
    sourceUri_ = "";
    sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new Reference();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_descriptor;
  }

  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.Reference.class, com.google.genomics.v1.Reference.Builder.class);
  }

  public static final int ID_FIELD_NUMBER = 1;
  @SuppressWarnings("serial")
  private volatile java.lang.Object id_ = "";
  /**
   * <pre>
   * The server-generated reference ID, unique across all references.
   * </pre>
   *
   * <code>string id = 1;</code>
   * @return The id.
   */
  @java.lang.Override
  public java.lang.String getId() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      id_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The server-generated reference ID, unique across all references.
   * </pre>
   *
   * <code>string id = 1;</code>
   * @return The bytes for id.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getIdBytes() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      id_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int LENGTH_FIELD_NUMBER = 2;
  private long length_ = 0L;
  /**
   * <pre>
   * The length of this reference's sequence.
   * </pre>
   *
   * <code>int64 length = 2;</code>
   * @return The length.
   */
  @java.lang.Override
  public long getLength() {
    return length_;
  }

  public static final int MD5CHECKSUM_FIELD_NUMBER = 3;
  @SuppressWarnings("serial")
  private volatile java.lang.Object md5Checksum_ = "";
  /**
   * <pre>
   * MD5 of the upper-case sequence excluding all whitespace characters (this
   * is equivalent to SQ:M5 in SAM). This value is represented in lower case
   * hexadecimal format.
   * </pre>
   *
   * <code>string md5checksum = 3;</code>
   * @return The md5checksum.
   */
  @java.lang.Override
  public java.lang.String getMd5Checksum() {
    java.lang.Object ref = md5Checksum_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      md5Checksum_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * MD5 of the upper-case sequence excluding all whitespace characters (this
   * is equivalent to SQ:M5 in SAM). This value is represented in lower case
   * hexadecimal format.
   * </pre>
   *
   * <code>string md5checksum = 3;</code>
   * @return The bytes for md5checksum.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getMd5ChecksumBytes() {
    java.lang.Object ref = md5Checksum_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      md5Checksum_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int NAME_FIELD_NUMBER = 4;
  @SuppressWarnings("serial")
  private volatile java.lang.Object name_ = "";
  /**
   * <pre>
   * The name of this reference, for example `22`.
   * </pre>
   *
   * <code>string name = 4;</code>
   * @return The name.
   */
  @java.lang.Override
  public java.lang.String getName() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      name_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The name of this reference, for example `22`.
   * </pre>
   *
   * <code>string name = 4;</code>
   * @return The bytes for name.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getNameBytes() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      name_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int SOURCE_URI_FIELD_NUMBER = 5;
  @SuppressWarnings("serial")
  private volatile java.lang.Object sourceUri_ = "";
  /**
   * <pre>
   * The URI from which the sequence was obtained. Typically specifies a FASTA
   * format file.
   * </pre>
   *
   * <code>string source_uri = 5;</code>
   * @return The sourceUri.
   */
  @java.lang.Override
  public java.lang.String getSourceUri() {
    java.lang.Object ref = sourceUri_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      sourceUri_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The URI from which the sequence was obtained. Typically specifies a FASTA
   * format file.
   * </pre>
   *
   * <code>string source_uri = 5;</code>
   * @return The bytes for sourceUri.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getSourceUriBytes() {
    java.lang.Object ref = sourceUri_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      sourceUri_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int SOURCE_ACCESSIONS_FIELD_NUMBER = 6;
  @SuppressWarnings("serial")
  private com.google.protobuf.LazyStringList sourceAccessions_;
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   * @return A list containing the sourceAccessions.
   */
  public com.google.protobuf.ProtocolStringList
      getSourceAccessionsList() {
    return sourceAccessions_;
  }
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   * @return The count of sourceAccessions.
   */
  public int getSourceAccessionsCount() {
    return sourceAccessions_.size();
  }
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   * @param index The index of the element to return.
   * @return The sourceAccessions at the given index.
   */
  public java.lang.String getSourceAccessions(int index) {
    return sourceAccessions_.get(index);
  }
  /**
   * <pre>
   * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
   * with a version number, for example `GCF_000001405.26`.
   * </pre>
   *
   * <code>repeated string source_accessions = 6;</code>
   * @param index The index of the value to return.
   * @return The bytes of the sourceAccessions at the given index.
   */
  public com.google.protobuf.ByteString
      getSourceAccessionsBytes(int index) {
    return sourceAccessions_.getByteString(index);
  }

  public static final int NCBI_TAXON_ID_FIELD_NUMBER = 7;
  private int ncbiTaxonId_ = 0;
  /**
   * <pre>
   * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
   * </pre>
   *
   * <code>int32 ncbi_taxon_id = 7;</code>
   * @return The ncbiTaxonId.
   */
  @java.lang.Override
  public int getNcbiTaxonId() {
    return ncbiTaxonId_;
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 1, id_);
    }
    if (length_ != 0L) {
      output.writeInt64(2, length_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(md5Checksum_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 3, md5Checksum_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 4, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(sourceUri_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 5, sourceUri_);
    }
    for (int i = 0; i < sourceAccessions_.size(); i++) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 6, sourceAccessions_.getRaw(i));
    }
    if (ncbiTaxonId_ != 0) {
      output.writeInt32(7, ncbiTaxonId_);
    }
    getUnknownFields().writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, id_);
    }
    if (length_ != 0L) {
      size += com.google.protobuf.CodedOutputStream
        .computeInt64Size(2, length_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(md5Checksum_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(3, md5Checksum_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(4, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(sourceUri_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(5, sourceUri_);
    }
    {
      int dataSize = 0;
      for (int i = 0; i < sourceAccessions_.size(); i++) {
        dataSize += computeStringSizeNoTag(sourceAccessions_.getRaw(i));
      }
      size += dataSize;
      size += 1 * getSourceAccessionsList().size();
    }
    if (ncbiTaxonId_ != 0) {
      size += com.google.protobuf.CodedOutputStream
        .computeInt32Size(7, ncbiTaxonId_);
    }
    size += getUnknownFields().getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.Reference)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.Reference other = (com.google.genomics.v1.Reference) obj;

    if (!getId()
        .equals(other.getId())) return false;
    if (getLength()
        != other.getLength()) return false;
    if (!getMd5Checksum()
        .equals(other.getMd5Checksum())) return false;
    if (!getName()
        .equals(other.getName())) return false;
    if (!getSourceUri()
        .equals(other.getSourceUri())) return false;
    if (!getSourceAccessionsList()
        .equals(other.getSourceAccessionsList())) return false;
    if (getNcbiTaxonId()
        != other.getNcbiTaxonId()) return false;
    if (!getUnknownFields().equals(other.getUnknownFields())) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    hash = (37 * hash) + ID_FIELD_NUMBER;
    hash = (53 * hash) + getId().hashCode();
    hash = (37 * hash) + LENGTH_FIELD_NUMBER;
    hash = (53 * hash) + com.google.protobuf.Internal.hashLong(
        getLength());
    hash = (37 * hash) + MD5CHECKSUM_FIELD_NUMBER;
    hash = (53 * hash) + getMd5Checksum().hashCode();
    hash = (37 * hash) + NAME_FIELD_NUMBER;
    hash = (53 * hash) + getName().hashCode();
    hash = (37 * hash) + SOURCE_URI_FIELD_NUMBER;
    hash = (53 * hash) + getSourceUri().hashCode();
    if (getSourceAccessionsCount() > 0) {
      hash = (37 * hash) + SOURCE_ACCESSIONS_FIELD_NUMBER;
      hash = (53 * hash) + getSourceAccessionsList().hashCode();
    }
    hash = (37 * hash) + NCBI_TAXON_ID_FIELD_NUMBER;
    hash = (53 * hash) + getNcbiTaxonId();
    hash = (29 * hash) + getUnknownFields().hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.Reference parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.Reference parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.Reference parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.Reference prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A reference is a canonical assembled DNA sequence, intended to act as a
   * reference coordinate space for other genomic annotations. A single reference
   * might represent the human chromosome 1 or mitochandrial DNA, for instance. A
   * reference belongs to one or more reference sets.
   * For more genomics resource definitions, see [Fundamentals of Google
   * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.Reference}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.Reference)
      com.google.genomics.v1.ReferenceOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.Reference.class, com.google.genomics.v1.Reference.Builder.class);
    }

    // Construct using com.google.genomics.v1.Reference.newBuilder()
    private Builder() {

    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);

    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      bitField0_ = 0;
      id_ = "";
      length_ = 0L;
      md5Checksum_ = "";
      name_ = "";
      sourceUri_ = "";
      sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      bitField0_ = (bitField0_ & ~0x00000020);
      ncbiTaxonId_ = 0;
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReferencesProto.internal_static_google_genomics_v1_Reference_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.Reference getDefaultInstanceForType() {
      return com.google.genomics.v1.Reference.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.Reference build() {
      com.google.genomics.v1.Reference result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.Reference buildPartial() {
      com.google.genomics.v1.Reference result = new com.google.genomics.v1.Reference(this);
      buildPartialRepeatedFields(result);
      if (bitField0_ != 0) { buildPartial0(result); }
      onBuilt();
      return result;
    }

    private void buildPartialRepeatedFields(com.google.genomics.v1.Reference result) {
      if (((bitField0_ & 0x00000020) != 0)) {
        sourceAccessions_ = sourceAccessions_.getUnmodifiableView();
        bitField0_ = (bitField0_ & ~0x00000020);
      }
      result.sourceAccessions_ = sourceAccessions_;
    }

    private void buildPartial0(com.google.genomics.v1.Reference result) {
      int from_bitField0_ = bitField0_;
      if (((from_bitField0_ & 0x00000001) != 0)) {
        result.id_ = id_;
      }
      if (((from_bitField0_ & 0x00000002) != 0)) {
        result.length_ = length_;
      }
      if (((from_bitField0_ & 0x00000004) != 0)) {
        result.md5Checksum_ = md5Checksum_;
      }
      if (((from_bitField0_ & 0x00000008) != 0)) {
        result.name_ = name_;
      }
      if (((from_bitField0_ & 0x00000010) != 0)) {
        result.sourceUri_ = sourceUri_;
      }
      if (((from_bitField0_ & 0x00000040) != 0)) {
        result.ncbiTaxonId_ = ncbiTaxonId_;
      }
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.Reference) {
        return mergeFrom((com.google.genomics.v1.Reference)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.Reference other) {
      if (other == com.google.genomics.v1.Reference.getDefaultInstance()) return this;
      if (!other.getId().isEmpty()) {
        id_ = other.id_;
        bitField0_ |= 0x00000001;
        onChanged();
      }
      if (other.getLength() != 0L) {
        setLength(other.getLength());
      }
      if (!other.getMd5Checksum().isEmpty()) {
        md5Checksum_ = other.md5Checksum_;
        bitField0_ |= 0x00000004;
        onChanged();
      }
      if (!other.getName().isEmpty()) {
        name_ = other.name_;
        bitField0_ |= 0x00000008;
        onChanged();
      }
      if (!other.getSourceUri().isEmpty()) {
        sourceUri_ = other.sourceUri_;
        bitField0_ |= 0x00000010;
        onChanged();
      }
      if (!other.sourceAccessions_.isEmpty()) {
        if (sourceAccessions_.isEmpty()) {
          sourceAccessions_ = other.sourceAccessions_;
          bitField0_ = (bitField0_ & ~0x00000020);
        } else {
          ensureSourceAccessionsIsMutable();
          sourceAccessions_.addAll(other.sourceAccessions_);
        }
        onChanged();
      }
      if (other.getNcbiTaxonId() != 0) {
        setNcbiTaxonId(other.getNcbiTaxonId());
      }
      this.mergeUnknownFields(other.getUnknownFields());
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 10: {
              id_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000001;
              break;
            } // case 10
            case 16: {
              length_ = input.readInt64();
              bitField0_ |= 0x00000002;
              break;
            } // case 16
            case 26: {
              md5Checksum_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000004;
              break;
            } // case 26
            case 34: {
              name_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000008;
              break;
            } // case 34
            case 42: {
              sourceUri_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000010;
              break;
            } // case 42
            case 50: {
              java.lang.String s = input.readStringRequireUtf8();
              ensureSourceAccessionsIsMutable();
              sourceAccessions_.add(s);
              break;
            } // case 50
            case 56: {
              ncbiTaxonId_ = input.readInt32();
              bitField0_ |= 0x00000040;
              break;
            } // case 56
            default: {
              if (!super.parseUnknownField(input, extensionRegistry, tag)) {
                done = true; // was an endgroup tag
              }
              break;
            } // default:
          } // switch (tag)
        } // while (!done)
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.unwrapIOException();
      } finally {
        onChanged();
      } // finally
      return this;
    }
    private int bitField0_;

    private java.lang.Object id_ = "";
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return The id.
     */
    public java.lang.String getId() {
      java.lang.Object ref = id_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        id_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return The bytes for id.
     */
    public com.google.protobuf.ByteString
        getIdBytes() {
      java.lang.Object ref = id_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        id_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @param value The id to set.
     * @return This builder for chaining.
     */
    public Builder setId(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      id_ = value;
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @return This builder for chaining.
     */
    public Builder clearId() {
      id_ = getDefaultInstance().getId();
      bitField0_ = (bitField0_ & ~0x00000001);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated reference ID, unique across all references.
     * </pre>
     *
     * <code>string id = 1;</code>
     * @param value The bytes for id to set.
     * @return This builder for chaining.
     */
    public Builder setIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      id_ = value;
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }

    private long length_ ;
    /**
     * <pre>
     * The length of this reference's sequence.
     * </pre>
     *
     * <code>int64 length = 2;</code>
     * @return The length.
     */
    @java.lang.Override
    public long getLength() {
      return length_;
    }
    /**
     * <pre>
     * The length of this reference's sequence.
     * </pre>
     *
     * <code>int64 length = 2;</code>
     * @param value The length to set.
     * @return This builder for chaining.
     */
    public Builder setLength(long value) {
      
      length_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The length of this reference's sequence.
     * </pre>
     *
     * <code>int64 length = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearLength() {
      bitField0_ = (bitField0_ & ~0x00000002);
      length_ = 0L;
      onChanged();
      return this;
    }

    private java.lang.Object md5Checksum_ = "";
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     * @return The md5checksum.
     */
    public java.lang.String getMd5Checksum() {
      java.lang.Object ref = md5Checksum_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        md5Checksum_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     * @return The bytes for md5checksum.
     */
    public com.google.protobuf.ByteString
        getMd5ChecksumBytes() {
      java.lang.Object ref = md5Checksum_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        md5Checksum_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     * @param value The md5checksum to set.
     * @return This builder for chaining.
     */
    public Builder setMd5Checksum(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      md5Checksum_ = value;
      bitField0_ |= 0x00000004;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     * @return This builder for chaining.
     */
    public Builder clearMd5Checksum() {
      md5Checksum_ = getDefaultInstance().getMd5Checksum();
      bitField0_ = (bitField0_ & ~0x00000004);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * MD5 of the upper-case sequence excluding all whitespace characters (this
     * is equivalent to SQ:M5 in SAM). This value is represented in lower case
     * hexadecimal format.
     * </pre>
     *
     * <code>string md5checksum = 3;</code>
     * @param value The bytes for md5checksum to set.
     * @return This builder for chaining.
     */
    public Builder setMd5ChecksumBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      md5Checksum_ = value;
      bitField0_ |= 0x00000004;
      onChanged();
      return this;
    }

    private java.lang.Object name_ = "";
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @return The name.
     */
    public java.lang.String getName() {
      java.lang.Object ref = name_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        name_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @return The bytes for name.
     */
    public com.google.protobuf.ByteString
        getNameBytes() {
      java.lang.Object ref = name_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        name_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @param value The name to set.
     * @return This builder for chaining.
     */
    public Builder setName(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      name_ = value;
      bitField0_ |= 0x00000008;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @return This builder for chaining.
     */
    public Builder clearName() {
      name_ = getDefaultInstance().getName();
      bitField0_ = (bitField0_ & ~0x00000008);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The name of this reference, for example `22`.
     * </pre>
     *
     * <code>string name = 4;</code>
     * @param value The bytes for name to set.
     * @return This builder for chaining.
     */
    public Builder setNameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      name_ = value;
      bitField0_ |= 0x00000008;
      onChanged();
      return this;
    }

    private java.lang.Object sourceUri_ = "";
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     * @return The sourceUri.
     */
    public java.lang.String getSourceUri() {
      java.lang.Object ref = sourceUri_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        sourceUri_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     * @return The bytes for sourceUri.
     */
    public com.google.protobuf.ByteString
        getSourceUriBytes() {
      java.lang.Object ref = sourceUri_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        sourceUri_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     * @param value The sourceUri to set.
     * @return This builder for chaining.
     */
    public Builder setSourceUri(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      sourceUri_ = value;
      bitField0_ |= 0x00000010;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     * @return This builder for chaining.
     */
    public Builder clearSourceUri() {
      sourceUri_ = getDefaultInstance().getSourceUri();
      bitField0_ = (bitField0_ & ~0x00000010);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The URI from which the sequence was obtained. Typically specifies a FASTA
     * format file.
     * </pre>
     *
     * <code>string source_uri = 5;</code>
     * @param value The bytes for sourceUri to set.
     * @return This builder for chaining.
     */
    public Builder setSourceUriBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      sourceUri_ = value;
      bitField0_ |= 0x00000010;
      onChanged();
      return this;
    }

    private com.google.protobuf.LazyStringList sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    private void ensureSourceAccessionsIsMutable() {
      if (!((bitField0_ & 0x00000020) != 0)) {
        sourceAccessions_ = new com.google.protobuf.LazyStringArrayList(sourceAccessions_);
        bitField0_ |= 0x00000020;
       }
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @return A list containing the sourceAccessions.
     */
    public com.google.protobuf.ProtocolStringList
        getSourceAccessionsList() {
      return sourceAccessions_.getUnmodifiableView();
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @return The count of sourceAccessions.
     */
    public int getSourceAccessionsCount() {
      return sourceAccessions_.size();
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @param index The index of the element to return.
     * @return The sourceAccessions at the given index.
     */
    public java.lang.String getSourceAccessions(int index) {
      return sourceAccessions_.get(index);
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @param index The index of the value to return.
     * @return The bytes of the sourceAccessions at the given index.
     */
    public com.google.protobuf.ByteString
        getSourceAccessionsBytes(int index) {
      return sourceAccessions_.getByteString(index);
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @param index The index to set the value at.
     * @param value The sourceAccessions to set.
     * @return This builder for chaining.
     */
    public Builder setSourceAccessions(
        int index, java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      ensureSourceAccessionsIsMutable();
      sourceAccessions_.set(index, value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @param value The sourceAccessions to add.
     * @return This builder for chaining.
     */
    public Builder addSourceAccessions(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      ensureSourceAccessionsIsMutable();
      sourceAccessions_.add(value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @param values The sourceAccessions to add.
     * @return This builder for chaining.
     */
    public Builder addAllSourceAccessions(
        java.lang.Iterable<java.lang.String> values) {
      ensureSourceAccessionsIsMutable();
      com.google.protobuf.AbstractMessageLite.Builder.addAll(
          values, sourceAccessions_);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @return This builder for chaining.
     */
    public Builder clearSourceAccessions() {
      sourceAccessions_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      bitField0_ = (bitField0_ & ~0x00000020);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * All known corresponding accession IDs in INSDC (GenBank/ENA/DDBJ) ideally
     * with a version number, for example `GCF_000001405.26`.
     * </pre>
     *
     * <code>repeated string source_accessions = 6;</code>
     * @param value The bytes of the sourceAccessions to add.
     * @return This builder for chaining.
     */
    public Builder addSourceAccessionsBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      ensureSourceAccessionsIsMutable();
      sourceAccessions_.add(value);
      onChanged();
      return this;
    }

    private int ncbiTaxonId_ ;
    /**
     * <pre>
     * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
     * </pre>
     *
     * <code>int32 ncbi_taxon_id = 7;</code>
     * @return The ncbiTaxonId.
     */
    @java.lang.Override
    public int getNcbiTaxonId() {
      return ncbiTaxonId_;
    }
    /**
     * <pre>
     * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
     * </pre>
     *
     * <code>int32 ncbi_taxon_id = 7;</code>
     * @param value The ncbiTaxonId to set.
     * @return This builder for chaining.
     */
    public Builder setNcbiTaxonId(int value) {
      
      ncbiTaxonId_ = value;
      bitField0_ |= 0x00000040;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * ID from http://www.ncbi.nlm.nih.gov/taxonomy. For example, 9606 for human.
     * </pre>
     *
     * <code>int32 ncbi_taxon_id = 7;</code>
     * @return This builder for chaining.
     */
    public Builder clearNcbiTaxonId() {
      bitField0_ = (bitField0_ & ~0x00000040);
      ncbiTaxonId_ = 0;
      onChanged();
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.Reference)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.Reference)
  private static final com.google.genomics.v1.Reference DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.Reference();
  }

  public static com.google.genomics.v1.Reference getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<Reference>
      PARSER = new com.google.protobuf.AbstractParser<Reference>() {
    @java.lang.Override
    public Reference parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      Builder builder = newBuilder();
      try {
        builder.mergeFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(builder.buildPartial());
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(builder.buildPartial());
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(e)
            .setUnfinishedMessage(builder.buildPartial());
      }
      return builder.buildPartial();
    }
  };

  public static com.google.protobuf.Parser<Reference> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<Reference> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.Reference getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

