// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/reads.proto

package com.google.genomics.v1;

/**
 * <pre>
 * The read search response.
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.SearchReadsResponse}
 */
public final class SearchReadsResponse extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.SearchReadsResponse)
    SearchReadsResponseOrBuilder {
private static final long serialVersionUID = 0L;
  // Use SearchReadsResponse.newBuilder() to construct.
  private SearchReadsResponse(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private SearchReadsResponse() {
    alignments_ = java.util.Collections.emptyList();
    nextPageToken_ = "";
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new SearchReadsResponse();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_SearchReadsResponse_descriptor;
  }

  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_SearchReadsResponse_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.SearchReadsResponse.class, com.google.genomics.v1.SearchReadsResponse.Builder.class);
  }

  public static final int ALIGNMENTS_FIELD_NUMBER = 1;
  @SuppressWarnings("serial")
  private java.util.List<com.google.genomics.v1.Read> alignments_;
  /**
   * <pre>
   * The list of matching alignments sorted by mapped genomic coordinate,
   * if any, ascending in position within the same reference. Unmapped reads,
   * which have no position, are returned contiguously and are sorted in
   * ascending lexicographic order by fragment name.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
   */
  @java.lang.Override
  public java.util.List<com.google.genomics.v1.Read> getAlignmentsList() {
    return alignments_;
  }
  /**
   * <pre>
   * The list of matching alignments sorted by mapped genomic coordinate,
   * if any, ascending in position within the same reference. Unmapped reads,
   * which have no position, are returned contiguously and are sorted in
   * ascending lexicographic order by fragment name.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
   */
  @java.lang.Override
  public java.util.List<? extends com.google.genomics.v1.ReadOrBuilder> 
      getAlignmentsOrBuilderList() {
    return alignments_;
  }
  /**
   * <pre>
   * The list of matching alignments sorted by mapped genomic coordinate,
   * if any, ascending in position within the same reference. Unmapped reads,
   * which have no position, are returned contiguously and are sorted in
   * ascending lexicographic order by fragment name.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
   */
  @java.lang.Override
  public int getAlignmentsCount() {
    return alignments_.size();
  }
  /**
   * <pre>
   * The list of matching alignments sorted by mapped genomic coordinate,
   * if any, ascending in position within the same reference. Unmapped reads,
   * which have no position, are returned contiguously and are sorted in
   * ascending lexicographic order by fragment name.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.Read getAlignments(int index) {
    return alignments_.get(index);
  }
  /**
   * <pre>
   * The list of matching alignments sorted by mapped genomic coordinate,
   * if any, ascending in position within the same reference. Unmapped reads,
   * which have no position, are returned contiguously and are sorted in
   * ascending lexicographic order by fragment name.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReadOrBuilder getAlignmentsOrBuilder(
      int index) {
    return alignments_.get(index);
  }

  public static final int NEXT_PAGE_TOKEN_FIELD_NUMBER = 2;
  @SuppressWarnings("serial")
  private volatile java.lang.Object nextPageToken_ = "";
  /**
   * <pre>
   * The continuation token, which is used to page through large result sets.
   * Provide this value in a subsequent request to return the next page of
   * results. This field will be empty if there aren't any additional results.
   * </pre>
   *
   * <code>string next_page_token = 2;</code>
   * @return The nextPageToken.
   */
  @java.lang.Override
  public java.lang.String getNextPageToken() {
    java.lang.Object ref = nextPageToken_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      nextPageToken_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The continuation token, which is used to page through large result sets.
   * Provide this value in a subsequent request to return the next page of
   * results. This field will be empty if there aren't any additional results.
   * </pre>
   *
   * <code>string next_page_token = 2;</code>
   * @return The bytes for nextPageToken.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getNextPageTokenBytes() {
    java.lang.Object ref = nextPageToken_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      nextPageToken_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    for (int i = 0; i < alignments_.size(); i++) {
      output.writeMessage(1, alignments_.get(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(nextPageToken_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 2, nextPageToken_);
    }
    getUnknownFields().writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    for (int i = 0; i < alignments_.size(); i++) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(1, alignments_.get(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(nextPageToken_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(2, nextPageToken_);
    }
    size += getUnknownFields().getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.SearchReadsResponse)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.SearchReadsResponse other = (com.google.genomics.v1.SearchReadsResponse) obj;

    if (!getAlignmentsList()
        .equals(other.getAlignmentsList())) return false;
    if (!getNextPageToken()
        .equals(other.getNextPageToken())) return false;
    if (!getUnknownFields().equals(other.getUnknownFields())) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    if (getAlignmentsCount() > 0) {
      hash = (37 * hash) + ALIGNMENTS_FIELD_NUMBER;
      hash = (53 * hash) + getAlignmentsList().hashCode();
    }
    hash = (37 * hash) + NEXT_PAGE_TOKEN_FIELD_NUMBER;
    hash = (53 * hash) + getNextPageToken().hashCode();
    hash = (29 * hash) + getUnknownFields().hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.SearchReadsResponse parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.SearchReadsResponse prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * The read search response.
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.SearchReadsResponse}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.SearchReadsResponse)
      com.google.genomics.v1.SearchReadsResponseOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_SearchReadsResponse_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_SearchReadsResponse_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.SearchReadsResponse.class, com.google.genomics.v1.SearchReadsResponse.Builder.class);
    }

    // Construct using com.google.genomics.v1.SearchReadsResponse.newBuilder()
    private Builder() {

    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);

    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      bitField0_ = 0;
      if (alignmentsBuilder_ == null) {
        alignments_ = java.util.Collections.emptyList();
      } else {
        alignments_ = null;
        alignmentsBuilder_.clear();
      }
      bitField0_ = (bitField0_ & ~0x00000001);
      nextPageToken_ = "";
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.ReadsProto.internal_static_google_genomics_v1_SearchReadsResponse_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.SearchReadsResponse getDefaultInstanceForType() {
      return com.google.genomics.v1.SearchReadsResponse.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.SearchReadsResponse build() {
      com.google.genomics.v1.SearchReadsResponse result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.SearchReadsResponse buildPartial() {
      com.google.genomics.v1.SearchReadsResponse result = new com.google.genomics.v1.SearchReadsResponse(this);
      buildPartialRepeatedFields(result);
      if (bitField0_ != 0) { buildPartial0(result); }
      onBuilt();
      return result;
    }

    private void buildPartialRepeatedFields(com.google.genomics.v1.SearchReadsResponse result) {
      if (alignmentsBuilder_ == null) {
        if (((bitField0_ & 0x00000001) != 0)) {
          alignments_ = java.util.Collections.unmodifiableList(alignments_);
          bitField0_ = (bitField0_ & ~0x00000001);
        }
        result.alignments_ = alignments_;
      } else {
        result.alignments_ = alignmentsBuilder_.build();
      }
    }

    private void buildPartial0(com.google.genomics.v1.SearchReadsResponse result) {
      int from_bitField0_ = bitField0_;
      if (((from_bitField0_ & 0x00000002) != 0)) {
        result.nextPageToken_ = nextPageToken_;
      }
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.SearchReadsResponse) {
        return mergeFrom((com.google.genomics.v1.SearchReadsResponse)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.SearchReadsResponse other) {
      if (other == com.google.genomics.v1.SearchReadsResponse.getDefaultInstance()) return this;
      if (alignmentsBuilder_ == null) {
        if (!other.alignments_.isEmpty()) {
          if (alignments_.isEmpty()) {
            alignments_ = other.alignments_;
            bitField0_ = (bitField0_ & ~0x00000001);
          } else {
            ensureAlignmentsIsMutable();
            alignments_.addAll(other.alignments_);
          }
          onChanged();
        }
      } else {
        if (!other.alignments_.isEmpty()) {
          if (alignmentsBuilder_.isEmpty()) {
            alignmentsBuilder_.dispose();
            alignmentsBuilder_ = null;
            alignments_ = other.alignments_;
            bitField0_ = (bitField0_ & ~0x00000001);
            alignmentsBuilder_ = 
              com.google.protobuf.GeneratedMessageV3.alwaysUseFieldBuilders ?
                 getAlignmentsFieldBuilder() : null;
          } else {
            alignmentsBuilder_.addAllMessages(other.alignments_);
          }
        }
      }
      if (!other.getNextPageToken().isEmpty()) {
        nextPageToken_ = other.nextPageToken_;
        bitField0_ |= 0x00000002;
        onChanged();
      }
      this.mergeUnknownFields(other.getUnknownFields());
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 10: {
              com.google.genomics.v1.Read m =
                  input.readMessage(
                      com.google.genomics.v1.Read.parser(),
                      extensionRegistry);
              if (alignmentsBuilder_ == null) {
                ensureAlignmentsIsMutable();
                alignments_.add(m);
              } else {
                alignmentsBuilder_.addMessage(m);
              }
              break;
            } // case 10
            case 18: {
              nextPageToken_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000002;
              break;
            } // case 18
            default: {
              if (!super.parseUnknownField(input, extensionRegistry, tag)) {
                done = true; // was an endgroup tag
              }
              break;
            } // default:
          } // switch (tag)
        } // while (!done)
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.unwrapIOException();
      } finally {
        onChanged();
      } // finally
      return this;
    }
    private int bitField0_;

    private java.util.List<com.google.genomics.v1.Read> alignments_ =
      java.util.Collections.emptyList();
    private void ensureAlignmentsIsMutable() {
      if (!((bitField0_ & 0x00000001) != 0)) {
        alignments_ = new java.util.ArrayList<com.google.genomics.v1.Read>(alignments_);
        bitField0_ |= 0x00000001;
       }
    }

    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.Read, com.google.genomics.v1.Read.Builder, com.google.genomics.v1.ReadOrBuilder> alignmentsBuilder_;

    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public java.util.List<com.google.genomics.v1.Read> getAlignmentsList() {
      if (alignmentsBuilder_ == null) {
        return java.util.Collections.unmodifiableList(alignments_);
      } else {
        return alignmentsBuilder_.getMessageList();
      }
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public int getAlignmentsCount() {
      if (alignmentsBuilder_ == null) {
        return alignments_.size();
      } else {
        return alignmentsBuilder_.getCount();
      }
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public com.google.genomics.v1.Read getAlignments(int index) {
      if (alignmentsBuilder_ == null) {
        return alignments_.get(index);
      } else {
        return alignmentsBuilder_.getMessage(index);
      }
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder setAlignments(
        int index, com.google.genomics.v1.Read value) {
      if (alignmentsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureAlignmentsIsMutable();
        alignments_.set(index, value);
        onChanged();
      } else {
        alignmentsBuilder_.setMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder setAlignments(
        int index, com.google.genomics.v1.Read.Builder builderForValue) {
      if (alignmentsBuilder_ == null) {
        ensureAlignmentsIsMutable();
        alignments_.set(index, builderForValue.build());
        onChanged();
      } else {
        alignmentsBuilder_.setMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder addAlignments(com.google.genomics.v1.Read value) {
      if (alignmentsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureAlignmentsIsMutable();
        alignments_.add(value);
        onChanged();
      } else {
        alignmentsBuilder_.addMessage(value);
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder addAlignments(
        int index, com.google.genomics.v1.Read value) {
      if (alignmentsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureAlignmentsIsMutable();
        alignments_.add(index, value);
        onChanged();
      } else {
        alignmentsBuilder_.addMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder addAlignments(
        com.google.genomics.v1.Read.Builder builderForValue) {
      if (alignmentsBuilder_ == null) {
        ensureAlignmentsIsMutable();
        alignments_.add(builderForValue.build());
        onChanged();
      } else {
        alignmentsBuilder_.addMessage(builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder addAlignments(
        int index, com.google.genomics.v1.Read.Builder builderForValue) {
      if (alignmentsBuilder_ == null) {
        ensureAlignmentsIsMutable();
        alignments_.add(index, builderForValue.build());
        onChanged();
      } else {
        alignmentsBuilder_.addMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder addAllAlignments(
        java.lang.Iterable<? extends com.google.genomics.v1.Read> values) {
      if (alignmentsBuilder_ == null) {
        ensureAlignmentsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, alignments_);
        onChanged();
      } else {
        alignmentsBuilder_.addAllMessages(values);
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder clearAlignments() {
      if (alignmentsBuilder_ == null) {
        alignments_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000001);
        onChanged();
      } else {
        alignmentsBuilder_.clear();
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public Builder removeAlignments(int index) {
      if (alignmentsBuilder_ == null) {
        ensureAlignmentsIsMutable();
        alignments_.remove(index);
        onChanged();
      } else {
        alignmentsBuilder_.remove(index);
      }
      return this;
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public com.google.genomics.v1.Read.Builder getAlignmentsBuilder(
        int index) {
      return getAlignmentsFieldBuilder().getBuilder(index);
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public com.google.genomics.v1.ReadOrBuilder getAlignmentsOrBuilder(
        int index) {
      if (alignmentsBuilder_ == null) {
        return alignments_.get(index);  } else {
        return alignmentsBuilder_.getMessageOrBuilder(index);
      }
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public java.util.List<? extends com.google.genomics.v1.ReadOrBuilder> 
         getAlignmentsOrBuilderList() {
      if (alignmentsBuilder_ != null) {
        return alignmentsBuilder_.getMessageOrBuilderList();
      } else {
        return java.util.Collections.unmodifiableList(alignments_);
      }
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public com.google.genomics.v1.Read.Builder addAlignmentsBuilder() {
      return getAlignmentsFieldBuilder().addBuilder(
          com.google.genomics.v1.Read.getDefaultInstance());
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public com.google.genomics.v1.Read.Builder addAlignmentsBuilder(
        int index) {
      return getAlignmentsFieldBuilder().addBuilder(
          index, com.google.genomics.v1.Read.getDefaultInstance());
    }
    /**
     * <pre>
     * The list of matching alignments sorted by mapped genomic coordinate,
     * if any, ascending in position within the same reference. Unmapped reads,
     * which have no position, are returned contiguously and are sorted in
     * ascending lexicographic order by fragment name.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.Read alignments = 1;</code>
     */
    public java.util.List<com.google.genomics.v1.Read.Builder> 
         getAlignmentsBuilderList() {
      return getAlignmentsFieldBuilder().getBuilderList();
    }
    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.Read, com.google.genomics.v1.Read.Builder, com.google.genomics.v1.ReadOrBuilder> 
        getAlignmentsFieldBuilder() {
      if (alignmentsBuilder_ == null) {
        alignmentsBuilder_ = new com.google.protobuf.RepeatedFieldBuilderV3<
            com.google.genomics.v1.Read, com.google.genomics.v1.Read.Builder, com.google.genomics.v1.ReadOrBuilder>(
                alignments_,
                ((bitField0_ & 0x00000001) != 0),
                getParentForChildren(),
                isClean());
        alignments_ = null;
      }
      return alignmentsBuilder_;
    }

    private java.lang.Object nextPageToken_ = "";
    /**
     * <pre>
     * The continuation token, which is used to page through large result sets.
     * Provide this value in a subsequent request to return the next page of
     * results. This field will be empty if there aren't any additional results.
     * </pre>
     *
     * <code>string next_page_token = 2;</code>
     * @return The nextPageToken.
     */
    public java.lang.String getNextPageToken() {
      java.lang.Object ref = nextPageToken_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        nextPageToken_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The continuation token, which is used to page through large result sets.
     * Provide this value in a subsequent request to return the next page of
     * results. This field will be empty if there aren't any additional results.
     * </pre>
     *
     * <code>string next_page_token = 2;</code>
     * @return The bytes for nextPageToken.
     */
    public com.google.protobuf.ByteString
        getNextPageTokenBytes() {
      java.lang.Object ref = nextPageToken_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        nextPageToken_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The continuation token, which is used to page through large result sets.
     * Provide this value in a subsequent request to return the next page of
     * results. This field will be empty if there aren't any additional results.
     * </pre>
     *
     * <code>string next_page_token = 2;</code>
     * @param value The nextPageToken to set.
     * @return This builder for chaining.
     */
    public Builder setNextPageToken(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      nextPageToken_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The continuation token, which is used to page through large result sets.
     * Provide this value in a subsequent request to return the next page of
     * results. This field will be empty if there aren't any additional results.
     * </pre>
     *
     * <code>string next_page_token = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearNextPageToken() {
      nextPageToken_ = getDefaultInstance().getNextPageToken();
      bitField0_ = (bitField0_ & ~0x00000002);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The continuation token, which is used to page through large result sets.
     * Provide this value in a subsequent request to return the next page of
     * results. This field will be empty if there aren't any additional results.
     * </pre>
     *
     * <code>string next_page_token = 2;</code>
     * @param value The bytes for nextPageToken to set.
     * @return This builder for chaining.
     */
    public Builder setNextPageTokenBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      nextPageToken_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.SearchReadsResponse)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.SearchReadsResponse)
  private static final com.google.genomics.v1.SearchReadsResponse DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.SearchReadsResponse();
  }

  public static com.google.genomics.v1.SearchReadsResponse getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<SearchReadsResponse>
      PARSER = new com.google.protobuf.AbstractParser<SearchReadsResponse>() {
    @java.lang.Override
    public SearchReadsResponse parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      Builder builder = newBuilder();
      try {
        builder.mergeFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(builder.buildPartial());
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(builder.buildPartial());
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(e)
            .setUnfinishedMessage(builder.buildPartial());
      }
      return builder.buildPartial();
    }
  };

  public static com.google.protobuf.Parser<SearchReadsResponse> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<SearchReadsResponse> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.SearchReadsResponse getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

