// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/annotations.proto

package com.google.genomics.v1;

public interface TranscriptOrBuilder extends
    // @@protoc_insertion_point(interface_extends:google.genomics.v1.Transcript)
    com.google.protobuf.MessageOrBuilder {

  /**
   * <pre>
   * The annotation ID of the gene from which this transcript is transcribed.
   * </pre>
   *
   * <code>string gene_id = 1;</code>
   * @return The geneId.
   */
  java.lang.String getGeneId();
  /**
   * <pre>
   * The annotation ID of the gene from which this transcript is transcribed.
   * </pre>
   *
   * <code>string gene_id = 1;</code>
   * @return The bytes for geneId.
   */
  com.google.protobuf.ByteString
      getGeneIdBytes();

  /**
   * <pre>
   * The &lt;a href="http://en.wikipedia.org/wiki/Exon"&gt;exons&lt;/a&gt; that compose
   * this transcript. This field should be unset for genomes where transcript
   * splicing does not occur, for example prokaryotes.
   * Introns are regions of the transcript that are not included in the
   * spliced RNA product. Though not explicitly modeled here, intron ranges can
   * be deduced; all regions of this transcript that are not exons are introns.
   * Exonic sequences do not necessarily code for a translational product
   * (amino acids). Only the regions of exons bounded by the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] correspond
   * to coding DNA sequence.
   * Exons are ordered by start position and may not overlap.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Transcript.Exon exons = 2;</code>
   */
  java.util.List<com.google.genomics.v1.Transcript.Exon> 
      getExonsList();
  /**
   * <pre>
   * The &lt;a href="http://en.wikipedia.org/wiki/Exon"&gt;exons&lt;/a&gt; that compose
   * this transcript. This field should be unset for genomes where transcript
   * splicing does not occur, for example prokaryotes.
   * Introns are regions of the transcript that are not included in the
   * spliced RNA product. Though not explicitly modeled here, intron ranges can
   * be deduced; all regions of this transcript that are not exons are introns.
   * Exonic sequences do not necessarily code for a translational product
   * (amino acids). Only the regions of exons bounded by the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] correspond
   * to coding DNA sequence.
   * Exons are ordered by start position and may not overlap.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Transcript.Exon exons = 2;</code>
   */
  com.google.genomics.v1.Transcript.Exon getExons(int index);
  /**
   * <pre>
   * The &lt;a href="http://en.wikipedia.org/wiki/Exon"&gt;exons&lt;/a&gt; that compose
   * this transcript. This field should be unset for genomes where transcript
   * splicing does not occur, for example prokaryotes.
   * Introns are regions of the transcript that are not included in the
   * spliced RNA product. Though not explicitly modeled here, intron ranges can
   * be deduced; all regions of this transcript that are not exons are introns.
   * Exonic sequences do not necessarily code for a translational product
   * (amino acids). Only the regions of exons bounded by the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] correspond
   * to coding DNA sequence.
   * Exons are ordered by start position and may not overlap.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Transcript.Exon exons = 2;</code>
   */
  int getExonsCount();
  /**
   * <pre>
   * The &lt;a href="http://en.wikipedia.org/wiki/Exon"&gt;exons&lt;/a&gt; that compose
   * this transcript. This field should be unset for genomes where transcript
   * splicing does not occur, for example prokaryotes.
   * Introns are regions of the transcript that are not included in the
   * spliced RNA product. Though not explicitly modeled here, intron ranges can
   * be deduced; all regions of this transcript that are not exons are introns.
   * Exonic sequences do not necessarily code for a translational product
   * (amino acids). Only the regions of exons bounded by the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] correspond
   * to coding DNA sequence.
   * Exons are ordered by start position and may not overlap.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Transcript.Exon exons = 2;</code>
   */
  java.util.List<? extends com.google.genomics.v1.Transcript.ExonOrBuilder> 
      getExonsOrBuilderList();
  /**
   * <pre>
   * The &lt;a href="http://en.wikipedia.org/wiki/Exon"&gt;exons&lt;/a&gt; that compose
   * this transcript. This field should be unset for genomes where transcript
   * splicing does not occur, for example prokaryotes.
   * Introns are regions of the transcript that are not included in the
   * spliced RNA product. Though not explicitly modeled here, intron ranges can
   * be deduced; all regions of this transcript that are not exons are introns.
   * Exonic sequences do not necessarily code for a translational product
   * (amino acids). Only the regions of exons bounded by the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] correspond
   * to coding DNA sequence.
   * Exons are ordered by start position and may not overlap.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.Transcript.Exon exons = 2;</code>
   */
  com.google.genomics.v1.Transcript.ExonOrBuilder getExonsOrBuilder(
      int index);

  /**
   * <pre>
   * The range of the coding sequence for this transcript, if any. To determine
   * the exact ranges of coding sequence, intersect this range with those of the
   * [exons][google.genomics.v1.Transcript.exons], if any. If there are any
   * [exons][google.genomics.v1.Transcript.exons], the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] must start
   * and end within them.
   * Note that in some cases, the reference genome will not exactly match the
   * observed mRNA transcript e.g. due to variance in the source genome from
   * reference. In these cases,
   * [exon.frame][google.genomics.v1.Transcript.Exon.frame] will not necessarily
   * match the expected reference reading frame and coding exon reference bases
   * cannot necessarily be concatenated to produce the original transcript mRNA.
   * </pre>
   *
   * <code>.google.genomics.v1.Transcript.CodingSequence coding_sequence = 3;</code>
   * @return Whether the codingSequence field is set.
   */
  boolean hasCodingSequence();
  /**
   * <pre>
   * The range of the coding sequence for this transcript, if any. To determine
   * the exact ranges of coding sequence, intersect this range with those of the
   * [exons][google.genomics.v1.Transcript.exons], if any. If there are any
   * [exons][google.genomics.v1.Transcript.exons], the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] must start
   * and end within them.
   * Note that in some cases, the reference genome will not exactly match the
   * observed mRNA transcript e.g. due to variance in the source genome from
   * reference. In these cases,
   * [exon.frame][google.genomics.v1.Transcript.Exon.frame] will not necessarily
   * match the expected reference reading frame and coding exon reference bases
   * cannot necessarily be concatenated to produce the original transcript mRNA.
   * </pre>
   *
   * <code>.google.genomics.v1.Transcript.CodingSequence coding_sequence = 3;</code>
   * @return The codingSequence.
   */
  com.google.genomics.v1.Transcript.CodingSequence getCodingSequence();
  /**
   * <pre>
   * The range of the coding sequence for this transcript, if any. To determine
   * the exact ranges of coding sequence, intersect this range with those of the
   * [exons][google.genomics.v1.Transcript.exons], if any. If there are any
   * [exons][google.genomics.v1.Transcript.exons], the
   * [codingSequence][google.genomics.v1.Transcript.coding_sequence] must start
   * and end within them.
   * Note that in some cases, the reference genome will not exactly match the
   * observed mRNA transcript e.g. due to variance in the source genome from
   * reference. In these cases,
   * [exon.frame][google.genomics.v1.Transcript.Exon.frame] will not necessarily
   * match the expected reference reading frame and coding exon reference bases
   * cannot necessarily be concatenated to produce the original transcript mRNA.
   * </pre>
   *
   * <code>.google.genomics.v1.Transcript.CodingSequence coding_sequence = 3;</code>
   */
  com.google.genomics.v1.Transcript.CodingSequenceOrBuilder getCodingSequenceOrBuilder();
}
