// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/variants.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A variant set is a collection of call sets and variants. It contains summary
 * statistics of those contents. A variant set belongs to a dataset.
 * For more genomics resource definitions, see [Fundamentals of Google
 * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.VariantSet}
 */
public final class VariantSet extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.VariantSet)
    VariantSetOrBuilder {
private static final long serialVersionUID = 0L;
  // Use VariantSet.newBuilder() to construct.
  private VariantSet(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private VariantSet() {
    datasetId_ = "";
    id_ = "";
    referenceSetId_ = "";
    referenceBounds_ = java.util.Collections.emptyList();
    metadata_ = java.util.Collections.emptyList();
    name_ = "";
    description_ = "";
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new VariantSet();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantSet_descriptor;
  }

  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantSet_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.VariantSet.class, com.google.genomics.v1.VariantSet.Builder.class);
  }

  public static final int DATASET_ID_FIELD_NUMBER = 1;
  @SuppressWarnings("serial")
  private volatile java.lang.Object datasetId_ = "";
  /**
   * <pre>
   * The dataset to which this variant set belongs.
   * </pre>
   *
   * <code>string dataset_id = 1;</code>
   * @return The datasetId.
   */
  @java.lang.Override
  public java.lang.String getDatasetId() {
    java.lang.Object ref = datasetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      datasetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The dataset to which this variant set belongs.
   * </pre>
   *
   * <code>string dataset_id = 1;</code>
   * @return The bytes for datasetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getDatasetIdBytes() {
    java.lang.Object ref = datasetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      datasetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int ID_FIELD_NUMBER = 2;
  @SuppressWarnings("serial")
  private volatile java.lang.Object id_ = "";
  /**
   * <pre>
   * The server-generated variant set ID, unique across all variant sets.
   * </pre>
   *
   * <code>string id = 2;</code>
   * @return The id.
   */
  @java.lang.Override
  public java.lang.String getId() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      id_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The server-generated variant set ID, unique across all variant sets.
   * </pre>
   *
   * <code>string id = 2;</code>
   * @return The bytes for id.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getIdBytes() {
    java.lang.Object ref = id_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      id_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int REFERENCE_SET_ID_FIELD_NUMBER = 6;
  @SuppressWarnings("serial")
  private volatile java.lang.Object referenceSetId_ = "";
  /**
   * <pre>
   * The reference set to which the variant set is mapped. The reference set
   * describes the alignment provenance of the variant set, while the
   * `referenceBounds` describe the shape of the actual variant data. The
   * reference set's reference names are a superset of those found in the
   * `referenceBounds`.
   * For example, given a variant set that is mapped to the GRCh38 reference set
   * and contains a single variant on reference 'X', `referenceBounds` would
   * contain only an entry for 'X', while the associated reference set
   * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
   * </pre>
   *
   * <code>string reference_set_id = 6;</code>
   * @return The referenceSetId.
   */
  @java.lang.Override
  public java.lang.String getReferenceSetId() {
    java.lang.Object ref = referenceSetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      referenceSetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The reference set to which the variant set is mapped. The reference set
   * describes the alignment provenance of the variant set, while the
   * `referenceBounds` describe the shape of the actual variant data. The
   * reference set's reference names are a superset of those found in the
   * `referenceBounds`.
   * For example, given a variant set that is mapped to the GRCh38 reference set
   * and contains a single variant on reference 'X', `referenceBounds` would
   * contain only an entry for 'X', while the associated reference set
   * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
   * </pre>
   *
   * <code>string reference_set_id = 6;</code>
   * @return The bytes for referenceSetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getReferenceSetIdBytes() {
    java.lang.Object ref = referenceSetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      referenceSetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int REFERENCE_BOUNDS_FIELD_NUMBER = 5;
  @SuppressWarnings("serial")
  private java.util.List<com.google.genomics.v1.ReferenceBound> referenceBounds_;
  /**
   * <pre>
   * A list of all references used by the variants in a variant set
   * with associated coordinate upper bounds for each one.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
   */
  @java.lang.Override
  public java.util.List<com.google.genomics.v1.ReferenceBound> getReferenceBoundsList() {
    return referenceBounds_;
  }
  /**
   * <pre>
   * A list of all references used by the variants in a variant set
   * with associated coordinate upper bounds for each one.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
   */
  @java.lang.Override
  public java.util.List<? extends com.google.genomics.v1.ReferenceBoundOrBuilder> 
      getReferenceBoundsOrBuilderList() {
    return referenceBounds_;
  }
  /**
   * <pre>
   * A list of all references used by the variants in a variant set
   * with associated coordinate upper bounds for each one.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
   */
  @java.lang.Override
  public int getReferenceBoundsCount() {
    return referenceBounds_.size();
  }
  /**
   * <pre>
   * A list of all references used by the variants in a variant set
   * with associated coordinate upper bounds for each one.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReferenceBound getReferenceBounds(int index) {
    return referenceBounds_.get(index);
  }
  /**
   * <pre>
   * A list of all references used by the variants in a variant set
   * with associated coordinate upper bounds for each one.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.ReferenceBoundOrBuilder getReferenceBoundsOrBuilder(
      int index) {
    return referenceBounds_.get(index);
  }

  public static final int METADATA_FIELD_NUMBER = 4;
  @SuppressWarnings("serial")
  private java.util.List<com.google.genomics.v1.VariantSetMetadata> metadata_;
  /**
   * <pre>
   * The metadata associated with this variant set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
   */
  @java.lang.Override
  public java.util.List<com.google.genomics.v1.VariantSetMetadata> getMetadataList() {
    return metadata_;
  }
  /**
   * <pre>
   * The metadata associated with this variant set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
   */
  @java.lang.Override
  public java.util.List<? extends com.google.genomics.v1.VariantSetMetadataOrBuilder> 
      getMetadataOrBuilderList() {
    return metadata_;
  }
  /**
   * <pre>
   * The metadata associated with this variant set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
   */
  @java.lang.Override
  public int getMetadataCount() {
    return metadata_.size();
  }
  /**
   * <pre>
   * The metadata associated with this variant set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.VariantSetMetadata getMetadata(int index) {
    return metadata_.get(index);
  }
  /**
   * <pre>
   * The metadata associated with this variant set.
   * </pre>
   *
   * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
   */
  @java.lang.Override
  public com.google.genomics.v1.VariantSetMetadataOrBuilder getMetadataOrBuilder(
      int index) {
    return metadata_.get(index);
  }

  public static final int NAME_FIELD_NUMBER = 7;
  @SuppressWarnings("serial")
  private volatile java.lang.Object name_ = "";
  /**
   * <pre>
   * User-specified, mutable name.
   * </pre>
   *
   * <code>string name = 7;</code>
   * @return The name.
   */
  @java.lang.Override
  public java.lang.String getName() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      name_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * User-specified, mutable name.
   * </pre>
   *
   * <code>string name = 7;</code>
   * @return The bytes for name.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getNameBytes() {
    java.lang.Object ref = name_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      name_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int DESCRIPTION_FIELD_NUMBER = 8;
  @SuppressWarnings("serial")
  private volatile java.lang.Object description_ = "";
  /**
   * <pre>
   * A textual description of this variant set.
   * </pre>
   *
   * <code>string description = 8;</code>
   * @return The description.
   */
  @java.lang.Override
  public java.lang.String getDescription() {
    java.lang.Object ref = description_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      description_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * A textual description of this variant set.
   * </pre>
   *
   * <code>string description = 8;</code>
   * @return The bytes for description.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getDescriptionBytes() {
    java.lang.Object ref = description_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      description_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(datasetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 1, datasetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 2, id_);
    }
    for (int i = 0; i < metadata_.size(); i++) {
      output.writeMessage(4, metadata_.get(i));
    }
    for (int i = 0; i < referenceBounds_.size(); i++) {
      output.writeMessage(5, referenceBounds_.get(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(referenceSetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 6, referenceSetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 7, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(description_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 8, description_);
    }
    getUnknownFields().writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(datasetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(1, datasetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(id_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(2, id_);
    }
    for (int i = 0; i < metadata_.size(); i++) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(4, metadata_.get(i));
    }
    for (int i = 0; i < referenceBounds_.size(); i++) {
      size += com.google.protobuf.CodedOutputStream
        .computeMessageSize(5, referenceBounds_.get(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(referenceSetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(6, referenceSetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(name_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(7, name_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(description_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(8, description_);
    }
    size += getUnknownFields().getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.VariantSet)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.VariantSet other = (com.google.genomics.v1.VariantSet) obj;

    if (!getDatasetId()
        .equals(other.getDatasetId())) return false;
    if (!getId()
        .equals(other.getId())) return false;
    if (!getReferenceSetId()
        .equals(other.getReferenceSetId())) return false;
    if (!getReferenceBoundsList()
        .equals(other.getReferenceBoundsList())) return false;
    if (!getMetadataList()
        .equals(other.getMetadataList())) return false;
    if (!getName()
        .equals(other.getName())) return false;
    if (!getDescription()
        .equals(other.getDescription())) return false;
    if (!getUnknownFields().equals(other.getUnknownFields())) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    hash = (37 * hash) + DATASET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getDatasetId().hashCode();
    hash = (37 * hash) + ID_FIELD_NUMBER;
    hash = (53 * hash) + getId().hashCode();
    hash = (37 * hash) + REFERENCE_SET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getReferenceSetId().hashCode();
    if (getReferenceBoundsCount() > 0) {
      hash = (37 * hash) + REFERENCE_BOUNDS_FIELD_NUMBER;
      hash = (53 * hash) + getReferenceBoundsList().hashCode();
    }
    if (getMetadataCount() > 0) {
      hash = (37 * hash) + METADATA_FIELD_NUMBER;
      hash = (53 * hash) + getMetadataList().hashCode();
    }
    hash = (37 * hash) + NAME_FIELD_NUMBER;
    hash = (53 * hash) + getName().hashCode();
    hash = (37 * hash) + DESCRIPTION_FIELD_NUMBER;
    hash = (53 * hash) + getDescription().hashCode();
    hash = (29 * hash) + getUnknownFields().hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.VariantSet parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantSet parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.VariantSet parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.VariantSet parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.VariantSet prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A variant set is a collection of call sets and variants. It contains summary
   * statistics of those contents. A variant set belongs to a dataset.
   * For more genomics resource definitions, see [Fundamentals of Google
   * Genomics](https://cloud.google.com/genomics/fundamentals-of-google-genomics)
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.VariantSet}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.VariantSet)
      com.google.genomics.v1.VariantSetOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantSet_descriptor;
    }

    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantSet_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.VariantSet.class, com.google.genomics.v1.VariantSet.Builder.class);
    }

    // Construct using com.google.genomics.v1.VariantSet.newBuilder()
    private Builder() {

    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);

    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      bitField0_ = 0;
      datasetId_ = "";
      id_ = "";
      referenceSetId_ = "";
      if (referenceBoundsBuilder_ == null) {
        referenceBounds_ = java.util.Collections.emptyList();
      } else {
        referenceBounds_ = null;
        referenceBoundsBuilder_.clear();
      }
      bitField0_ = (bitField0_ & ~0x00000008);
      if (metadataBuilder_ == null) {
        metadata_ = java.util.Collections.emptyList();
      } else {
        metadata_ = null;
        metadataBuilder_.clear();
      }
      bitField0_ = (bitField0_ & ~0x00000010);
      name_ = "";
      description_ = "";
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantSet_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.VariantSet getDefaultInstanceForType() {
      return com.google.genomics.v1.VariantSet.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.VariantSet build() {
      com.google.genomics.v1.VariantSet result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.VariantSet buildPartial() {
      com.google.genomics.v1.VariantSet result = new com.google.genomics.v1.VariantSet(this);
      buildPartialRepeatedFields(result);
      if (bitField0_ != 0) { buildPartial0(result); }
      onBuilt();
      return result;
    }

    private void buildPartialRepeatedFields(com.google.genomics.v1.VariantSet result) {
      if (referenceBoundsBuilder_ == null) {
        if (((bitField0_ & 0x00000008) != 0)) {
          referenceBounds_ = java.util.Collections.unmodifiableList(referenceBounds_);
          bitField0_ = (bitField0_ & ~0x00000008);
        }
        result.referenceBounds_ = referenceBounds_;
      } else {
        result.referenceBounds_ = referenceBoundsBuilder_.build();
      }
      if (metadataBuilder_ == null) {
        if (((bitField0_ & 0x00000010) != 0)) {
          metadata_ = java.util.Collections.unmodifiableList(metadata_);
          bitField0_ = (bitField0_ & ~0x00000010);
        }
        result.metadata_ = metadata_;
      } else {
        result.metadata_ = metadataBuilder_.build();
      }
    }

    private void buildPartial0(com.google.genomics.v1.VariantSet result) {
      int from_bitField0_ = bitField0_;
      if (((from_bitField0_ & 0x00000001) != 0)) {
        result.datasetId_ = datasetId_;
      }
      if (((from_bitField0_ & 0x00000002) != 0)) {
        result.id_ = id_;
      }
      if (((from_bitField0_ & 0x00000004) != 0)) {
        result.referenceSetId_ = referenceSetId_;
      }
      if (((from_bitField0_ & 0x00000020) != 0)) {
        result.name_ = name_;
      }
      if (((from_bitField0_ & 0x00000040) != 0)) {
        result.description_ = description_;
      }
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.VariantSet) {
        return mergeFrom((com.google.genomics.v1.VariantSet)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.VariantSet other) {
      if (other == com.google.genomics.v1.VariantSet.getDefaultInstance()) return this;
      if (!other.getDatasetId().isEmpty()) {
        datasetId_ = other.datasetId_;
        bitField0_ |= 0x00000001;
        onChanged();
      }
      if (!other.getId().isEmpty()) {
        id_ = other.id_;
        bitField0_ |= 0x00000002;
        onChanged();
      }
      if (!other.getReferenceSetId().isEmpty()) {
        referenceSetId_ = other.referenceSetId_;
        bitField0_ |= 0x00000004;
        onChanged();
      }
      if (referenceBoundsBuilder_ == null) {
        if (!other.referenceBounds_.isEmpty()) {
          if (referenceBounds_.isEmpty()) {
            referenceBounds_ = other.referenceBounds_;
            bitField0_ = (bitField0_ & ~0x00000008);
          } else {
            ensureReferenceBoundsIsMutable();
            referenceBounds_.addAll(other.referenceBounds_);
          }
          onChanged();
        }
      } else {
        if (!other.referenceBounds_.isEmpty()) {
          if (referenceBoundsBuilder_.isEmpty()) {
            referenceBoundsBuilder_.dispose();
            referenceBoundsBuilder_ = null;
            referenceBounds_ = other.referenceBounds_;
            bitField0_ = (bitField0_ & ~0x00000008);
            referenceBoundsBuilder_ = 
              com.google.protobuf.GeneratedMessageV3.alwaysUseFieldBuilders ?
                 getReferenceBoundsFieldBuilder() : null;
          } else {
            referenceBoundsBuilder_.addAllMessages(other.referenceBounds_);
          }
        }
      }
      if (metadataBuilder_ == null) {
        if (!other.metadata_.isEmpty()) {
          if (metadata_.isEmpty()) {
            metadata_ = other.metadata_;
            bitField0_ = (bitField0_ & ~0x00000010);
          } else {
            ensureMetadataIsMutable();
            metadata_.addAll(other.metadata_);
          }
          onChanged();
        }
      } else {
        if (!other.metadata_.isEmpty()) {
          if (metadataBuilder_.isEmpty()) {
            metadataBuilder_.dispose();
            metadataBuilder_ = null;
            metadata_ = other.metadata_;
            bitField0_ = (bitField0_ & ~0x00000010);
            metadataBuilder_ = 
              com.google.protobuf.GeneratedMessageV3.alwaysUseFieldBuilders ?
                 getMetadataFieldBuilder() : null;
          } else {
            metadataBuilder_.addAllMessages(other.metadata_);
          }
        }
      }
      if (!other.getName().isEmpty()) {
        name_ = other.name_;
        bitField0_ |= 0x00000020;
        onChanged();
      }
      if (!other.getDescription().isEmpty()) {
        description_ = other.description_;
        bitField0_ |= 0x00000040;
        onChanged();
      }
      this.mergeUnknownFields(other.getUnknownFields());
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 10: {
              datasetId_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000001;
              break;
            } // case 10
            case 18: {
              id_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000002;
              break;
            } // case 18
            case 34: {
              com.google.genomics.v1.VariantSetMetadata m =
                  input.readMessage(
                      com.google.genomics.v1.VariantSetMetadata.parser(),
                      extensionRegistry);
              if (metadataBuilder_ == null) {
                ensureMetadataIsMutable();
                metadata_.add(m);
              } else {
                metadataBuilder_.addMessage(m);
              }
              break;
            } // case 34
            case 42: {
              com.google.genomics.v1.ReferenceBound m =
                  input.readMessage(
                      com.google.genomics.v1.ReferenceBound.parser(),
                      extensionRegistry);
              if (referenceBoundsBuilder_ == null) {
                ensureReferenceBoundsIsMutable();
                referenceBounds_.add(m);
              } else {
                referenceBoundsBuilder_.addMessage(m);
              }
              break;
            } // case 42
            case 50: {
              referenceSetId_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000004;
              break;
            } // case 50
            case 58: {
              name_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000020;
              break;
            } // case 58
            case 66: {
              description_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000040;
              break;
            } // case 66
            default: {
              if (!super.parseUnknownField(input, extensionRegistry, tag)) {
                done = true; // was an endgroup tag
              }
              break;
            } // default:
          } // switch (tag)
        } // while (!done)
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.unwrapIOException();
      } finally {
        onChanged();
      } // finally
      return this;
    }
    private int bitField0_;

    private java.lang.Object datasetId_ = "";
    /**
     * <pre>
     * The dataset to which this variant set belongs.
     * </pre>
     *
     * <code>string dataset_id = 1;</code>
     * @return The datasetId.
     */
    public java.lang.String getDatasetId() {
      java.lang.Object ref = datasetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        datasetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The dataset to which this variant set belongs.
     * </pre>
     *
     * <code>string dataset_id = 1;</code>
     * @return The bytes for datasetId.
     */
    public com.google.protobuf.ByteString
        getDatasetIdBytes() {
      java.lang.Object ref = datasetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        datasetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The dataset to which this variant set belongs.
     * </pre>
     *
     * <code>string dataset_id = 1;</code>
     * @param value The datasetId to set.
     * @return This builder for chaining.
     */
    public Builder setDatasetId(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      datasetId_ = value;
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The dataset to which this variant set belongs.
     * </pre>
     *
     * <code>string dataset_id = 1;</code>
     * @return This builder for chaining.
     */
    public Builder clearDatasetId() {
      datasetId_ = getDefaultInstance().getDatasetId();
      bitField0_ = (bitField0_ & ~0x00000001);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The dataset to which this variant set belongs.
     * </pre>
     *
     * <code>string dataset_id = 1;</code>
     * @param value The bytes for datasetId to set.
     * @return This builder for chaining.
     */
    public Builder setDatasetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      datasetId_ = value;
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }

    private java.lang.Object id_ = "";
    /**
     * <pre>
     * The server-generated variant set ID, unique across all variant sets.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return The id.
     */
    public java.lang.String getId() {
      java.lang.Object ref = id_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        id_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The server-generated variant set ID, unique across all variant sets.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return The bytes for id.
     */
    public com.google.protobuf.ByteString
        getIdBytes() {
      java.lang.Object ref = id_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        id_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The server-generated variant set ID, unique across all variant sets.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @param value The id to set.
     * @return This builder for chaining.
     */
    public Builder setId(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      id_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated variant set ID, unique across all variant sets.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @return This builder for chaining.
     */
    public Builder clearId() {
      id_ = getDefaultInstance().getId();
      bitField0_ = (bitField0_ & ~0x00000002);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The server-generated variant set ID, unique across all variant sets.
     * </pre>
     *
     * <code>string id = 2;</code>
     * @param value The bytes for id to set.
     * @return This builder for chaining.
     */
    public Builder setIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      id_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }

    private java.lang.Object referenceSetId_ = "";
    /**
     * <pre>
     * The reference set to which the variant set is mapped. The reference set
     * describes the alignment provenance of the variant set, while the
     * `referenceBounds` describe the shape of the actual variant data. The
     * reference set's reference names are a superset of those found in the
     * `referenceBounds`.
     * For example, given a variant set that is mapped to the GRCh38 reference set
     * and contains a single variant on reference 'X', `referenceBounds` would
     * contain only an entry for 'X', while the associated reference set
     * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
     * </pre>
     *
     * <code>string reference_set_id = 6;</code>
     * @return The referenceSetId.
     */
    public java.lang.String getReferenceSetId() {
      java.lang.Object ref = referenceSetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        referenceSetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The reference set to which the variant set is mapped. The reference set
     * describes the alignment provenance of the variant set, while the
     * `referenceBounds` describe the shape of the actual variant data. The
     * reference set's reference names are a superset of those found in the
     * `referenceBounds`.
     * For example, given a variant set that is mapped to the GRCh38 reference set
     * and contains a single variant on reference 'X', `referenceBounds` would
     * contain only an entry for 'X', while the associated reference set
     * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
     * </pre>
     *
     * <code>string reference_set_id = 6;</code>
     * @return The bytes for referenceSetId.
     */
    public com.google.protobuf.ByteString
        getReferenceSetIdBytes() {
      java.lang.Object ref = referenceSetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        referenceSetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The reference set to which the variant set is mapped. The reference set
     * describes the alignment provenance of the variant set, while the
     * `referenceBounds` describe the shape of the actual variant data. The
     * reference set's reference names are a superset of those found in the
     * `referenceBounds`.
     * For example, given a variant set that is mapped to the GRCh38 reference set
     * and contains a single variant on reference 'X', `referenceBounds` would
     * contain only an entry for 'X', while the associated reference set
     * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
     * </pre>
     *
     * <code>string reference_set_id = 6;</code>
     * @param value The referenceSetId to set.
     * @return This builder for chaining.
     */
    public Builder setReferenceSetId(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      referenceSetId_ = value;
      bitField0_ |= 0x00000004;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The reference set to which the variant set is mapped. The reference set
     * describes the alignment provenance of the variant set, while the
     * `referenceBounds` describe the shape of the actual variant data. The
     * reference set's reference names are a superset of those found in the
     * `referenceBounds`.
     * For example, given a variant set that is mapped to the GRCh38 reference set
     * and contains a single variant on reference 'X', `referenceBounds` would
     * contain only an entry for 'X', while the associated reference set
     * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
     * </pre>
     *
     * <code>string reference_set_id = 6;</code>
     * @return This builder for chaining.
     */
    public Builder clearReferenceSetId() {
      referenceSetId_ = getDefaultInstance().getReferenceSetId();
      bitField0_ = (bitField0_ & ~0x00000004);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The reference set to which the variant set is mapped. The reference set
     * describes the alignment provenance of the variant set, while the
     * `referenceBounds` describe the shape of the actual variant data. The
     * reference set's reference names are a superset of those found in the
     * `referenceBounds`.
     * For example, given a variant set that is mapped to the GRCh38 reference set
     * and contains a single variant on reference 'X', `referenceBounds` would
     * contain only an entry for 'X', while the associated reference set
     * enumerates all possible references: '1', '2', 'X', 'Y', 'MT', etc.
     * </pre>
     *
     * <code>string reference_set_id = 6;</code>
     * @param value The bytes for referenceSetId to set.
     * @return This builder for chaining.
     */
    public Builder setReferenceSetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      referenceSetId_ = value;
      bitField0_ |= 0x00000004;
      onChanged();
      return this;
    }

    private java.util.List<com.google.genomics.v1.ReferenceBound> referenceBounds_ =
      java.util.Collections.emptyList();
    private void ensureReferenceBoundsIsMutable() {
      if (!((bitField0_ & 0x00000008) != 0)) {
        referenceBounds_ = new java.util.ArrayList<com.google.genomics.v1.ReferenceBound>(referenceBounds_);
        bitField0_ |= 0x00000008;
       }
    }

    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.ReferenceBound, com.google.genomics.v1.ReferenceBound.Builder, com.google.genomics.v1.ReferenceBoundOrBuilder> referenceBoundsBuilder_;

    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public java.util.List<com.google.genomics.v1.ReferenceBound> getReferenceBoundsList() {
      if (referenceBoundsBuilder_ == null) {
        return java.util.Collections.unmodifiableList(referenceBounds_);
      } else {
        return referenceBoundsBuilder_.getMessageList();
      }
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public int getReferenceBoundsCount() {
      if (referenceBoundsBuilder_ == null) {
        return referenceBounds_.size();
      } else {
        return referenceBoundsBuilder_.getCount();
      }
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public com.google.genomics.v1.ReferenceBound getReferenceBounds(int index) {
      if (referenceBoundsBuilder_ == null) {
        return referenceBounds_.get(index);
      } else {
        return referenceBoundsBuilder_.getMessage(index);
      }
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder setReferenceBounds(
        int index, com.google.genomics.v1.ReferenceBound value) {
      if (referenceBoundsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureReferenceBoundsIsMutable();
        referenceBounds_.set(index, value);
        onChanged();
      } else {
        referenceBoundsBuilder_.setMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder setReferenceBounds(
        int index, com.google.genomics.v1.ReferenceBound.Builder builderForValue) {
      if (referenceBoundsBuilder_ == null) {
        ensureReferenceBoundsIsMutable();
        referenceBounds_.set(index, builderForValue.build());
        onChanged();
      } else {
        referenceBoundsBuilder_.setMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder addReferenceBounds(com.google.genomics.v1.ReferenceBound value) {
      if (referenceBoundsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureReferenceBoundsIsMutable();
        referenceBounds_.add(value);
        onChanged();
      } else {
        referenceBoundsBuilder_.addMessage(value);
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder addReferenceBounds(
        int index, com.google.genomics.v1.ReferenceBound value) {
      if (referenceBoundsBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureReferenceBoundsIsMutable();
        referenceBounds_.add(index, value);
        onChanged();
      } else {
        referenceBoundsBuilder_.addMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder addReferenceBounds(
        com.google.genomics.v1.ReferenceBound.Builder builderForValue) {
      if (referenceBoundsBuilder_ == null) {
        ensureReferenceBoundsIsMutable();
        referenceBounds_.add(builderForValue.build());
        onChanged();
      } else {
        referenceBoundsBuilder_.addMessage(builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder addReferenceBounds(
        int index, com.google.genomics.v1.ReferenceBound.Builder builderForValue) {
      if (referenceBoundsBuilder_ == null) {
        ensureReferenceBoundsIsMutable();
        referenceBounds_.add(index, builderForValue.build());
        onChanged();
      } else {
        referenceBoundsBuilder_.addMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder addAllReferenceBounds(
        java.lang.Iterable<? extends com.google.genomics.v1.ReferenceBound> values) {
      if (referenceBoundsBuilder_ == null) {
        ensureReferenceBoundsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, referenceBounds_);
        onChanged();
      } else {
        referenceBoundsBuilder_.addAllMessages(values);
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder clearReferenceBounds() {
      if (referenceBoundsBuilder_ == null) {
        referenceBounds_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000008);
        onChanged();
      } else {
        referenceBoundsBuilder_.clear();
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public Builder removeReferenceBounds(int index) {
      if (referenceBoundsBuilder_ == null) {
        ensureReferenceBoundsIsMutable();
        referenceBounds_.remove(index);
        onChanged();
      } else {
        referenceBoundsBuilder_.remove(index);
      }
      return this;
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public com.google.genomics.v1.ReferenceBound.Builder getReferenceBoundsBuilder(
        int index) {
      return getReferenceBoundsFieldBuilder().getBuilder(index);
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public com.google.genomics.v1.ReferenceBoundOrBuilder getReferenceBoundsOrBuilder(
        int index) {
      if (referenceBoundsBuilder_ == null) {
        return referenceBounds_.get(index);  } else {
        return referenceBoundsBuilder_.getMessageOrBuilder(index);
      }
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public java.util.List<? extends com.google.genomics.v1.ReferenceBoundOrBuilder> 
         getReferenceBoundsOrBuilderList() {
      if (referenceBoundsBuilder_ != null) {
        return referenceBoundsBuilder_.getMessageOrBuilderList();
      } else {
        return java.util.Collections.unmodifiableList(referenceBounds_);
      }
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public com.google.genomics.v1.ReferenceBound.Builder addReferenceBoundsBuilder() {
      return getReferenceBoundsFieldBuilder().addBuilder(
          com.google.genomics.v1.ReferenceBound.getDefaultInstance());
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public com.google.genomics.v1.ReferenceBound.Builder addReferenceBoundsBuilder(
        int index) {
      return getReferenceBoundsFieldBuilder().addBuilder(
          index, com.google.genomics.v1.ReferenceBound.getDefaultInstance());
    }
    /**
     * <pre>
     * A list of all references used by the variants in a variant set
     * with associated coordinate upper bounds for each one.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.ReferenceBound reference_bounds = 5;</code>
     */
    public java.util.List<com.google.genomics.v1.ReferenceBound.Builder> 
         getReferenceBoundsBuilderList() {
      return getReferenceBoundsFieldBuilder().getBuilderList();
    }
    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.ReferenceBound, com.google.genomics.v1.ReferenceBound.Builder, com.google.genomics.v1.ReferenceBoundOrBuilder> 
        getReferenceBoundsFieldBuilder() {
      if (referenceBoundsBuilder_ == null) {
        referenceBoundsBuilder_ = new com.google.protobuf.RepeatedFieldBuilderV3<
            com.google.genomics.v1.ReferenceBound, com.google.genomics.v1.ReferenceBound.Builder, com.google.genomics.v1.ReferenceBoundOrBuilder>(
                referenceBounds_,
                ((bitField0_ & 0x00000008) != 0),
                getParentForChildren(),
                isClean());
        referenceBounds_ = null;
      }
      return referenceBoundsBuilder_;
    }

    private java.util.List<com.google.genomics.v1.VariantSetMetadata> metadata_ =
      java.util.Collections.emptyList();
    private void ensureMetadataIsMutable() {
      if (!((bitField0_ & 0x00000010) != 0)) {
        metadata_ = new java.util.ArrayList<com.google.genomics.v1.VariantSetMetadata>(metadata_);
        bitField0_ |= 0x00000010;
       }
    }

    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.VariantSetMetadata, com.google.genomics.v1.VariantSetMetadata.Builder, com.google.genomics.v1.VariantSetMetadataOrBuilder> metadataBuilder_;

    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public java.util.List<com.google.genomics.v1.VariantSetMetadata> getMetadataList() {
      if (metadataBuilder_ == null) {
        return java.util.Collections.unmodifiableList(metadata_);
      } else {
        return metadataBuilder_.getMessageList();
      }
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public int getMetadataCount() {
      if (metadataBuilder_ == null) {
        return metadata_.size();
      } else {
        return metadataBuilder_.getCount();
      }
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public com.google.genomics.v1.VariantSetMetadata getMetadata(int index) {
      if (metadataBuilder_ == null) {
        return metadata_.get(index);
      } else {
        return metadataBuilder_.getMessage(index);
      }
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder setMetadata(
        int index, com.google.genomics.v1.VariantSetMetadata value) {
      if (metadataBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureMetadataIsMutable();
        metadata_.set(index, value);
        onChanged();
      } else {
        metadataBuilder_.setMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder setMetadata(
        int index, com.google.genomics.v1.VariantSetMetadata.Builder builderForValue) {
      if (metadataBuilder_ == null) {
        ensureMetadataIsMutable();
        metadata_.set(index, builderForValue.build());
        onChanged();
      } else {
        metadataBuilder_.setMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder addMetadata(com.google.genomics.v1.VariantSetMetadata value) {
      if (metadataBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureMetadataIsMutable();
        metadata_.add(value);
        onChanged();
      } else {
        metadataBuilder_.addMessage(value);
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder addMetadata(
        int index, com.google.genomics.v1.VariantSetMetadata value) {
      if (metadataBuilder_ == null) {
        if (value == null) {
          throw new NullPointerException();
        }
        ensureMetadataIsMutable();
        metadata_.add(index, value);
        onChanged();
      } else {
        metadataBuilder_.addMessage(index, value);
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder addMetadata(
        com.google.genomics.v1.VariantSetMetadata.Builder builderForValue) {
      if (metadataBuilder_ == null) {
        ensureMetadataIsMutable();
        metadata_.add(builderForValue.build());
        onChanged();
      } else {
        metadataBuilder_.addMessage(builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder addMetadata(
        int index, com.google.genomics.v1.VariantSetMetadata.Builder builderForValue) {
      if (metadataBuilder_ == null) {
        ensureMetadataIsMutable();
        metadata_.add(index, builderForValue.build());
        onChanged();
      } else {
        metadataBuilder_.addMessage(index, builderForValue.build());
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder addAllMetadata(
        java.lang.Iterable<? extends com.google.genomics.v1.VariantSetMetadata> values) {
      if (metadataBuilder_ == null) {
        ensureMetadataIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, metadata_);
        onChanged();
      } else {
        metadataBuilder_.addAllMessages(values);
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder clearMetadata() {
      if (metadataBuilder_ == null) {
        metadata_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000010);
        onChanged();
      } else {
        metadataBuilder_.clear();
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public Builder removeMetadata(int index) {
      if (metadataBuilder_ == null) {
        ensureMetadataIsMutable();
        metadata_.remove(index);
        onChanged();
      } else {
        metadataBuilder_.remove(index);
      }
      return this;
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public com.google.genomics.v1.VariantSetMetadata.Builder getMetadataBuilder(
        int index) {
      return getMetadataFieldBuilder().getBuilder(index);
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public com.google.genomics.v1.VariantSetMetadataOrBuilder getMetadataOrBuilder(
        int index) {
      if (metadataBuilder_ == null) {
        return metadata_.get(index);  } else {
        return metadataBuilder_.getMessageOrBuilder(index);
      }
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public java.util.List<? extends com.google.genomics.v1.VariantSetMetadataOrBuilder> 
         getMetadataOrBuilderList() {
      if (metadataBuilder_ != null) {
        return metadataBuilder_.getMessageOrBuilderList();
      } else {
        return java.util.Collections.unmodifiableList(metadata_);
      }
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public com.google.genomics.v1.VariantSetMetadata.Builder addMetadataBuilder() {
      return getMetadataFieldBuilder().addBuilder(
          com.google.genomics.v1.VariantSetMetadata.getDefaultInstance());
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public com.google.genomics.v1.VariantSetMetadata.Builder addMetadataBuilder(
        int index) {
      return getMetadataFieldBuilder().addBuilder(
          index, com.google.genomics.v1.VariantSetMetadata.getDefaultInstance());
    }
    /**
     * <pre>
     * The metadata associated with this variant set.
     * </pre>
     *
     * <code>repeated .google.genomics.v1.VariantSetMetadata metadata = 4;</code>
     */
    public java.util.List<com.google.genomics.v1.VariantSetMetadata.Builder> 
         getMetadataBuilderList() {
      return getMetadataFieldBuilder().getBuilderList();
    }
    private com.google.protobuf.RepeatedFieldBuilderV3<
        com.google.genomics.v1.VariantSetMetadata, com.google.genomics.v1.VariantSetMetadata.Builder, com.google.genomics.v1.VariantSetMetadataOrBuilder> 
        getMetadataFieldBuilder() {
      if (metadataBuilder_ == null) {
        metadataBuilder_ = new com.google.protobuf.RepeatedFieldBuilderV3<
            com.google.genomics.v1.VariantSetMetadata, com.google.genomics.v1.VariantSetMetadata.Builder, com.google.genomics.v1.VariantSetMetadataOrBuilder>(
                metadata_,
                ((bitField0_ & 0x00000010) != 0),
                getParentForChildren(),
                isClean());
        metadata_ = null;
      }
      return metadataBuilder_;
    }

    private java.lang.Object name_ = "";
    /**
     * <pre>
     * User-specified, mutable name.
     * </pre>
     *
     * <code>string name = 7;</code>
     * @return The name.
     */
    public java.lang.String getName() {
      java.lang.Object ref = name_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        name_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * User-specified, mutable name.
     * </pre>
     *
     * <code>string name = 7;</code>
     * @return The bytes for name.
     */
    public com.google.protobuf.ByteString
        getNameBytes() {
      java.lang.Object ref = name_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        name_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * User-specified, mutable name.
     * </pre>
     *
     * <code>string name = 7;</code>
     * @param value The name to set.
     * @return This builder for chaining.
     */
    public Builder setName(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      name_ = value;
      bitField0_ |= 0x00000020;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * User-specified, mutable name.
     * </pre>
     *
     * <code>string name = 7;</code>
     * @return This builder for chaining.
     */
    public Builder clearName() {
      name_ = getDefaultInstance().getName();
      bitField0_ = (bitField0_ & ~0x00000020);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * User-specified, mutable name.
     * </pre>
     *
     * <code>string name = 7;</code>
     * @param value The bytes for name to set.
     * @return This builder for chaining.
     */
    public Builder setNameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      name_ = value;
      bitField0_ |= 0x00000020;
      onChanged();
      return this;
    }

    private java.lang.Object description_ = "";
    /**
     * <pre>
     * A textual description of this variant set.
     * </pre>
     *
     * <code>string description = 8;</code>
     * @return The description.
     */
    public java.lang.String getDescription() {
      java.lang.Object ref = description_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        description_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * A textual description of this variant set.
     * </pre>
     *
     * <code>string description = 8;</code>
     * @return The bytes for description.
     */
    public com.google.protobuf.ByteString
        getDescriptionBytes() {
      java.lang.Object ref = description_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        description_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * A textual description of this variant set.
     * </pre>
     *
     * <code>string description = 8;</code>
     * @param value The description to set.
     * @return This builder for chaining.
     */
    public Builder setDescription(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      description_ = value;
      bitField0_ |= 0x00000040;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * A textual description of this variant set.
     * </pre>
     *
     * <code>string description = 8;</code>
     * @return This builder for chaining.
     */
    public Builder clearDescription() {
      description_ = getDefaultInstance().getDescription();
      bitField0_ = (bitField0_ & ~0x00000040);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * A textual description of this variant set.
     * </pre>
     *
     * <code>string description = 8;</code>
     * @param value The bytes for description to set.
     * @return This builder for chaining.
     */
    public Builder setDescriptionBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      description_ = value;
      bitField0_ |= 0x00000040;
      onChanged();
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.VariantSet)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.VariantSet)
  private static final com.google.genomics.v1.VariantSet DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.VariantSet();
  }

  public static com.google.genomics.v1.VariantSet getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<VariantSet>
      PARSER = new com.google.protobuf.AbstractParser<VariantSet>() {
    @java.lang.Override
    public VariantSet parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      Builder builder = newBuilder();
      try {
        builder.mergeFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(builder.buildPartial());
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(builder.buildPartial());
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(e)
            .setUnfinishedMessage(builder.buildPartial());
      }
      return builder.buildPartial();
    }
  };

  public static com.google.protobuf.Parser<VariantSet> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<VariantSet> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.VariantSet getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

