// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: google/genomics/v1/variants.proto

package com.google.genomics.v1;

/**
 * <pre>
 * A call represents the determination of genotype with respect to a particular
 * variant. It may include associated information such as quality and phasing.
 * For example, a call might assign a probability of 0.32 to the occurrence of
 * a SNP named rs1234 in a call set with the name NA12345.
 * </pre>
 *
 * Protobuf type {@code google.genomics.v1.VariantCall}
 */
public final class VariantCall extends
    com.google.protobuf.GeneratedMessageV3 implements
    // @@protoc_insertion_point(message_implements:google.genomics.v1.VariantCall)
    VariantCallOrBuilder {
private static final long serialVersionUID = 0L;
  // Use VariantCall.newBuilder() to construct.
  private VariantCall(com.google.protobuf.GeneratedMessageV3.Builder<?> builder) {
    super(builder);
  }
  private VariantCall() {
    callSetId_ = "";
    callSetName_ = "";
    genotype_ = emptyIntList();
    phaseset_ = "";
    genotypeLikelihood_ = emptyDoubleList();
  }

  @java.lang.Override
  @SuppressWarnings({"unused"})
  protected java.lang.Object newInstance(
      UnusedPrivateParameter unused) {
    return new VariantCall();
  }

  @java.lang.Override
  public final com.google.protobuf.UnknownFieldSet
  getUnknownFields() {
    return this.unknownFields;
  }
  public static final com.google.protobuf.Descriptors.Descriptor
      getDescriptor() {
    return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantCall_descriptor;
  }

  @SuppressWarnings({"rawtypes"})
  @java.lang.Override
  protected com.google.protobuf.MapField internalGetMapField(
      int number) {
    switch (number) {
      case 2:
        return internalGetInfo();
      default:
        throw new RuntimeException(
            "Invalid map field number: " + number);
    }
  }
  @java.lang.Override
  protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
      internalGetFieldAccessorTable() {
    return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantCall_fieldAccessorTable
        .ensureFieldAccessorsInitialized(
            com.google.genomics.v1.VariantCall.class, com.google.genomics.v1.VariantCall.Builder.class);
  }

  public static final int CALL_SET_ID_FIELD_NUMBER = 8;
  @SuppressWarnings("serial")
  private volatile java.lang.Object callSetId_ = "";
  /**
   * <pre>
   * The ID of the call set this variant call belongs to.
   * </pre>
   *
   * <code>string call_set_id = 8;</code>
   * @return The callSetId.
   */
  @java.lang.Override
  public java.lang.String getCallSetId() {
    java.lang.Object ref = callSetId_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      callSetId_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The ID of the call set this variant call belongs to.
   * </pre>
   *
   * <code>string call_set_id = 8;</code>
   * @return The bytes for callSetId.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getCallSetIdBytes() {
    java.lang.Object ref = callSetId_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      callSetId_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int CALL_SET_NAME_FIELD_NUMBER = 9;
  @SuppressWarnings("serial")
  private volatile java.lang.Object callSetName_ = "";
  /**
   * <pre>
   * The name of the call set this variant call belongs to.
   * </pre>
   *
   * <code>string call_set_name = 9;</code>
   * @return The callSetName.
   */
  @java.lang.Override
  public java.lang.String getCallSetName() {
    java.lang.Object ref = callSetName_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      callSetName_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * The name of the call set this variant call belongs to.
   * </pre>
   *
   * <code>string call_set_name = 9;</code>
   * @return The bytes for callSetName.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getCallSetNameBytes() {
    java.lang.Object ref = callSetName_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      callSetName_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int GENOTYPE_FIELD_NUMBER = 7;
  @SuppressWarnings("serial")
  private com.google.protobuf.Internal.IntList genotype_;
  /**
   * <pre>
   * The genotype of this variant call. Each value represents either the value
   * of the `referenceBases` field or a 1-based index into
   * `alternateBases`. If a variant had a `referenceBases`
   * value of `T` and an `alternateBases`
   * value of `["A", "C"]`, and the `genotype` was
   * `[2, 1]`, that would mean the call
   * represented the heterozygous value `CA` for this variant.
   * If the `genotype` was instead `[0, 1]`, the
   * represented value would be `TA`. Ordering of the
   * genotype values is important if the `phaseset` is present.
   * If a genotype is not called (that is, a `.` is present in the
   * GT string) -1 is returned.
   * </pre>
   *
   * <code>repeated int32 genotype = 7;</code>
   * @return A list containing the genotype.
   */
  @java.lang.Override
  public java.util.List<java.lang.Integer>
      getGenotypeList() {
    return genotype_;
  }
  /**
   * <pre>
   * The genotype of this variant call. Each value represents either the value
   * of the `referenceBases` field or a 1-based index into
   * `alternateBases`. If a variant had a `referenceBases`
   * value of `T` and an `alternateBases`
   * value of `["A", "C"]`, and the `genotype` was
   * `[2, 1]`, that would mean the call
   * represented the heterozygous value `CA` for this variant.
   * If the `genotype` was instead `[0, 1]`, the
   * represented value would be `TA`. Ordering of the
   * genotype values is important if the `phaseset` is present.
   * If a genotype is not called (that is, a `.` is present in the
   * GT string) -1 is returned.
   * </pre>
   *
   * <code>repeated int32 genotype = 7;</code>
   * @return The count of genotype.
   */
  public int getGenotypeCount() {
    return genotype_.size();
  }
  /**
   * <pre>
   * The genotype of this variant call. Each value represents either the value
   * of the `referenceBases` field or a 1-based index into
   * `alternateBases`. If a variant had a `referenceBases`
   * value of `T` and an `alternateBases`
   * value of `["A", "C"]`, and the `genotype` was
   * `[2, 1]`, that would mean the call
   * represented the heterozygous value `CA` for this variant.
   * If the `genotype` was instead `[0, 1]`, the
   * represented value would be `TA`. Ordering of the
   * genotype values is important if the `phaseset` is present.
   * If a genotype is not called (that is, a `.` is present in the
   * GT string) -1 is returned.
   * </pre>
   *
   * <code>repeated int32 genotype = 7;</code>
   * @param index The index of the element to return.
   * @return The genotype at the given index.
   */
  public int getGenotype(int index) {
    return genotype_.getInt(index);
  }
  private int genotypeMemoizedSerializedSize = -1;

  public static final int PHASESET_FIELD_NUMBER = 5;
  @SuppressWarnings("serial")
  private volatile java.lang.Object phaseset_ = "";
  /**
   * <pre>
   * If this field is present, this variant call's genotype ordering implies
   * the phase of the bases and is consistent with any other variant calls in
   * the same reference sequence which have the same phaseset value.
   * When importing data from VCF, if the genotype data was phased but no
   * phase set was specified this field will be set to `*`.
   * </pre>
   *
   * <code>string phaseset = 5;</code>
   * @return The phaseset.
   */
  @java.lang.Override
  public java.lang.String getPhaseset() {
    java.lang.Object ref = phaseset_;
    if (ref instanceof java.lang.String) {
      return (java.lang.String) ref;
    } else {
      com.google.protobuf.ByteString bs = 
          (com.google.protobuf.ByteString) ref;
      java.lang.String s = bs.toStringUtf8();
      phaseset_ = s;
      return s;
    }
  }
  /**
   * <pre>
   * If this field is present, this variant call's genotype ordering implies
   * the phase of the bases and is consistent with any other variant calls in
   * the same reference sequence which have the same phaseset value.
   * When importing data from VCF, if the genotype data was phased but no
   * phase set was specified this field will be set to `*`.
   * </pre>
   *
   * <code>string phaseset = 5;</code>
   * @return The bytes for phaseset.
   */
  @java.lang.Override
  public com.google.protobuf.ByteString
      getPhasesetBytes() {
    java.lang.Object ref = phaseset_;
    if (ref instanceof java.lang.String) {
      com.google.protobuf.ByteString b = 
          com.google.protobuf.ByteString.copyFromUtf8(
              (java.lang.String) ref);
      phaseset_ = b;
      return b;
    } else {
      return (com.google.protobuf.ByteString) ref;
    }
  }

  public static final int GENOTYPE_LIKELIHOOD_FIELD_NUMBER = 6;
  @SuppressWarnings("serial")
  private com.google.protobuf.Internal.DoubleList genotypeLikelihood_;
  /**
   * <pre>
   * The genotype likelihoods for this variant call. Each array entry
   * represents how likely a specific genotype is for this call. The value
   * ordering is defined by the GL tag in the VCF spec.
   * If Phred-scaled genotype likelihood scores (PL) are available and
   * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
   * to GL scores.  If both are available, PL scores are stored in `info`.
   * </pre>
   *
   * <code>repeated double genotype_likelihood = 6;</code>
   * @return A list containing the genotypeLikelihood.
   */
  @java.lang.Override
  public java.util.List<java.lang.Double>
      getGenotypeLikelihoodList() {
    return genotypeLikelihood_;
  }
  /**
   * <pre>
   * The genotype likelihoods for this variant call. Each array entry
   * represents how likely a specific genotype is for this call. The value
   * ordering is defined by the GL tag in the VCF spec.
   * If Phred-scaled genotype likelihood scores (PL) are available and
   * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
   * to GL scores.  If both are available, PL scores are stored in `info`.
   * </pre>
   *
   * <code>repeated double genotype_likelihood = 6;</code>
   * @return The count of genotypeLikelihood.
   */
  public int getGenotypeLikelihoodCount() {
    return genotypeLikelihood_.size();
  }
  /**
   * <pre>
   * The genotype likelihoods for this variant call. Each array entry
   * represents how likely a specific genotype is for this call. The value
   * ordering is defined by the GL tag in the VCF spec.
   * If Phred-scaled genotype likelihood scores (PL) are available and
   * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
   * to GL scores.  If both are available, PL scores are stored in `info`.
   * </pre>
   *
   * <code>repeated double genotype_likelihood = 6;</code>
   * @param index The index of the element to return.
   * @return The genotypeLikelihood at the given index.
   */
  public double getGenotypeLikelihood(int index) {
    return genotypeLikelihood_.getDouble(index);
  }
  private int genotypeLikelihoodMemoizedSerializedSize = -1;

  public static final int INFO_FIELD_NUMBER = 2;
  private static final class InfoDefaultEntryHolder {
    static final com.google.protobuf.MapEntry<
        java.lang.String, com.google.protobuf.ListValue> defaultEntry =
            com.google.protobuf.MapEntry
            .<java.lang.String, com.google.protobuf.ListValue>newDefaultInstance(
                com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantCall_InfoEntry_descriptor, 
                com.google.protobuf.WireFormat.FieldType.STRING,
                "",
                com.google.protobuf.WireFormat.FieldType.MESSAGE,
                com.google.protobuf.ListValue.getDefaultInstance());
  }
  @SuppressWarnings("serial")
  private com.google.protobuf.MapField<
      java.lang.String, com.google.protobuf.ListValue> info_;
  private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
  internalGetInfo() {
    if (info_ == null) {
      return com.google.protobuf.MapField.emptyMapField(
          InfoDefaultEntryHolder.defaultEntry);
    }
    return info_;
  }
  public int getInfoCount() {
    return internalGetInfo().getMap().size();
  }
  /**
   * <pre>
   * A map of additional variant call information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
   */
  @java.lang.Override
  public boolean containsInfo(
      java.lang.String key) {
    if (key == null) { throw new NullPointerException("map key"); }
    return internalGetInfo().getMap().containsKey(key);
  }
  /**
   * Use {@link #getInfoMap()} instead.
   */
  @java.lang.Override
  @java.lang.Deprecated
  public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfo() {
    return getInfoMap();
  }
  /**
   * <pre>
   * A map of additional variant call information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
   */
  @java.lang.Override
  public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfoMap() {
    return internalGetInfo().getMap();
  }
  /**
   * <pre>
   * A map of additional variant call information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
   */
  @java.lang.Override
  public /* nullable */
com.google.protobuf.ListValue getInfoOrDefault(
      java.lang.String key,
      /* nullable */
com.google.protobuf.ListValue defaultValue) {
    if (key == null) { throw new NullPointerException("map key"); }
    java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
        internalGetInfo().getMap();
    return map.containsKey(key) ? map.get(key) : defaultValue;
  }
  /**
   * <pre>
   * A map of additional variant call information. This must be of the form
   * map&lt;string, string[]&gt; (string key mapping to a list of string values).
   * </pre>
   *
   * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
   */
  @java.lang.Override
  public com.google.protobuf.ListValue getInfoOrThrow(
      java.lang.String key) {
    if (key == null) { throw new NullPointerException("map key"); }
    java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
        internalGetInfo().getMap();
    if (!map.containsKey(key)) {
      throw new java.lang.IllegalArgumentException();
    }
    return map.get(key);
  }

  private byte memoizedIsInitialized = -1;
  @java.lang.Override
  public final boolean isInitialized() {
    byte isInitialized = memoizedIsInitialized;
    if (isInitialized == 1) return true;
    if (isInitialized == 0) return false;

    memoizedIsInitialized = 1;
    return true;
  }

  @java.lang.Override
  public void writeTo(com.google.protobuf.CodedOutputStream output)
                      throws java.io.IOException {
    getSerializedSize();
    com.google.protobuf.GeneratedMessageV3
      .serializeStringMapTo(
        output,
        internalGetInfo(),
        InfoDefaultEntryHolder.defaultEntry,
        2);
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(phaseset_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 5, phaseset_);
    }
    if (getGenotypeLikelihoodList().size() > 0) {
      output.writeUInt32NoTag(50);
      output.writeUInt32NoTag(genotypeLikelihoodMemoizedSerializedSize);
    }
    for (int i = 0; i < genotypeLikelihood_.size(); i++) {
      output.writeDoubleNoTag(genotypeLikelihood_.getDouble(i));
    }
    if (getGenotypeList().size() > 0) {
      output.writeUInt32NoTag(58);
      output.writeUInt32NoTag(genotypeMemoizedSerializedSize);
    }
    for (int i = 0; i < genotype_.size(); i++) {
      output.writeInt32NoTag(genotype_.getInt(i));
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(callSetId_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 8, callSetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(callSetName_)) {
      com.google.protobuf.GeneratedMessageV3.writeString(output, 9, callSetName_);
    }
    getUnknownFields().writeTo(output);
  }

  @java.lang.Override
  public int getSerializedSize() {
    int size = memoizedSize;
    if (size != -1) return size;

    size = 0;
    for (java.util.Map.Entry<java.lang.String, com.google.protobuf.ListValue> entry
         : internalGetInfo().getMap().entrySet()) {
      com.google.protobuf.MapEntry<java.lang.String, com.google.protobuf.ListValue>
      info__ = InfoDefaultEntryHolder.defaultEntry.newBuilderForType()
          .setKey(entry.getKey())
          .setValue(entry.getValue())
          .build();
      size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(2, info__);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(phaseset_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(5, phaseset_);
    }
    {
      int dataSize = 0;
      dataSize = 8 * getGenotypeLikelihoodList().size();
      size += dataSize;
      if (!getGenotypeLikelihoodList().isEmpty()) {
        size += 1;
        size += com.google.protobuf.CodedOutputStream
            .computeInt32SizeNoTag(dataSize);
      }
      genotypeLikelihoodMemoizedSerializedSize = dataSize;
    }
    {
      int dataSize = 0;
      for (int i = 0; i < genotype_.size(); i++) {
        dataSize += com.google.protobuf.CodedOutputStream
          .computeInt32SizeNoTag(genotype_.getInt(i));
      }
      size += dataSize;
      if (!getGenotypeList().isEmpty()) {
        size += 1;
        size += com.google.protobuf.CodedOutputStream
            .computeInt32SizeNoTag(dataSize);
      }
      genotypeMemoizedSerializedSize = dataSize;
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(callSetId_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(8, callSetId_);
    }
    if (!com.google.protobuf.GeneratedMessageV3.isStringEmpty(callSetName_)) {
      size += com.google.protobuf.GeneratedMessageV3.computeStringSize(9, callSetName_);
    }
    size += getUnknownFields().getSerializedSize();
    memoizedSize = size;
    return size;
  }

  @java.lang.Override
  public boolean equals(final java.lang.Object obj) {
    if (obj == this) {
     return true;
    }
    if (!(obj instanceof com.google.genomics.v1.VariantCall)) {
      return super.equals(obj);
    }
    com.google.genomics.v1.VariantCall other = (com.google.genomics.v1.VariantCall) obj;

    if (!getCallSetId()
        .equals(other.getCallSetId())) return false;
    if (!getCallSetName()
        .equals(other.getCallSetName())) return false;
    if (!getGenotypeList()
        .equals(other.getGenotypeList())) return false;
    if (!getPhaseset()
        .equals(other.getPhaseset())) return false;
    if (!getGenotypeLikelihoodList()
        .equals(other.getGenotypeLikelihoodList())) return false;
    if (!internalGetInfo().equals(
        other.internalGetInfo())) return false;
    if (!getUnknownFields().equals(other.getUnknownFields())) return false;
    return true;
  }

  @java.lang.Override
  public int hashCode() {
    if (memoizedHashCode != 0) {
      return memoizedHashCode;
    }
    int hash = 41;
    hash = (19 * hash) + getDescriptor().hashCode();
    hash = (37 * hash) + CALL_SET_ID_FIELD_NUMBER;
    hash = (53 * hash) + getCallSetId().hashCode();
    hash = (37 * hash) + CALL_SET_NAME_FIELD_NUMBER;
    hash = (53 * hash) + getCallSetName().hashCode();
    if (getGenotypeCount() > 0) {
      hash = (37 * hash) + GENOTYPE_FIELD_NUMBER;
      hash = (53 * hash) + getGenotypeList().hashCode();
    }
    hash = (37 * hash) + PHASESET_FIELD_NUMBER;
    hash = (53 * hash) + getPhaseset().hashCode();
    if (getGenotypeLikelihoodCount() > 0) {
      hash = (37 * hash) + GENOTYPE_LIKELIHOOD_FIELD_NUMBER;
      hash = (53 * hash) + getGenotypeLikelihoodList().hashCode();
    }
    if (!internalGetInfo().getMap().isEmpty()) {
      hash = (37 * hash) + INFO_FIELD_NUMBER;
      hash = (53 * hash) + internalGetInfo().hashCode();
    }
    hash = (29 * hash) + getUnknownFields().hashCode();
    memoizedHashCode = hash;
    return hash;
  }

  public static com.google.genomics.v1.VariantCall parseFrom(
      java.nio.ByteBuffer data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      java.nio.ByteBuffer data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      com.google.protobuf.ByteString data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      com.google.protobuf.ByteString data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(byte[] data)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      byte[] data,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws com.google.protobuf.InvalidProtocolBufferException {
    return PARSER.parseFrom(data, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantCall parseDelimitedFrom(java.io.InputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.VariantCall parseDelimitedFrom(
      java.io.InputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseDelimitedWithIOException(PARSER, input, extensionRegistry);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      com.google.protobuf.CodedInputStream input)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input);
  }
  public static com.google.genomics.v1.VariantCall parseFrom(
      com.google.protobuf.CodedInputStream input,
      com.google.protobuf.ExtensionRegistryLite extensionRegistry)
      throws java.io.IOException {
    return com.google.protobuf.GeneratedMessageV3
        .parseWithIOException(PARSER, input, extensionRegistry);
  }

  @java.lang.Override
  public Builder newBuilderForType() { return newBuilder(); }
  public static Builder newBuilder() {
    return DEFAULT_INSTANCE.toBuilder();
  }
  public static Builder newBuilder(com.google.genomics.v1.VariantCall prototype) {
    return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
  }
  @java.lang.Override
  public Builder toBuilder() {
    return this == DEFAULT_INSTANCE
        ? new Builder() : new Builder().mergeFrom(this);
  }

  @java.lang.Override
  protected Builder newBuilderForType(
      com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
    Builder builder = new Builder(parent);
    return builder;
  }
  /**
   * <pre>
   * A call represents the determination of genotype with respect to a particular
   * variant. It may include associated information such as quality and phasing.
   * For example, a call might assign a probability of 0.32 to the occurrence of
   * a SNP named rs1234 in a call set with the name NA12345.
   * </pre>
   *
   * Protobuf type {@code google.genomics.v1.VariantCall}
   */
  public static final class Builder extends
      com.google.protobuf.GeneratedMessageV3.Builder<Builder> implements
      // @@protoc_insertion_point(builder_implements:google.genomics.v1.VariantCall)
      com.google.genomics.v1.VariantCallOrBuilder {
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantCall_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 2:
          return internalGetInfo();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMutableMapField(
        int number) {
      switch (number) {
        case 2:
          return internalGetMutableInfo();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    @java.lang.Override
    protected com.google.protobuf.GeneratedMessageV3.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantCall_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              com.google.genomics.v1.VariantCall.class, com.google.genomics.v1.VariantCall.Builder.class);
    }

    // Construct using com.google.genomics.v1.VariantCall.newBuilder()
    private Builder() {

    }

    private Builder(
        com.google.protobuf.GeneratedMessageV3.BuilderParent parent) {
      super(parent);

    }
    @java.lang.Override
    public Builder clear() {
      super.clear();
      bitField0_ = 0;
      callSetId_ = "";
      callSetName_ = "";
      genotype_ = emptyIntList();
      phaseset_ = "";
      genotypeLikelihood_ = emptyDoubleList();
      internalGetMutableInfo().clear();
      return this;
    }

    @java.lang.Override
    public com.google.protobuf.Descriptors.Descriptor
        getDescriptorForType() {
      return com.google.genomics.v1.VariantsProto.internal_static_google_genomics_v1_VariantCall_descriptor;
    }

    @java.lang.Override
    public com.google.genomics.v1.VariantCall getDefaultInstanceForType() {
      return com.google.genomics.v1.VariantCall.getDefaultInstance();
    }

    @java.lang.Override
    public com.google.genomics.v1.VariantCall build() {
      com.google.genomics.v1.VariantCall result = buildPartial();
      if (!result.isInitialized()) {
        throw newUninitializedMessageException(result);
      }
      return result;
    }

    @java.lang.Override
    public com.google.genomics.v1.VariantCall buildPartial() {
      com.google.genomics.v1.VariantCall result = new com.google.genomics.v1.VariantCall(this);
      buildPartialRepeatedFields(result);
      if (bitField0_ != 0) { buildPartial0(result); }
      onBuilt();
      return result;
    }

    private void buildPartialRepeatedFields(com.google.genomics.v1.VariantCall result) {
      if (((bitField0_ & 0x00000004) != 0)) {
        genotype_.makeImmutable();
        bitField0_ = (bitField0_ & ~0x00000004);
      }
      result.genotype_ = genotype_;
      if (((bitField0_ & 0x00000010) != 0)) {
        genotypeLikelihood_.makeImmutable();
        bitField0_ = (bitField0_ & ~0x00000010);
      }
      result.genotypeLikelihood_ = genotypeLikelihood_;
    }

    private void buildPartial0(com.google.genomics.v1.VariantCall result) {
      int from_bitField0_ = bitField0_;
      if (((from_bitField0_ & 0x00000001) != 0)) {
        result.callSetId_ = callSetId_;
      }
      if (((from_bitField0_ & 0x00000002) != 0)) {
        result.callSetName_ = callSetName_;
      }
      if (((from_bitField0_ & 0x00000008) != 0)) {
        result.phaseset_ = phaseset_;
      }
      if (((from_bitField0_ & 0x00000020) != 0)) {
        result.info_ = internalGetInfo();
        result.info_.makeImmutable();
      }
    }

    @java.lang.Override
    public Builder clone() {
      return super.clone();
    }
    @java.lang.Override
    public Builder setField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.setField(field, value);
    }
    @java.lang.Override
    public Builder clearField(
        com.google.protobuf.Descriptors.FieldDescriptor field) {
      return super.clearField(field);
    }
    @java.lang.Override
    public Builder clearOneof(
        com.google.protobuf.Descriptors.OneofDescriptor oneof) {
      return super.clearOneof(oneof);
    }
    @java.lang.Override
    public Builder setRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        int index, java.lang.Object value) {
      return super.setRepeatedField(field, index, value);
    }
    @java.lang.Override
    public Builder addRepeatedField(
        com.google.protobuf.Descriptors.FieldDescriptor field,
        java.lang.Object value) {
      return super.addRepeatedField(field, value);
    }
    @java.lang.Override
    public Builder mergeFrom(com.google.protobuf.Message other) {
      if (other instanceof com.google.genomics.v1.VariantCall) {
        return mergeFrom((com.google.genomics.v1.VariantCall)other);
      } else {
        super.mergeFrom(other);
        return this;
      }
    }

    public Builder mergeFrom(com.google.genomics.v1.VariantCall other) {
      if (other == com.google.genomics.v1.VariantCall.getDefaultInstance()) return this;
      if (!other.getCallSetId().isEmpty()) {
        callSetId_ = other.callSetId_;
        bitField0_ |= 0x00000001;
        onChanged();
      }
      if (!other.getCallSetName().isEmpty()) {
        callSetName_ = other.callSetName_;
        bitField0_ |= 0x00000002;
        onChanged();
      }
      if (!other.genotype_.isEmpty()) {
        if (genotype_.isEmpty()) {
          genotype_ = other.genotype_;
          bitField0_ = (bitField0_ & ~0x00000004);
        } else {
          ensureGenotypeIsMutable();
          genotype_.addAll(other.genotype_);
        }
        onChanged();
      }
      if (!other.getPhaseset().isEmpty()) {
        phaseset_ = other.phaseset_;
        bitField0_ |= 0x00000008;
        onChanged();
      }
      if (!other.genotypeLikelihood_.isEmpty()) {
        if (genotypeLikelihood_.isEmpty()) {
          genotypeLikelihood_ = other.genotypeLikelihood_;
          bitField0_ = (bitField0_ & ~0x00000010);
        } else {
          ensureGenotypeLikelihoodIsMutable();
          genotypeLikelihood_.addAll(other.genotypeLikelihood_);
        }
        onChanged();
      }
      internalGetMutableInfo().mergeFrom(
          other.internalGetInfo());
      bitField0_ |= 0x00000020;
      this.mergeUnknownFields(other.getUnknownFields());
      onChanged();
      return this;
    }

    @java.lang.Override
    public final boolean isInitialized() {
      return true;
    }

    @java.lang.Override
    public Builder mergeFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      if (extensionRegistry == null) {
        throw new java.lang.NullPointerException();
      }
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            case 18: {
              com.google.protobuf.MapEntry<java.lang.String, com.google.protobuf.ListValue>
              info__ = input.readMessage(
                  InfoDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              internalGetMutableInfo().getMutableMap().put(
                  info__.getKey(), info__.getValue());
              bitField0_ |= 0x00000020;
              break;
            } // case 18
            case 42: {
              phaseset_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000008;
              break;
            } // case 42
            case 49: {
              double v = input.readDouble();
              ensureGenotypeLikelihoodIsMutable();
              genotypeLikelihood_.addDouble(v);
              break;
            } // case 49
            case 50: {
              int length = input.readRawVarint32();
              int limit = input.pushLimit(length);
              ensureGenotypeLikelihoodIsMutable();
              while (input.getBytesUntilLimit() > 0) {
                genotypeLikelihood_.addDouble(input.readDouble());
              }
              input.popLimit(limit);
              break;
            } // case 50
            case 56: {
              int v = input.readInt32();
              ensureGenotypeIsMutable();
              genotype_.addInt(v);
              break;
            } // case 56
            case 58: {
              int length = input.readRawVarint32();
              int limit = input.pushLimit(length);
              ensureGenotypeIsMutable();
              while (input.getBytesUntilLimit() > 0) {
                genotype_.addInt(input.readInt32());
              }
              input.popLimit(limit);
              break;
            } // case 58
            case 66: {
              callSetId_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000001;
              break;
            } // case 66
            case 74: {
              callSetName_ = input.readStringRequireUtf8();
              bitField0_ |= 0x00000002;
              break;
            } // case 74
            default: {
              if (!super.parseUnknownField(input, extensionRegistry, tag)) {
                done = true; // was an endgroup tag
              }
              break;
            } // default:
          } // switch (tag)
        } // while (!done)
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.unwrapIOException();
      } finally {
        onChanged();
      } // finally
      return this;
    }
    private int bitField0_;

    private java.lang.Object callSetId_ = "";
    /**
     * <pre>
     * The ID of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_id = 8;</code>
     * @return The callSetId.
     */
    public java.lang.String getCallSetId() {
      java.lang.Object ref = callSetId_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        callSetId_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The ID of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_id = 8;</code>
     * @return The bytes for callSetId.
     */
    public com.google.protobuf.ByteString
        getCallSetIdBytes() {
      java.lang.Object ref = callSetId_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        callSetId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The ID of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_id = 8;</code>
     * @param value The callSetId to set.
     * @return This builder for chaining.
     */
    public Builder setCallSetId(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      callSetId_ = value;
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The ID of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_id = 8;</code>
     * @return This builder for chaining.
     */
    public Builder clearCallSetId() {
      callSetId_ = getDefaultInstance().getCallSetId();
      bitField0_ = (bitField0_ & ~0x00000001);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The ID of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_id = 8;</code>
     * @param value The bytes for callSetId to set.
     * @return This builder for chaining.
     */
    public Builder setCallSetIdBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      callSetId_ = value;
      bitField0_ |= 0x00000001;
      onChanged();
      return this;
    }

    private java.lang.Object callSetName_ = "";
    /**
     * <pre>
     * The name of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_name = 9;</code>
     * @return The callSetName.
     */
    public java.lang.String getCallSetName() {
      java.lang.Object ref = callSetName_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        callSetName_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * The name of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_name = 9;</code>
     * @return The bytes for callSetName.
     */
    public com.google.protobuf.ByteString
        getCallSetNameBytes() {
      java.lang.Object ref = callSetName_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        callSetName_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * The name of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_name = 9;</code>
     * @param value The callSetName to set.
     * @return This builder for chaining.
     */
    public Builder setCallSetName(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      callSetName_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The name of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_name = 9;</code>
     * @return This builder for chaining.
     */
    public Builder clearCallSetName() {
      callSetName_ = getDefaultInstance().getCallSetName();
      bitField0_ = (bitField0_ & ~0x00000002);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The name of the call set this variant call belongs to.
     * </pre>
     *
     * <code>string call_set_name = 9;</code>
     * @param value The bytes for callSetName to set.
     * @return This builder for chaining.
     */
    public Builder setCallSetNameBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      callSetName_ = value;
      bitField0_ |= 0x00000002;
      onChanged();
      return this;
    }

    private com.google.protobuf.Internal.IntList genotype_ = emptyIntList();
    private void ensureGenotypeIsMutable() {
      if (!((bitField0_ & 0x00000004) != 0)) {
        genotype_ = mutableCopy(genotype_);
        bitField0_ |= 0x00000004;
      }
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @return A list containing the genotype.
     */
    public java.util.List<java.lang.Integer>
        getGenotypeList() {
      return ((bitField0_ & 0x00000004) != 0) ?
               java.util.Collections.unmodifiableList(genotype_) : genotype_;
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @return The count of genotype.
     */
    public int getGenotypeCount() {
      return genotype_.size();
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @param index The index of the element to return.
     * @return The genotype at the given index.
     */
    public int getGenotype(int index) {
      return genotype_.getInt(index);
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @param index The index to set the value at.
     * @param value The genotype to set.
     * @return This builder for chaining.
     */
    public Builder setGenotype(
        int index, int value) {
      
      ensureGenotypeIsMutable();
      genotype_.setInt(index, value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @param value The genotype to add.
     * @return This builder for chaining.
     */
    public Builder addGenotype(int value) {
      
      ensureGenotypeIsMutable();
      genotype_.addInt(value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @param values The genotype to add.
     * @return This builder for chaining.
     */
    public Builder addAllGenotype(
        java.lang.Iterable<? extends java.lang.Integer> values) {
      ensureGenotypeIsMutable();
      com.google.protobuf.AbstractMessageLite.Builder.addAll(
          values, genotype_);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genotype of this variant call. Each value represents either the value
     * of the `referenceBases` field or a 1-based index into
     * `alternateBases`. If a variant had a `referenceBases`
     * value of `T` and an `alternateBases`
     * value of `["A", "C"]`, and the `genotype` was
     * `[2, 1]`, that would mean the call
     * represented the heterozygous value `CA` for this variant.
     * If the `genotype` was instead `[0, 1]`, the
     * represented value would be `TA`. Ordering of the
     * genotype values is important if the `phaseset` is present.
     * If a genotype is not called (that is, a `.` is present in the
     * GT string) -1 is returned.
     * </pre>
     *
     * <code>repeated int32 genotype = 7;</code>
     * @return This builder for chaining.
     */
    public Builder clearGenotype() {
      genotype_ = emptyIntList();
      bitField0_ = (bitField0_ & ~0x00000004);
      onChanged();
      return this;
    }

    private java.lang.Object phaseset_ = "";
    /**
     * <pre>
     * If this field is present, this variant call's genotype ordering implies
     * the phase of the bases and is consistent with any other variant calls in
     * the same reference sequence which have the same phaseset value.
     * When importing data from VCF, if the genotype data was phased but no
     * phase set was specified this field will be set to `*`.
     * </pre>
     *
     * <code>string phaseset = 5;</code>
     * @return The phaseset.
     */
    public java.lang.String getPhaseset() {
      java.lang.Object ref = phaseset_;
      if (!(ref instanceof java.lang.String)) {
        com.google.protobuf.ByteString bs =
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        phaseset_ = s;
        return s;
      } else {
        return (java.lang.String) ref;
      }
    }
    /**
     * <pre>
     * If this field is present, this variant call's genotype ordering implies
     * the phase of the bases and is consistent with any other variant calls in
     * the same reference sequence which have the same phaseset value.
     * When importing data from VCF, if the genotype data was phased but no
     * phase set was specified this field will be set to `*`.
     * </pre>
     *
     * <code>string phaseset = 5;</code>
     * @return The bytes for phaseset.
     */
    public com.google.protobuf.ByteString
        getPhasesetBytes() {
      java.lang.Object ref = phaseset_;
      if (ref instanceof String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        phaseset_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }
    /**
     * <pre>
     * If this field is present, this variant call's genotype ordering implies
     * the phase of the bases and is consistent with any other variant calls in
     * the same reference sequence which have the same phaseset value.
     * When importing data from VCF, if the genotype data was phased but no
     * phase set was specified this field will be set to `*`.
     * </pre>
     *
     * <code>string phaseset = 5;</code>
     * @param value The phaseset to set.
     * @return This builder for chaining.
     */
    public Builder setPhaseset(
        java.lang.String value) {
      if (value == null) { throw new NullPointerException(); }
      phaseset_ = value;
      bitField0_ |= 0x00000008;
      onChanged();
      return this;
    }
    /**
     * <pre>
     * If this field is present, this variant call's genotype ordering implies
     * the phase of the bases and is consistent with any other variant calls in
     * the same reference sequence which have the same phaseset value.
     * When importing data from VCF, if the genotype data was phased but no
     * phase set was specified this field will be set to `*`.
     * </pre>
     *
     * <code>string phaseset = 5;</code>
     * @return This builder for chaining.
     */
    public Builder clearPhaseset() {
      phaseset_ = getDefaultInstance().getPhaseset();
      bitField0_ = (bitField0_ & ~0x00000008);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * If this field is present, this variant call's genotype ordering implies
     * the phase of the bases and is consistent with any other variant calls in
     * the same reference sequence which have the same phaseset value.
     * When importing data from VCF, if the genotype data was phased but no
     * phase set was specified this field will be set to `*`.
     * </pre>
     *
     * <code>string phaseset = 5;</code>
     * @param value The bytes for phaseset to set.
     * @return This builder for chaining.
     */
    public Builder setPhasesetBytes(
        com.google.protobuf.ByteString value) {
      if (value == null) { throw new NullPointerException(); }
      checkByteStringIsUtf8(value);
      phaseset_ = value;
      bitField0_ |= 0x00000008;
      onChanged();
      return this;
    }

    private com.google.protobuf.Internal.DoubleList genotypeLikelihood_ = emptyDoubleList();
    private void ensureGenotypeLikelihoodIsMutable() {
      if (!((bitField0_ & 0x00000010) != 0)) {
        genotypeLikelihood_ = mutableCopy(genotypeLikelihood_);
        bitField0_ |= 0x00000010;
      }
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @return A list containing the genotypeLikelihood.
     */
    public java.util.List<java.lang.Double>
        getGenotypeLikelihoodList() {
      return ((bitField0_ & 0x00000010) != 0) ?
               java.util.Collections.unmodifiableList(genotypeLikelihood_) : genotypeLikelihood_;
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @return The count of genotypeLikelihood.
     */
    public int getGenotypeLikelihoodCount() {
      return genotypeLikelihood_.size();
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @param index The index of the element to return.
     * @return The genotypeLikelihood at the given index.
     */
    public double getGenotypeLikelihood(int index) {
      return genotypeLikelihood_.getDouble(index);
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @param index The index to set the value at.
     * @param value The genotypeLikelihood to set.
     * @return This builder for chaining.
     */
    public Builder setGenotypeLikelihood(
        int index, double value) {
      
      ensureGenotypeLikelihoodIsMutable();
      genotypeLikelihood_.setDouble(index, value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @param value The genotypeLikelihood to add.
     * @return This builder for chaining.
     */
    public Builder addGenotypeLikelihood(double value) {
      
      ensureGenotypeLikelihoodIsMutable();
      genotypeLikelihood_.addDouble(value);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @param values The genotypeLikelihood to add.
     * @return This builder for chaining.
     */
    public Builder addAllGenotypeLikelihood(
        java.lang.Iterable<? extends java.lang.Double> values) {
      ensureGenotypeLikelihoodIsMutable();
      com.google.protobuf.AbstractMessageLite.Builder.addAll(
          values, genotypeLikelihood_);
      onChanged();
      return this;
    }
    /**
     * <pre>
     * The genotype likelihoods for this variant call. Each array entry
     * represents how likely a specific genotype is for this call. The value
     * ordering is defined by the GL tag in the VCF spec.
     * If Phred-scaled genotype likelihood scores (PL) are available and
     * log10(P) genotype likelihood scores (GL) are not, PL scores are converted
     * to GL scores.  If both are available, PL scores are stored in `info`.
     * </pre>
     *
     * <code>repeated double genotype_likelihood = 6;</code>
     * @return This builder for chaining.
     */
    public Builder clearGenotypeLikelihood() {
      genotypeLikelihood_ = emptyDoubleList();
      bitField0_ = (bitField0_ & ~0x00000010);
      onChanged();
      return this;
    }

    private com.google.protobuf.MapField<
        java.lang.String, com.google.protobuf.ListValue> info_;
    private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
        internalGetInfo() {
      if (info_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            InfoDefaultEntryHolder.defaultEntry);
      }
      return info_;
    }
    private com.google.protobuf.MapField<java.lang.String, com.google.protobuf.ListValue>
        internalGetMutableInfo() {
      if (info_ == null) {
        info_ = com.google.protobuf.MapField.newMapField(
            InfoDefaultEntryHolder.defaultEntry);
      }
      if (!info_.isMutable()) {
        info_ = info_.copy();
      }
      bitField0_ |= 0x00000020;
      onChanged();
      return info_;
    }
    public int getInfoCount() {
      return internalGetInfo().getMap().size();
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    @java.lang.Override
    public boolean containsInfo(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      return internalGetInfo().getMap().containsKey(key);
    }
    /**
     * Use {@link #getInfoMap()} instead.
     */
    @java.lang.Override
    @java.lang.Deprecated
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfo() {
      return getInfoMap();
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    @java.lang.Override
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue> getInfoMap() {
      return internalGetInfo().getMap();
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    @java.lang.Override
    public /* nullable */
com.google.protobuf.ListValue getInfoOrDefault(
        java.lang.String key,
        /* nullable */
com.google.protobuf.ListValue defaultValue) {
      if (key == null) { throw new NullPointerException("map key"); }
      java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
          internalGetInfo().getMap();
      return map.containsKey(key) ? map.get(key) : defaultValue;
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    @java.lang.Override
    public com.google.protobuf.ListValue getInfoOrThrow(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      java.util.Map<java.lang.String, com.google.protobuf.ListValue> map =
          internalGetInfo().getMap();
      if (!map.containsKey(key)) {
        throw new java.lang.IllegalArgumentException();
      }
      return map.get(key);
    }
    public Builder clearInfo() {
      bitField0_ = (bitField0_ & ~0x00000020);
      internalGetMutableInfo().getMutableMap()
          .clear();
      return this;
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    public Builder removeInfo(
        java.lang.String key) {
      if (key == null) { throw new NullPointerException("map key"); }
      internalGetMutableInfo().getMutableMap()
          .remove(key);
      return this;
    }
    /**
     * Use alternate mutation accessors instead.
     */
    @java.lang.Deprecated
    public java.util.Map<java.lang.String, com.google.protobuf.ListValue>
        getMutableInfo() {
      bitField0_ |= 0x00000020;
      return internalGetMutableInfo().getMutableMap();
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    public Builder putInfo(
        java.lang.String key,
        com.google.protobuf.ListValue value) {
      if (key == null) { throw new NullPointerException("map key"); }
      if (value == null) { throw new NullPointerException("map value"); }
      internalGetMutableInfo().getMutableMap()
          .put(key, value);
      bitField0_ |= 0x00000020;
      return this;
    }
    /**
     * <pre>
     * A map of additional variant call information. This must be of the form
     * map&lt;string, string[]&gt; (string key mapping to a list of string values).
     * </pre>
     *
     * <code>map&lt;string, .google.protobuf.ListValue&gt; info = 2;</code>
     */
    public Builder putAllInfo(
        java.util.Map<java.lang.String, com.google.protobuf.ListValue> values) {
      internalGetMutableInfo().getMutableMap()
          .putAll(values);
      bitField0_ |= 0x00000020;
      return this;
    }
    @java.lang.Override
    public final Builder setUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.setUnknownFields(unknownFields);
    }

    @java.lang.Override
    public final Builder mergeUnknownFields(
        final com.google.protobuf.UnknownFieldSet unknownFields) {
      return super.mergeUnknownFields(unknownFields);
    }


    // @@protoc_insertion_point(builder_scope:google.genomics.v1.VariantCall)
  }

  // @@protoc_insertion_point(class_scope:google.genomics.v1.VariantCall)
  private static final com.google.genomics.v1.VariantCall DEFAULT_INSTANCE;
  static {
    DEFAULT_INSTANCE = new com.google.genomics.v1.VariantCall();
  }

  public static com.google.genomics.v1.VariantCall getDefaultInstance() {
    return DEFAULT_INSTANCE;
  }

  private static final com.google.protobuf.Parser<VariantCall>
      PARSER = new com.google.protobuf.AbstractParser<VariantCall>() {
    @java.lang.Override
    public VariantCall parsePartialFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      Builder builder = newBuilder();
      try {
        builder.mergeFrom(input, extensionRegistry);
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw e.setUnfinishedMessage(builder.buildPartial());
      } catch (com.google.protobuf.UninitializedMessageException e) {
        throw e.asInvalidProtocolBufferException().setUnfinishedMessage(builder.buildPartial());
      } catch (java.io.IOException e) {
        throw new com.google.protobuf.InvalidProtocolBufferException(e)
            .setUnfinishedMessage(builder.buildPartial());
      }
      return builder.buildPartial();
    }
  };

  public static com.google.protobuf.Parser<VariantCall> parser() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.protobuf.Parser<VariantCall> getParserForType() {
    return PARSER;
  }

  @java.lang.Override
  public com.google.genomics.v1.VariantCall getDefaultInstanceForType() {
    return DEFAULT_INSTANCE;
  }

}

